// molstar/lib/commonjs/extensions/mvs/tree/molstar/conversion.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.mvsSanityCheck = exports.mvsSanityCheckIssues = exports.convertMvsToMolstar = exports.ParseFormatMvsToMolstar = void 0;
var tslib_1 = require("tslib");
var tree_utils_1 = require("../generic/tree-utils");
var mvs_defaults_1 = require("../mvs/mvs-defaults");
var object_1 = require("../../../../mol-util/object");
/** Convert `format` parameter of `parse` node in `MolstarTree`
 * into `format` and `is_binary` parameters in `MolstarTree` */
exports.ParseFormatMvsToMolstar = {
    mmcif: { format: 'cif', is_binary: false },
    bcif: { format: 'cif', is_binary: true },
    pdb: { format: 'pdb', is_binary: false },
};
/** Conversion rules for conversion from `MVSTree` (with all parameter values) to `MolstarTree` */
var mvsToMolstarConversionRules = {
    'download': function (node) { return []; },
    'parse': function (node, parent) {
        var _a = exports.ParseFormatMvsToMolstar[node.params.format], format = _a.format, is_binary = _a.is_binary;
        var convertedNode = { kind: 'parse', params: tslib_1.__assign(tslib_1.__assign({}, node.params), { format: format }) };
        if ((parent === null || parent === void 0 ? void 0 : parent.kind) === 'download') {
            return [
                { kind: 'download', params: tslib_1.__assign(tslib_1.__assign({}, parent.params), { is_binary: is_binary }) },
                convertedNode,
            ];
        }
        else {
            console.warn('"parse" node is not being converted, this is suspicious');
            return [convertedNode];
        }
    },
    'structure': function (node, parent) {
        if ((parent === null || parent === void 0 ? void 0 : parent.kind) !== 'parse')
            throw new Error('Parent of "structure" must be "parse".');
        var format = exports.ParseFormatMvsToMolstar[parent.params.format].format;
        return [
            { kind: 'trajectory', params: tslib_1.__assign({ format: format }, (0, object_1.pickObjectKeys)(node.params, ['block_header', 'block_index'])) },
            { kind: 'model', params: (0, object_1.pickObjectKeys)(node.params, ['model_index']) },
            { kind: 'structure', params: (0, object_1.omitObjectKeys)(node.params, ['block_header', 'block_index', 'model_index']) },
        ];
    },
};
/** Node kinds in `MolstarTree` that it makes sense to condense */
var molstarNodesToCondense = new Set(['download', 'parse', 'trajectory', 'model']);
/** Convert MolViewSpec tree into MolStar tree */
function convertMvsToMolstar(mvsTree, sourceUrl) {
    var full = (0, tree_utils_1.addDefaults)(mvsTree, mvs_defaults_1.MVSDefaults);
    if (sourceUrl)
        (0, tree_utils_1.resolveUris)(full, sourceUrl, ['uri', 'url']);
    var converted = (0, tree_utils_1.convertTree)(full, mvsToMolstarConversionRules);
    if (converted.kind !== 'root')
        throw new Error("Root's type is not 'root' after conversion from MVS tree to Molstar tree.");
    var condensed = (0, tree_utils_1.condenseTree)(converted, molstarNodesToCondense);
    return condensed;
}
exports.convertMvsToMolstar = convertMvsToMolstar;
function fileExtensionMatches(filename, extensions) {
    filename = filename.toLowerCase();
    return extensions.some(function (ext) { return ext === '*' || filename.endsWith(ext); });
}
var StructureFormatExtensions = {
    mmcif: ['.cif', '.mmif'],
    bcif: ['.bcif'],
    pdb: ['.pdb', '.ent'],
};
/** Run some sanity check on a MVSTree. Return a list of potential problems (`undefined` if there are none) */
function mvsSanityCheckIssues(tree) {
    var result = [];
    (0, tree_utils_1.dfs)(tree, function (node, parent) {
        if (node.kind === 'parse' && (parent === null || parent === void 0 ? void 0 : parent.kind) === 'download') {
            var source = parent.params.url;
            var extensions = StructureFormatExtensions[node.params.format];
            if (!fileExtensionMatches(source, extensions)) {
                result.push("Parsing data from ".concat(source, " as ").concat(node.params.format, " format might be a mistake. The file extension doesn't match recommended file extensions (").concat(extensions.join(', '), ")"));
            }
        }
    });
    return result.length > 0 ? result : undefined;
}
exports.mvsSanityCheckIssues = mvsSanityCheckIssues;
/** Run some sanity check on a MVSTree and print potential issues to the console. */
function mvsSanityCheck(tree) {
    var issues = mvsSanityCheckIssues(tree);
    if (issues) {
        console.warn('There are potential issues in the MVS tree:');
        for (var _i = 0, issues_1 = issues; _i < issues_1.length; _i++) {
            var issue = issues_1[_i];
            console.warn(' ', issue);
        }
    }
}
exports.mvsSanityCheck = mvsSanityCheck;
