// molstar/lib/commonjs/mol-repr/structure/visual/nucleotide-ring-mesh.js
"use strict";
/**
 * Copyright (c) 2019-2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NucleotideRingVisual = exports.NucleotideRingParams = exports.DefaultNucleotideRingMeshProps = exports.NucleotideRingMeshParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var structure_1 = require("../../../mol-model/structure");
var mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
var mesh_builder_1 = require("../../../mol-geo/geometry/mesh/mesh-builder");
var int_1 = require("../../../mol-data/int");
var types_1 = require("../../../mol-model/structure/model/types");
var cylinder_1 = require("../../../mol-geo/geometry/mesh/builder/cylinder");
var sphere_1 = require("../../../mol-geo/geometry/mesh/builder/sphere");
var units_visual_1 = require("../units-visual");
var nucleotide_1 = require("./util/nucleotide");
var base_1 = require("../../../mol-geo/geometry/base");
var geometry_1 = require("../../../mol-math/geometry");
// TODO support rings for multiple locations (including from microheterogeneity)
var pTrace = (0, linear_algebra_1.Vec3)();
var pN1 = (0, linear_algebra_1.Vec3)();
var pC2 = (0, linear_algebra_1.Vec3)();
var pN3 = (0, linear_algebra_1.Vec3)();
var pC4 = (0, linear_algebra_1.Vec3)();
var pC5 = (0, linear_algebra_1.Vec3)();
var pC6 = (0, linear_algebra_1.Vec3)();
var pN7 = (0, linear_algebra_1.Vec3)();
var pC8 = (0, linear_algebra_1.Vec3)();
var pN9 = (0, linear_algebra_1.Vec3)();
var normal = (0, linear_algebra_1.Vec3)();
exports.NucleotideRingMeshParams = {
    sizeFactor: param_definition_1.ParamDefinition.Numeric(0.2, { min: 0, max: 10, step: 0.01 }),
    thicknessFactor: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 2, step: 0.01 }),
    radialSegments: param_definition_1.ParamDefinition.Numeric(16, { min: 2, max: 56, step: 2 }, base_1.BaseGeometry.CustomQualityParamInfo),
    detail: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 3, step: 1 }, base_1.BaseGeometry.CustomQualityParamInfo),
};
exports.DefaultNucleotideRingMeshProps = param_definition_1.ParamDefinition.getDefaultValues(exports.NucleotideRingMeshParams);
var positionsRing5_6 = new Float32Array(2 * 9 * 3);
var stripIndicesRing5_6 = new Uint32Array([0, 1, 2, 3, 4, 5, 6, 7, 16, 17, 14, 15, 12, 13, 8, 9, 10, 11, 0, 1]);
var fanIndicesTopRing5_6 = new Uint32Array([8, 12, 14, 16, 6, 4, 2, 0, 10]);
var fanIndicesBottomRing5_6 = new Uint32Array([9, 11, 1, 3, 5, 7, 17, 15, 13]);
var positionsRing6 = new Float32Array(2 * 6 * 3);
var stripIndicesRing6 = new Uint32Array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0, 1]);
var fanIndicesTopRing6 = new Uint32Array([0, 10, 8, 6, 4, 2]);
var fanIndicesBottomRing6 = new Uint32Array([1, 3, 5, 7, 9, 11]);
var tmpShiftV = (0, linear_algebra_1.Vec3)();
function shiftPositions(out, dir) {
    var positions = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        positions[_i - 2] = arguments[_i];
    }
    for (var i = 0, il = positions.length; i < il; ++i) {
        var v = positions[i];
        linear_algebra_1.Vec3.toArray(linear_algebra_1.Vec3.add(tmpShiftV, v, dir), out, (i * 2) * 3);
        linear_algebra_1.Vec3.toArray(linear_algebra_1.Vec3.sub(tmpShiftV, v, dir), out, (i * 2 + 1) * 3);
    }
}
function createNucleotideRingMesh(ctx, unit, structure, theme, props, mesh) {
    if (!structure_1.Unit.isAtomic(unit))
        return mesh_1.Mesh.createEmpty(mesh);
    var nucleotideElementCount = unit.nucleotideElements.length;
    if (!nucleotideElementCount)
        return mesh_1.Mesh.createEmpty(mesh);
    var sizeFactor = props.sizeFactor, thicknessFactor = props.thicknessFactor, radialSegments = props.radialSegments, detail = props.detail;
    var vertexCount = nucleotideElementCount * (26 + radialSegments * 2);
    var builderState = mesh_builder_1.MeshBuilder.createState(vertexCount, vertexCount / 4, mesh);
    var elements = unit.elements, model = unit.model;
    var _a = model.atomicHierarchy, chainAtomSegments = _a.chainAtomSegments, residueAtomSegments = _a.residueAtomSegments;
    var moleculeType = model.atomicHierarchy.derived.residue.moleculeType;
    var pos = unit.conformation.invariantPosition;
    var chainIt = int_1.Segmentation.transientSegments(chainAtomSegments, elements);
    var residueIt = int_1.Segmentation.transientSegments(residueAtomSegments, elements);
    var radius = 1 * sizeFactor;
    var thickness = thicknessFactor * sizeFactor;
    var cylinderProps = { radiusTop: radius, radiusBottom: radius, radialSegments: radialSegments };
    var i = 0;
    while (chainIt.hasNext) {
        residueIt.setSegment(chainIt.move());
        while (residueIt.hasNext) {
            var residueIndex = residueIt.move().index;
            if ((0, types_1.isNucleic)(moleculeType[residueIndex])) {
                var idx = (0, nucleotide_1.createNucleicIndices)();
                builderState.currentGroup = i;
                var _b = (0, nucleotide_1.getNucleotideBaseType)(unit, residueIndex), isPurine = _b.isPurine, isPyrimidine = _b.isPyrimidine;
                if (isPurine) {
                    (0, nucleotide_1.setPurinIndices)(idx, unit, residueIndex);
                    if (idx.N9 !== -1 && idx.trace !== -1) {
                        pos(idx.N9, pN9);
                        pos(idx.trace, pTrace);
                        builderState.currentGroup = i;
                        (0, cylinder_1.addCylinder)(builderState, pN9, pTrace, 1, cylinderProps);
                        (0, sphere_1.addSphere)(builderState, pN9, radius, detail);
                    }
                    if ((0, nucleotide_1.hasPurinIndices)(idx)) {
                        pos(idx.N1, pN1);
                        pos(idx.C2, pC2);
                        pos(idx.N3, pN3);
                        pos(idx.C4, pC4);
                        pos(idx.C5, pC5);
                        pos(idx.C6, pC6);
                        pos(idx.N7, pN7);
                        pos(idx.C8, pC8);
                        linear_algebra_1.Vec3.triangleNormal(normal, pN1, pC4, pC5);
                        linear_algebra_1.Vec3.scale(normal, normal, thickness);
                        shiftPositions(positionsRing5_6, normal, pN1, pC2, pN3, pC4, pC5, pC6, pN7, pC8, pN9);
                        mesh_builder_1.MeshBuilder.addTriangleStrip(builderState, positionsRing5_6, stripIndicesRing5_6);
                        mesh_builder_1.MeshBuilder.addTriangleFan(builderState, positionsRing5_6, fanIndicesTopRing5_6);
                        mesh_builder_1.MeshBuilder.addTriangleFan(builderState, positionsRing5_6, fanIndicesBottomRing5_6);
                    }
                }
                else if (isPyrimidine) {
                    (0, nucleotide_1.setPyrimidineIndices)(idx, unit, residueIndex);
                    if (idx.N1 !== -1 && idx.trace !== -1) {
                        pos(idx.N1, pN1);
                        pos(idx.trace, pTrace);
                        builderState.currentGroup = i;
                        (0, cylinder_1.addCylinder)(builderState, pN1, pTrace, 1, cylinderProps);
                        (0, sphere_1.addSphere)(builderState, pN1, radius, detail);
                    }
                    if ((0, nucleotide_1.hasPyrimidineIndices)(idx)) {
                        pos(idx.C2, pC2);
                        pos(idx.N3, pN3);
                        pos(idx.C4, pC4);
                        pos(idx.C5, pC5);
                        pos(idx.C6, pC6);
                        linear_algebra_1.Vec3.triangleNormal(normal, pN1, pC4, pC5);
                        linear_algebra_1.Vec3.scale(normal, normal, thickness);
                        shiftPositions(positionsRing6, normal, pN1, pC2, pN3, pC4, pC5, pC6);
                        mesh_builder_1.MeshBuilder.addTriangleStrip(builderState, positionsRing6, stripIndicesRing6);
                        mesh_builder_1.MeshBuilder.addTriangleFan(builderState, positionsRing6, fanIndicesTopRing6);
                        mesh_builder_1.MeshBuilder.addTriangleFan(builderState, positionsRing6, fanIndicesBottomRing6);
                    }
                }
                ++i;
            }
        }
    }
    var m = mesh_builder_1.MeshBuilder.getMesh(builderState);
    var sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, radius);
    m.setBoundingSphere(sphere);
    return m;
}
exports.NucleotideRingParams = tslib_1.__assign(tslib_1.__assign({}, units_visual_1.UnitsMeshParams), exports.NucleotideRingMeshParams);
function NucleotideRingVisual(materialId) {
    return (0, units_visual_1.UnitsMeshVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.NucleotideRingParams),
        createGeometry: createNucleotideRingMesh,
        createLocationIterator: nucleotide_1.NucleotideLocationIterator.fromGroup,
        getLoci: nucleotide_1.getNucleotideElementLoci,
        eachLocation: nucleotide_1.eachNucleotideElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = (newProps.sizeFactor !== currentProps.sizeFactor ||
                newProps.thicknessFactor !== currentProps.thicknessFactor ||
                newProps.radialSegments !== currentProps.radialSegments);
        }
    }, materialId);
}
exports.NucleotideRingVisual = NucleotideRingVisual;
