// molstar/lib/commonjs/mol-repr/structure/visual/nucleotide-atomic-bond.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Gianluca Tomasello <giagitom@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NucleotideAtomicBondMeshVisual = exports.NucleotideAtomicBondImpostorVisual = exports.NucleotideAtomicBondVisual = exports.NucleotideAtomicBondParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var structure_1 = require("../../../mol-model/structure");
var mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
var mesh_builder_1 = require("../../../mol-geo/geometry/mesh/mesh-builder");
var int_1 = require("../../../mol-data/int");
var types_1 = require("../../../mol-model/structure/model/types");
var cylinder_1 = require("../../../mol-geo/geometry/mesh/builder/cylinder");
var units_visual_1 = require("../units-visual");
var nucleotide_1 = require("./util/nucleotide");
var base_1 = require("../../../mol-geo/geometry/base");
var geometry_1 = require("../../../mol-math/geometry");
var cylinders_1 = require("../../../mol-geo/geometry/cylinders/cylinders");
var cylinders_builder_1 = require("../../../mol-geo/geometry/cylinders/cylinders-builder");
var pTrace = (0, linear_algebra_1.Vec3)();
var pN1 = (0, linear_algebra_1.Vec3)();
var pC2 = (0, linear_algebra_1.Vec3)();
var pN3 = (0, linear_algebra_1.Vec3)();
var pC4 = (0, linear_algebra_1.Vec3)();
var pC5 = (0, linear_algebra_1.Vec3)();
var pC6 = (0, linear_algebra_1.Vec3)();
var pN7 = (0, linear_algebra_1.Vec3)();
var pC8 = (0, linear_algebra_1.Vec3)();
var pN9 = (0, linear_algebra_1.Vec3)();
var pC1_1 = (0, linear_algebra_1.Vec3)();
var pC2_1 = (0, linear_algebra_1.Vec3)();
var pC3_1 = (0, linear_algebra_1.Vec3)();
var pC4_1 = (0, linear_algebra_1.Vec3)();
var pO4_1 = (0, linear_algebra_1.Vec3)();
exports.NucleotideAtomicBondParams = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, units_visual_1.UnitsMeshParams), units_visual_1.UnitsCylindersParams), { sizeFactor: param_definition_1.ParamDefinition.Numeric(0.3, { min: 0, max: 10, step: 0.01 }), radialSegments: param_definition_1.ParamDefinition.Numeric(16, { min: 2, max: 56, step: 2 }, base_1.BaseGeometry.CustomQualityParamInfo), tryUseImpostor: param_definition_1.ParamDefinition.Boolean(true) });
function NucleotideAtomicBondVisual(materialId, structure, props, webgl) {
    return props.tryUseImpostor && webgl && webgl.extensions.fragDepth
        ? NucleotideAtomicBondImpostorVisual(materialId)
        : NucleotideAtomicBondMeshVisual(materialId);
}
exports.NucleotideAtomicBondVisual = NucleotideAtomicBondVisual;
function createNucleotideAtomicBondImpostor(ctx, unit, structure, theme, props, cylinders) {
    if (!structure_1.Unit.isAtomic(unit))
        return cylinders_1.Cylinders.createEmpty(cylinders);
    var nucleotideElementCount = unit.nucleotideElements.length;
    if (!nucleotideElementCount)
        return cylinders_1.Cylinders.createEmpty(cylinders);
    var cylindersCountEstimate = nucleotideElementCount * 15; // 15 is the average purine (17) & pirimidine (13) bonds
    var builder = cylinders_builder_1.CylindersBuilder.create(cylindersCountEstimate, cylindersCountEstimate / 4, cylinders);
    var elements = unit.elements, model = unit.model;
    var _a = model.atomicHierarchy, chainAtomSegments = _a.chainAtomSegments, residueAtomSegments = _a.residueAtomSegments;
    var moleculeType = model.atomicHierarchy.derived.residue.moleculeType;
    var pos = unit.conformation.invariantPosition;
    var chainIt = int_1.Segmentation.transientSegments(chainAtomSegments, elements);
    var residueIt = int_1.Segmentation.transientSegments(residueAtomSegments, elements);
    var i = 0;
    var colorModeFlag = 2.0;
    while (chainIt.hasNext) {
        residueIt.setSegment(chainIt.move());
        while (residueIt.hasNext) {
            var residueIndex = residueIt.move().index;
            if ((0, types_1.isNucleic)(moleculeType[residueIndex])) {
                var idx = (0, nucleotide_1.createNucleicIndices)();
                (0, nucleotide_1.setSugarIndices)(idx, unit, residueIndex);
                if ((0, nucleotide_1.hasSugarIndices)(idx)) {
                    pos(idx.C1_1, pC1_1);
                    pos(idx.C2_1, pC2_1);
                    pos(idx.C3_1, pC3_1);
                    pos(idx.C4_1, pC4_1);
                    pos(idx.O4_1, pO4_1);
                    // trace cylinder
                    pos(idx.trace, pTrace);
                    builder.add(pC3_1[0], pC3_1[1], pC3_1[2], pTrace[0], pTrace[1], pTrace[2], 1, true, true, colorModeFlag, i);
                    // sugar ring
                    builder.add(pC3_1[0], pC3_1[1], pC3_1[2], pC4_1[0], pC4_1[1], pC4_1[2], 1, true, true, colorModeFlag, i);
                    builder.add(pC4_1[0], pC4_1[1], pC4_1[2], pO4_1[0], pO4_1[1], pO4_1[2], 1, true, true, colorModeFlag, i);
                    builder.add(pO4_1[0], pO4_1[1], pO4_1[2], pC1_1[0], pC1_1[1], pC1_1[2], 1, true, true, colorModeFlag, i);
                    builder.add(pC1_1[0], pC1_1[1], pC1_1[2], pC2_1[0], pC2_1[1], pC2_1[2], 1, true, true, colorModeFlag, i);
                    builder.add(pC2_1[0], pC2_1[1], pC2_1[2], pC3_1[0], pC3_1[1], pC3_1[2], 1, true, true, colorModeFlag, i);
                }
                var _b = (0, nucleotide_1.getNucleotideBaseType)(unit, residueIndex), isPurine = _b.isPurine, isPyrimidine = _b.isPyrimidine;
                if (isPurine) {
                    (0, nucleotide_1.setPurinIndices)(idx, unit, residueIndex);
                    if (idx.C1_1 !== -1 && idx.N9 !== -1) {
                        pos(idx.C1_1, pC1_1);
                        pos(idx.N9, pN9);
                        builder.add(pN9[0], pN9[1], pN9[2], pC1_1[0], pC1_1[1], pC1_1[2], 1, true, true, colorModeFlag, i);
                    }
                    else if (idx.N9 !== -1 && idx.trace !== -1) {
                        pos(idx.N9, pN9);
                        pos(idx.trace, pTrace);
                        builder.add(pN9[0], pN9[1], pN9[2], pTrace[0], pTrace[1], pTrace[2], 1, true, true, colorModeFlag, i);
                    }
                    if ((0, nucleotide_1.hasPurinIndices)(idx)) {
                        pos(idx.N1, pN1);
                        pos(idx.C2, pC2);
                        pos(idx.N3, pN3);
                        pos(idx.C4, pC4);
                        pos(idx.C5, pC5);
                        pos(idx.C6, pC6);
                        pos(idx.N7, pN7);
                        pos(idx.C8, pC8);
                        pos(idx.N9, pN9);
                        // base ring
                        builder.add(pN9[0], pN9[1], pN9[2], pC8[0], pC8[1], pC8[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC8[0], pC8[1], pC8[2], pN7[0], pN7[1], pN7[2], 1, true, true, colorModeFlag, i);
                        builder.add(pN7[0], pN7[1], pN7[2], pC5[0], pC5[1], pC5[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC5[0], pC5[1], pC5[2], pC6[0], pC6[1], pC6[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC6[0], pC6[1], pC6[2], pN1[0], pN1[1], pN1[2], 1, true, true, colorModeFlag, i);
                        builder.add(pN1[0], pN1[1], pN1[2], pC2[0], pC2[1], pC2[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC2[0], pC2[1], pC2[2], pN3[0], pN3[1], pN3[2], 1, true, true, colorModeFlag, i);
                        builder.add(pN3[0], pN3[1], pN3[2], pC4[0], pC4[1], pC4[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC4[0], pC4[1], pC4[2], pC5[0], pC5[1], pC5[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC4[0], pC4[1], pC4[2], pN9[0], pN9[1], pN9[2], 1, true, true, colorModeFlag, i);
                    }
                }
                else if (isPyrimidine) {
                    (0, nucleotide_1.setPyrimidineIndices)(idx, unit, residueIndex);
                    if (idx.C1_1 !== -1 && idx.N1 !== -1) {
                        pos(idx.N1, pN1);
                        pos(idx.C1_1, pC1_1);
                        builder.add(pN1[0], pN1[1], pN1[2], pC1_1[0], pC1_1[1], pC1_1[2], 1, true, true, colorModeFlag, i);
                    }
                    else if (idx.N1 !== -1 && idx.trace !== -1) {
                        pos(idx.N1, pN1);
                        pos(idx.trace, pTrace);
                        builder.add(pN1[0], pN1[1], pN1[2], pTrace[0], pTrace[1], pTrace[2], 1, true, true, colorModeFlag, i);
                    }
                    if ((0, nucleotide_1.hasPyrimidineIndices)(idx)) {
                        pos(idx.N1, pN1);
                        pos(idx.C2, pC2);
                        pos(idx.N3, pN3);
                        pos(idx.C4, pC4);
                        pos(idx.C5, pC5);
                        pos(idx.C6, pC6);
                        // base ring
                        builder.add(pN1[0], pN1[1], pN1[2], pC6[0], pC6[1], pC6[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC6[0], pC6[1], pC6[2], pC5[0], pC5[1], pC5[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC5[0], pC5[1], pC5[2], pC4[0], pC4[1], pC4[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC4[0], pC4[1], pC4[2], pN3[0], pN3[1], pN3[2], 1, true, true, colorModeFlag, i);
                        builder.add(pN3[0], pN3[1], pN3[2], pC2[0], pC2[1], pC2[2], 1, true, true, colorModeFlag, i);
                        builder.add(pC2[0], pC2[1], pC2[2], pN1[0], pN1[1], pN1[2], 1, true, true, colorModeFlag, i);
                    }
                }
                ++i;
            }
        }
    }
    var c = builder.getCylinders();
    var sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, 1 * props.sizeFactor);
    c.setBoundingSphere(sphere);
    return c;
}
function NucleotideAtomicBondImpostorVisual(materialId) {
    return (0, units_visual_1.UnitsCylindersVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.NucleotideAtomicBondParams),
        createGeometry: createNucleotideAtomicBondImpostor,
        createLocationIterator: nucleotide_1.NucleotideLocationIterator.fromGroup,
        getLoci: nucleotide_1.getNucleotideElementLoci,
        eachLocation: nucleotide_1.eachNucleotideElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = (newProps.sizeFactor !== currentProps.sizeFactor);
        },
        mustRecreate: function (structureGroup, props, webgl) {
            return !props.tryUseImpostor || !webgl;
        }
    }, materialId);
}
exports.NucleotideAtomicBondImpostorVisual = NucleotideAtomicBondImpostorVisual;
function createNucleotideAtomicBondMesh(ctx, unit, structure, theme, props, mesh) {
    if (!structure_1.Unit.isAtomic(unit))
        return mesh_1.Mesh.createEmpty(mesh);
    var nucleotideElementCount = unit.nucleotideElements.length;
    if (!nucleotideElementCount)
        return mesh_1.Mesh.createEmpty(mesh);
    var sizeFactor = props.sizeFactor, radialSegments = props.radialSegments;
    var vertexCount = nucleotideElementCount * (radialSegments * 15); // 15 is the average purine (17) & pirimidine (13) bonds
    var builderState = mesh_builder_1.MeshBuilder.createState(vertexCount, vertexCount / 4, mesh);
    var elements = unit.elements, model = unit.model;
    var _a = model.atomicHierarchy, chainAtomSegments = _a.chainAtomSegments, residueAtomSegments = _a.residueAtomSegments;
    var moleculeType = model.atomicHierarchy.derived.residue.moleculeType;
    var pos = unit.conformation.invariantPosition;
    var chainIt = int_1.Segmentation.transientSegments(chainAtomSegments, elements);
    var residueIt = int_1.Segmentation.transientSegments(residueAtomSegments, elements);
    var cylinderProps = { radiusTop: 1 * sizeFactor, radiusBottom: 1 * sizeFactor, radialSegments: radialSegments };
    var i = 0;
    while (chainIt.hasNext) {
        residueIt.setSegment(chainIt.move());
        while (residueIt.hasNext) {
            var residueIndex = residueIt.move().index;
            if ((0, types_1.isNucleic)(moleculeType[residueIndex])) {
                var idx = (0, nucleotide_1.createNucleicIndices)();
                builderState.currentGroup = i;
                (0, nucleotide_1.setSugarIndices)(idx, unit, residueIndex);
                if ((0, nucleotide_1.hasSugarIndices)(idx)) {
                    pos(idx.C1_1, pC1_1);
                    pos(idx.C2_1, pC2_1);
                    pos(idx.C3_1, pC3_1);
                    pos(idx.C4_1, pC4_1);
                    pos(idx.O4_1, pO4_1);
                    // trace cylinder
                    pos(idx.trace, pTrace);
                    (0, cylinder_1.addCylinder)(builderState, pC3_1, pTrace, 1, cylinderProps);
                    // sugar ring
                    (0, cylinder_1.addCylinder)(builderState, pC3_1, pC4_1, 1, cylinderProps);
                    (0, cylinder_1.addCylinder)(builderState, pC4_1, pO4_1, 1, cylinderProps);
                    (0, cylinder_1.addCylinder)(builderState, pO4_1, pC1_1, 1, cylinderProps);
                    (0, cylinder_1.addCylinder)(builderState, pC1_1, pC2_1, 1, cylinderProps);
                    (0, cylinder_1.addCylinder)(builderState, pC2_1, pC3_1, 1, cylinderProps);
                }
                var _b = (0, nucleotide_1.getNucleotideBaseType)(unit, residueIndex), isPurine = _b.isPurine, isPyrimidine = _b.isPyrimidine;
                if (isPurine) {
                    (0, nucleotide_1.setPurinIndices)(idx, unit, residueIndex);
                    if (idx.C1_1 !== -1 && idx.N9 !== -1) {
                        pos(idx.C1_1, pC1_1);
                        pos(idx.N9, pN9);
                        (0, cylinder_1.addCylinder)(builderState, pN9, pC1_1, 1, cylinderProps);
                    }
                    else if (idx.N9 !== -1 && idx.trace !== -1) {
                        pos(idx.N9, pN9);
                        pos(idx.trace, pTrace);
                        (0, cylinder_1.addCylinder)(builderState, pN9, pTrace, 1, cylinderProps);
                    }
                    if ((0, nucleotide_1.hasPurinIndices)(idx)) {
                        pos(idx.N1, pN1);
                        pos(idx.C2, pC2);
                        pos(idx.N3, pN3);
                        pos(idx.C4, pC4);
                        pos(idx.C5, pC5);
                        pos(idx.C6, pC6);
                        pos(idx.N7, pN7);
                        pos(idx.C8, pC8);
                        pos(idx.N9, pN9);
                        // base ring
                        (0, cylinder_1.addCylinder)(builderState, pN9, pC8, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC8, pN7, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pN7, pC5, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC5, pC6, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC6, pN1, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pN1, pC2, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC2, pN3, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pN3, pC4, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC4, pC5, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC4, pN9, 1, cylinderProps);
                    }
                }
                else if (isPyrimidine) {
                    (0, nucleotide_1.setPyrimidineIndices)(idx, unit, residueIndex);
                    if (idx.C1_1 !== -1 && idx.N1 !== -1) {
                        pos(idx.N1, pN1);
                        pos(idx.C1_1, pC1_1);
                        (0, cylinder_1.addCylinder)(builderState, pN1, pC1_1, 1, cylinderProps);
                    }
                    else if (idx.N1 !== -1 && idx.trace !== -1) {
                        pos(idx.N1, pN1);
                        pos(idx.trace, pTrace);
                        (0, cylinder_1.addCylinder)(builderState, pN1, pTrace, 1, cylinderProps);
                    }
                    if ((0, nucleotide_1.hasPyrimidineIndices)(idx)) {
                        pos(idx.N1, pN1);
                        pos(idx.C2, pC2);
                        pos(idx.N3, pN3);
                        pos(idx.C4, pC4);
                        pos(idx.C5, pC5);
                        pos(idx.C6, pC6);
                        // base ring
                        (0, cylinder_1.addCylinder)(builderState, pN1, pC6, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC6, pC5, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC5, pC4, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC4, pN3, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pN3, pC2, 1, cylinderProps);
                        (0, cylinder_1.addCylinder)(builderState, pC2, pN1, 1, cylinderProps);
                    }
                }
                ++i;
            }
        }
    }
    var m = mesh_builder_1.MeshBuilder.getMesh(builderState);
    var sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, 1 * props.sizeFactor);
    m.setBoundingSphere(sphere);
    return m;
}
function NucleotideAtomicBondMeshVisual(materialId) {
    return (0, units_visual_1.UnitsMeshVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.NucleotideAtomicBondParams),
        createGeometry: createNucleotideAtomicBondMesh,
        createLocationIterator: nucleotide_1.NucleotideLocationIterator.fromGroup,
        getLoci: nucleotide_1.getNucleotideElementLoci,
        eachLocation: nucleotide_1.eachNucleotideElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = (newProps.sizeFactor !== currentProps.sizeFactor ||
                newProps.radialSegments !== currentProps.radialSegments);
        },
        mustRecreate: function (structureGroup, props, webgl) {
            return props.tryUseImpostor && !!webgl;
        }
    }, materialId);
}
exports.NucleotideAtomicBondMeshVisual = NucleotideAtomicBondMeshVisual;
