// molstar/lib/mol-util/param-choice.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { ParamDefinition } from './param-definition';
/**
 * Represents a set of values to choose from, with a default value. Example:
 * ```
 * export const MyChoice = new Choice({ yes: 'I agree', no: 'Nope' }, 'yes');
 * export type MyChoiceType = Choice.Values<typeof MyChoice>; // 'yes'|'no'
 * ```
 */
var Choice = /** @class */ (function () {
    function Choice(opts, defaultValue) {
        this.defaultValue = defaultValue;
        this.options = Object.keys(opts).map(function (k) { return [k, opts[k]]; });
        this.nameDict = opts;
    }
    Choice.prototype.PDSelect = function (defaultValue, info) {
        return ParamDefinition.Select(defaultValue !== null && defaultValue !== void 0 ? defaultValue : this.defaultValue, this.options, info);
    };
    Choice.prototype.prettyName = function (value) {
        return this.nameDict[value];
    };
    Object.defineProperty(Choice.prototype, "values", {
        get: function () {
            return this.options.map(function (_a) {
                var value = _a[0], pretty = _a[1];
                return value;
            });
        },
        enumerable: false,
        configurable: true
    });
    return Choice;
}());
export { Choice };
