// molstar/lib/commonjs/extensions/mvs/components/annotation-label/visual.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MVSAnnotationLabelTextVisual = exports.MVSAnnotationLabelTextParams = void 0;
var tslib_1 = require("tslib");
var text_builder_1 = require("../../../../mol-geo/geometry/text/text-builder");
var complex_visual_1 = require("../../../../mol-repr/structure/complex-visual");
var Original = tslib_1.__importStar(require("../../../../mol-repr/structure/visual/label-text"));
var element_1 = require("../../../../mol-repr/structure/visual/util/element");
var names_1 = require("../../../../mol-util/color/names");
var object_1 = require("../../../../mol-util/object");
var param_definition_1 = require("../../../../mol-util/param-definition");
var label_text_1 = require("../../helpers/label-text");
var selections_1 = require("../../helpers/selections");
var annotation_prop_1 = require("../annotation-prop");
exports.MVSAnnotationLabelTextParams = tslib_1.__assign(tslib_1.__assign({ annotationId: param_definition_1.ParamDefinition.Text('', { description: 'Reference to "Annotation" custom model property', isEssential: true }), fieldName: param_definition_1.ParamDefinition.Text('label', { description: 'Annotation field (column) from which to take label contents', isEssential: true }) }, (0, object_1.omitObjectKeys)(Original.LabelTextParams, ['level', 'chainScale', 'residueScale', 'elementScale'])), { borderColor: tslib_1.__assign(tslib_1.__assign({}, Original.LabelTextParams.borderColor), { defaultValue: names_1.ColorNames.black }) });
/** Create "label-text" visual for "MVS Annotation Label" representation */
function MVSAnnotationLabelTextVisual(materialId) {
    return (0, complex_visual_1.ComplexTextVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.MVSAnnotationLabelTextParams),
        createGeometry: createLabelText,
        createLocationIterator: element_1.ElementIterator.fromStructure,
        getLoci: element_1.getSerialElementLoci,
        eachLocation: element_1.eachSerialElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = newProps.annotationId !== currentProps.annotationId || newProps.fieldName !== currentProps.fieldName;
        }
    }, materialId);
}
exports.MVSAnnotationLabelTextVisual = MVSAnnotationLabelTextVisual;
function createLabelText(ctx, structure, theme, props, text) {
    var _a;
    var _b = (0, annotation_prop_1.getMVSAnnotationForStructure)(structure, props.annotationId), annotation = _b.annotation, model = _b.model;
    var rows = (_a = annotation === null || annotation === void 0 ? void 0 : annotation.getRows()) !== null && _a !== void 0 ? _a : [];
    var _c = (0, selections_1.groupRows)(rows), count = _c.count, offsets = _c.offsets, grouped = _c.grouped;
    var builder = text_builder_1.TextBuilder.create(props, count, count / 2, text);
    for (var iGroup = 0; iGroup < count; iGroup++) {
        var iFirstRowInGroup = grouped[offsets[iGroup]];
        var labelText = annotation.getValueForRow(iFirstRowInGroup, props.fieldName);
        if (!labelText)
            continue;
        var rowsInGroup = grouped.slice(offsets[iGroup], offsets[iGroup + 1]).map(function (j) { return rows[j]; });
        var p = (0, label_text_1.textPropsForSelection)(structure, theme.size.size, rowsInGroup, model);
        if (!p)
            continue;
        builder.add(labelText, p.center[0], p.center[1], p.center[2], p.depth, p.scale, p.group);
    }
    return builder.getText();
}
