// molstar/lib/commonjs/extensions/mvs/tree/generic/params-schema.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.paramsValidationIssues = exports.AllRequired = exports.fieldValidationIssues = exports.OptionalField = exports.RequiredField = exports.literal = exports.nullable = exports.union = exports.list = exports.tuple = exports.bool = exports.float = exports.int = exports.str = void 0;
var tslib_1 = require("tslib");
var iots = tslib_1.__importStar(require("io-ts"));
var PathReporter_1 = require("io-ts/PathReporter");
var object_1 = require("../../../../mol-util/object");
var json_1 = require("../../../../mol-util/json");
/** Type definition for a string  */
exports.str = iots.string;
/** Type definition for an integer  */
exports.int = iots.Integer;
/** Type definition for a float or integer number  */
exports.float = iots.number;
/** Type definition for a boolean  */
exports.bool = iots.boolean;
/** Type definition for a tuple, e.g. `tuple([str, int, int])`  */
exports.tuple = iots.tuple;
/** Type definition for a list/array, e.g. `list(str)`  */
exports.list = iots.array;
/** Type definition for union types, e.g. `union([str, int])` means string or integer  */
exports.union = iots.union;
/** Type definition for nullable types, e.g. `nullable(str)` means string or `null`  */
function nullable(type) {
    return (0, exports.union)([type, iots.null]);
}
exports.nullable = nullable;
/** Type definition for literal types, e.g. `literal('red', 'green', 'blue')` means 'red' or 'green' or 'blue'  */
function literal() {
    var values = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        values[_i] = arguments[_i];
    }
    if (values.length === 0) {
        throw new Error("literal type must have at least one value");
    }
    var typeName = "(".concat(values.map(function (v) { return (0, json_1.onelinerJsonString)(v); }).join(' | '), ")");
    return new iots.Type(typeName, (function (value) { return values.includes(value); }), function (value, ctx) { return values.includes(value) ? { _tag: 'Right', right: value } : { _tag: 'Left', left: [{ value: value, context: ctx, message: "\"".concat(value, "\" is not a valid value for literal type ").concat(typeName) }] }; }, function (value) { return value; });
}
exports.literal = literal;
function RequiredField(type, description) {
    return { type: type, required: true, description: description };
}
exports.RequiredField = RequiredField;
function OptionalField(type, description) {
    return { type: type, required: false, description: description };
}
exports.OptionalField = OptionalField;
/** Return `undefined` if `value` has correct type for `field`, regardsless of if required or optional.
 * Return description of validation issues, if `value` has wrong type. */
function fieldValidationIssues(field, value) {
    var validation = field.type.decode(value);
    if (validation._tag === 'Right') {
        return undefined;
    }
    else {
        return PathReporter_1.PathReporter.report(validation);
    }
}
exports.fieldValidationIssues = fieldValidationIssues;
function AllRequired(paramsSchema) {
    return (0, object_1.mapObjectMap)(paramsSchema, function (field) { return RequiredField(field.type, field.description); });
}
exports.AllRequired = AllRequired;
/** Return `undefined` if `values` contains correct value types for `schema`,
 * return description of validation issues, if `values` have wrong type.
 * If `options.requireAll`, all parameters (including optional) must have a value provided.
 * If `options.noExtra` is true, presence of any extra parameters is treated as an issue.
 */
function paramsValidationIssues(schema, values, options) {
    if (options === void 0) { options = {}; }
    if (!(0, object_1.isPlainObject)(values))
        return ["Parameters must be an object, not ".concat(values)];
    for (var key in schema) {
        var paramDef = schema[key];
        if (Object.hasOwn(values, key)) {
            var value = values[key];
            var issues = fieldValidationIssues(paramDef, value);
            if (issues)
                return tslib_1.__spreadArray(["Invalid type for parameter \"".concat(key, "\":")], issues.map(function (s) { return '  ' + s; }), true);
        }
        else {
            if (paramDef.required)
                return ["Missing required parameter \"".concat(key, "\".")];
            if (options.requireAll)
                return ["Missing optional parameter \"".concat(key, "\".")];
        }
    }
    if (options.noExtra) {
        for (var key in values) {
            if (!Object.hasOwn(schema, key))
                return ["Unknown parameter \"".concat(key, "\".")];
        }
    }
    return undefined;
}
exports.paramsValidationIssues = paramsValidationIssues;
