// molstar/lib/commonjs/extensions/mvs/tree/generic/tree-schema.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.treeSchemaToMarkdown = exports.treeSchemaToString = exports.validateTree = exports.treeValidationIssues = exports.TreeSchemaWithAllRequired = exports.TreeSchema = exports.getChildren = exports.getParams = void 0;
var tslib_1 = require("tslib");
var json_1 = require("../../../../mol-util/json");
var object_1 = require("../../../../mol-util/object");
var params_schema_1 = require("./params-schema");
var tree_utils_1 = require("./tree-utils");
/** Get params from a tree node */
function getParams(node) {
    var _a;
    return (_a = node.params) !== null && _a !== void 0 ? _a : {};
}
exports.getParams = getParams;
/** Get children from a tree node */
function getChildren(tree) {
    var _a;
    return (_a = tree.children) !== null && _a !== void 0 ? _a : [];
}
exports.getChildren = getChildren;
function TreeSchema(schema) {
    return schema;
}
exports.TreeSchema = TreeSchema;
function TreeSchemaWithAllRequired(schema) {
    return tslib_1.__assign(tslib_1.__assign({}, schema), { nodes: (0, object_1.mapObjectMap)(schema.nodes, function (node) { return (tslib_1.__assign(tslib_1.__assign({}, node), { params: (0, params_schema_1.AllRequired)(node.params) })); }) });
}
exports.TreeSchemaWithAllRequired = TreeSchemaWithAllRequired;
/** Return `undefined` if a tree conforms to the given schema,
 * return validation issues (as a list of lines) if it does not conform.
 * If `options.requireAll`, all parameters (including optional) must have a value provided.
 * If `options.noExtra` is true, presence of any extra parameters is treated as an issue.
 * If `options.anyRoot` is true, the kind of the root node is not enforced.
 */
function treeValidationIssues(schema, tree, options) {
    if (options === void 0) { options = {}; }
    if (!(0, object_1.isPlainObject)(tree))
        return ["Node must be an object, not ".concat(tree)];
    if (!options.anyRoot && tree.kind !== schema.rootKind)
        return ["Invalid root node kind \"".concat(tree.kind, "\", root must be of kind \"").concat(schema.rootKind, "\"")];
    var nodeSchema = schema.nodes[tree.kind];
    if (!nodeSchema)
        return ["Unknown node kind \"".concat(tree.kind, "\"")];
    if (nodeSchema.parent && (options.parent !== undefined) && !nodeSchema.parent.includes(options.parent)) {
        return ["Node of kind \"".concat(tree.kind, "\" cannot appear as a child of \"").concat(options.parent, "\". Allowed parents for \"").concat(tree.kind, "\" are: ").concat(nodeSchema.parent.map(function (s) { return "\"".concat(s, "\""); }).join(', '))];
    }
    var issues = (0, params_schema_1.paramsValidationIssues)(nodeSchema.params, getParams(tree), options);
    if (issues)
        return tslib_1.__spreadArray(["Invalid parameters for node of kind \"".concat(tree.kind, "\":")], issues.map(function (s) { return '  ' + s; }), true);
    for (var _i = 0, _a = getChildren(tree); _i < _a.length; _i++) {
        var child = _a[_i];
        var issues_1 = treeValidationIssues(schema, child, tslib_1.__assign(tslib_1.__assign({}, options), { anyRoot: true, parent: tree.kind }));
        if (issues_1)
            return issues_1;
    }
    return undefined;
}
exports.treeValidationIssues = treeValidationIssues;
/** Validate a tree against the given schema.
 * Do nothing if OK; print validation issues on console and throw an error is the tree does not conform.
 * Include `label` in the printed output. */
function validateTree(schema, tree, label) {
    var issues = treeValidationIssues(schema, tree, { noExtra: true });
    if (issues) {
        console.warn("Invalid ".concat(label, " tree:\n").concat((0, tree_utils_1.treeToString)(tree)));
        console.error("".concat(label, " tree validation issues:"));
        for (var _i = 0, issues_2 = issues; _i < issues_2.length; _i++) {
            var line = issues_2[_i];
            console.error(' ', line);
        }
        throw new Error('FormatError');
    }
}
exports.validateTree = validateTree;
/** Return documentation for a tree schema as plain text */
function treeSchemaToString(schema, defaults) {
    return treeSchemaToString_(schema, defaults, false);
}
exports.treeSchemaToString = treeSchemaToString;
/** Return documentation for a tree schema as markdown text */
function treeSchemaToMarkdown(schema, defaults) {
    return treeSchemaToString_(schema, defaults, true);
}
exports.treeSchemaToMarkdown = treeSchemaToMarkdown;
function treeSchemaToString_(schema, defaults, markdown) {
    var _a;
    if (markdown === void 0) { markdown = false; }
    var out = [];
    var bold = function (str) { return markdown ? "**".concat(str, "**") : str; };
    var code = function (str) { return markdown ? "`".concat(str, "`") : str; };
    var h1 = markdown ? '## ' : '  - ';
    var p1 = markdown ? '' : '    ';
    var h2 = markdown ? '- ' : '      - ';
    var p2 = markdown ? '  ' : '        ';
    var newline = markdown ? '\n\n' : '\n';
    out.push("Tree schema:");
    for (var kind in schema.nodes) {
        var _b = schema.nodes[kind], description = _b.description, params = _b.params, parent_1 = _b.parent;
        out.push("".concat(h1).concat(code(kind)));
        if (kind === schema.rootKind) {
            out.push("".concat(p1, "[Root of the tree must be of this kind]"));
        }
        if (description) {
            out.push("".concat(p1).concat(description));
        }
        out.push("".concat(p1, "Parent: ").concat(!parent_1 ? 'any' : parent_1.length === 0 ? 'none' : parent_1.map(code).join(' or ')));
        out.push("".concat(p1, "Params:").concat(Object.keys(params).length > 0 ? '' : ' none'));
        for (var key in params) {
            var field = params[key];
            var typeString = field.type.name;
            if (typeString.startsWith('(') && typeString.endsWith(')')) {
                typeString = typeString.slice(1, -1);
            }
            out.push("".concat(h2).concat(bold(code(key + (field.required ? ': ' : '?: ')))).concat(code(typeString)));
            var defaultValue = (_a = defaults === null || defaults === void 0 ? void 0 : defaults[kind]) === null || _a === void 0 ? void 0 : _a[key];
            if (field.description) {
                out.push("".concat(p2).concat(field.description));
            }
            if (defaultValue !== undefined) {
                out.push("".concat(p2, "Default: ").concat(code((0, json_1.onelinerJsonString)(defaultValue))));
            }
        }
    }
    return out.join(newline);
}
