// molstar/lib/commonjs/cli/mvs/mvs-render.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 *
 * Command-line application for rendering images from MolViewSpec files
 * Build: npm install --no-save canvas gl jpeg-js pngjs  // these packages are not listed in Mol* dependencies for performance reasons
 *        npm run build
 * Run:   node lib/commonjs/cli/mvs/mvs-render -i examples/mvs/1cbs.mvsj -o ../outputs/1cbs.png --size 800x600 --molj
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var argparse_1 = require("argparse");
var fs_1 = tslib_1.__importDefault(require("fs"));
var gl_1 = tslib_1.__importDefault(require("gl"));
var jpeg_js_1 = tslib_1.__importDefault(require("jpeg-js"));
var path_1 = tslib_1.__importDefault(require("path"));
var pngjs_1 = tslib_1.__importDefault(require("pngjs"));
var canvas3d_1 = require("../../mol-canvas3d/canvas3d");
var headless_plugin_context_1 = require("../../mol-plugin/headless-plugin-context");
var spec_1 = require("../../mol-plugin/spec");
var headless_screenshot_1 = require("../../mol-plugin/util/headless-screenshot");
var data_source_1 = require("../../mol-util/data-source");
var json_1 = require("../../mol-util/json");
var param_definition_1 = require("../../mol-util/param-definition");
// MolViewSpec must be imported after HeadlessPluginContext
var behavior_1 = require("../../extensions/mvs/behavior");
var load_1 = require("../../extensions/mvs/load");
var mvs_data_1 = require("../../extensions/mvs/mvs-data");
var font_atlas_1 = require("../../mol-geo/geometry/text/font-atlas");
(0, data_source_1.setFSModule)(fs_1.default);
(0, font_atlas_1.setCanvasModule)(require('canvas'));
var DEFAULT_SIZE = '800x800';
/** Return parsed command line arguments for `main` */
function parseArguments() {
    var parser = new argparse_1.ArgumentParser({ description: 'Command-line application for rendering images from MolViewSpec files' });
    parser.add_argument('-i', '--input', { required: true, nargs: '+', help: 'Input file(s) in .mvsj format' });
    parser.add_argument('-o', '--output', { required: true, nargs: '+', help: 'File path(s) for output files (one output path for each input file). Output format is inferred from the file extension (.png or .jpg)' });
    parser.add_argument('-s', '--size', { help: "Output image resolution, {width}x{height}. Default: ".concat(DEFAULT_SIZE, "."), default: DEFAULT_SIZE });
    parser.add_argument('-m', '--molj', { action: 'store_true', help: "Save Mol* state (.molj) in addition to rendered images (use the same output file paths but with .molj extension)" });
    var args = parser.parse_args();
    try {
        var parts = args.size.split('x');
        if (parts.length !== 2)
            throw new Error('Must contain two x-separated parts');
        args.size = { width: parseIntStrict(parts[0]), height: parseIntStrict(parts[1]) };
    }
    catch (_a) {
        parser.error("argument: --size: invalid image size string: '".concat(args.size, "' (must be two x-separated integers (width and height), e.g. '400x300')"));
    }
    if (args.input.length !== args.output.length) {
        parser.error("argument: --output: must specify the same number of input and output file paths (specified ".concat(args.input.length, " input path").concat(args.input.length !== 1 ? 's' : '', " but ").concat(args.output.length, " output path").concat(args.output.length !== 1 ? 's' : '', ")"));
    }
    return tslib_1.__assign({}, args);
}
/** Main workflow for rendering images from MolViewSpec files */
function main(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var plugin, i, input, output, data, mvsData;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, createHeadlessPlugin(args)];
                case 1:
                    plugin = _a.sent();
                    i = 0;
                    _a.label = 2;
                case 2:
                    if (!(i < args.input.length)) return [3 /*break*/, 8];
                    input = args.input[i];
                    output = args.output[i];
                    console.log("Processing ".concat(input, " -> ").concat(output));
                    data = fs_1.default.readFileSync(input, { encoding: 'utf8' });
                    mvsData = mvs_data_1.MVSData.fromMVSJ(data);
                    return [4 /*yield*/, (0, load_1.loadMVS)(plugin, mvsData, { sanityChecks: true, replaceExisting: true, sourceUrl: "file://".concat(path_1.default.resolve(input)) })];
                case 3:
                    _a.sent();
                    fs_1.default.mkdirSync(path_1.default.dirname(output), { recursive: true });
                    if (!args.molj) return [3 /*break*/, 5];
                    return [4 /*yield*/, plugin.saveStateSnapshot(withExtension(output, '.molj'))];
                case 4:
                    _a.sent();
                    _a.label = 5;
                case 5: return [4 /*yield*/, plugin.saveImage(output)];
                case 6:
                    _a.sent();
                    checkState(plugin);
                    _a.label = 7;
                case 7:
                    i++;
                    return [3 /*break*/, 2];
                case 8: return [4 /*yield*/, plugin.clear()];
                case 9:
                    _a.sent();
                    plugin.dispose();
                    return [2 /*return*/];
            }
        });
    });
}
/** Return a new and initiatized HeadlessPlugin */
function createHeadlessPlugin(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var externalModules, spec, headlessCanvasOptions, canvasOptions, plugin, error_1;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    externalModules = { gl: gl_1.default, pngjs: pngjs_1.default, 'jpeg-js': jpeg_js_1.default };
                    spec = (0, spec_1.DefaultPluginSpec)();
                    spec.behaviors.push(spec_1.PluginSpec.Behavior(behavior_1.MolViewSpec));
                    headlessCanvasOptions = (0, headless_screenshot_1.defaultCanvas3DParams)();
                    canvasOptions = tslib_1.__assign(tslib_1.__assign({}, param_definition_1.ParamDefinition.getDefaultValues(canvas3d_1.Canvas3DParams)), { cameraResetDurationMs: headlessCanvasOptions.cameraResetDurationMs, postprocessing: headlessCanvasOptions.postprocessing });
                    plugin = new headless_plugin_context_1.HeadlessPluginContext(externalModules, spec, args.size, { canvas: canvasOptions });
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, plugin.init()];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 4];
                case 3:
                    error_1 = _a.sent();
                    plugin.dispose();
                    throw error_1;
                case 4: return [2 /*return*/, plugin];
            }
        });
    });
}
/** Parse integer, fail early. */
function parseIntStrict(str) {
    if (str === '')
        throw new Error('Is empty string');
    var result = Number(str);
    if (isNaN(result))
        throw new Error('Is NaN');
    if (Math.floor(result) !== result)
        throw new Error('Is not integer');
    return result;
}
/** Replace the file extension in `filename` by `extension`. If `filename` has no extension, add it. */
function withExtension(filename, extension) {
    var oldExtension = path_1.default.extname(filename);
    return filename.slice(0, -oldExtension.length) + extension;
}
/** Check Mol* state, print and throw error if any cell is not OK. */
function checkState(plugin) {
    var cells = Array.from(plugin.state.data.cells.values());
    var badCell = cells.find(function (cell) { return cell.status !== 'ok'; });
    if (badCell) {
        console.error("Building Mol* state failed");
        console.error("    Transformer: ".concat(badCell.transform.transformer.id));
        console.error("    Params: ".concat((0, json_1.onelinerJsonString)(badCell.transform.params)));
        console.error("    Error: ".concat(badCell.errorText));
        console.error("");
        throw new Error("Building Mol* state failed: ".concat(badCell.errorText));
    }
}
main(parseArguments());
