// molstar/lib/commonjs/mol-util/object.js
"use strict";
/**
 * Copyright (c) 2017-2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.promiseAllObj = exports.isPlainObject = exports.objectFromKeysAndValues = exports.omitObjectKeys = exports.pickObjectKeys = exports.objectForEach = exports.mapArrayToObject = exports.mapObjectMap = exports.deepClone = exports.shallowMergeArray = exports.shallowMerge = exports.shallowEqual = exports.shallowMerge2 = exports.assignIfUndefined = void 0;
var tslib_1 = require("tslib");
var hasOwnProperty = Object.prototype.hasOwnProperty;
/** Assign to the object if a given property in update is undefined */
function assignIfUndefined(to, full) {
    for (var _i = 0, _a = Object.keys(full); _i < _a.length; _i++) {
        var k = _a[_i];
        if (!hasOwnProperty.call(full, k))
            continue;
        if (typeof to[k] === 'undefined') {
            to[k] = full[k];
        }
    }
    return to;
}
exports.assignIfUndefined = assignIfUndefined;
/** Create new object if any property in "update" changes in "source". */
function shallowMerge2(source, update) {
    // Adapted from LiteMol (https://github.com/dsehnal/LiteMol)
    var changed = false;
    for (var _i = 0, _a = Object.keys(update); _i < _a.length; _i++) {
        var k = _a[_i];
        if (!hasOwnProperty.call(update, k))
            continue;
        if (update[k] !== source[k]) {
            changed = true;
            break;
        }
    }
    if (!changed)
        return source;
    return Object.assign({}, source, update);
}
exports.shallowMerge2 = shallowMerge2;
function shallowEqual(a, b) {
    if (!a) {
        if (!b)
            return true;
        return false;
    }
    if (!b)
        return false;
    var keys = Object.keys(a);
    if (Object.keys(b).length !== keys.length)
        return false;
    for (var _i = 0, keys_1 = keys; _i < keys_1.length; _i++) {
        var k = keys_1[_i];
        if (!hasOwnProperty.call(a, k) || a[k] !== b[k])
            return false;
    }
    return true;
}
exports.shallowEqual = shallowEqual;
function shallowMerge(source) {
    var rest = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        rest[_i - 1] = arguments[_i];
    }
    return shallowMergeArray(source, rest);
}
exports.shallowMerge = shallowMerge;
function shallowMergeArray(source, rest) {
    // Adapted from LiteMol (https://github.com/dsehnal/LiteMol)
    var ret = source;
    for (var s = 0; s < rest.length; s++) {
        if (!rest[s])
            continue;
        ret = shallowMerge2(source, rest[s]);
        if (ret !== source) {
            for (var i = s + 1; i < rest.length; i++) {
                ret = Object.assign(ret, rest[i]);
            }
            break;
        }
    }
    return ret;
}
exports.shallowMergeArray = shallowMergeArray;
/** Simple deep clone for number, boolean, string, null, undefined, object, array */
function deepClone(source) {
    if (null === source || 'object' !== typeof source)
        return source;
    if (source instanceof Array) {
        var copy = [];
        for (var i = 0, len = source.length; i < len; i++) {
            copy[i] = deepClone(source[i]);
        }
        return copy;
    }
    // `instanceof Object` does not find `Object.create(null)`
    if (typeof source === 'object' && !('prototype' in source)) {
        var copy = {};
        for (var k in source) {
            if (hasOwnProperty.call(source, k))
                copy[k] = deepClone(source[k]);
        }
        return copy;
    }
    throw new Error("Can't clone, type \"".concat(typeof source, "\" unsupported"));
}
exports.deepClone = deepClone;
/** Return a new object with the same keys, where function `f` is applied to each value.
 * Equivalent to Pythonic `{k: f(v) for k, v in obj.items()}` */
function mapObjectMap(obj, f) {
    var ret = {};
    for (var _i = 0, _a = Object.keys(obj); _i < _a.length; _i++) {
        var k = _a[_i];
        ret[k] = f(obj[k]);
    }
    return ret;
}
exports.mapObjectMap = mapObjectMap;
/** Return an object with keys being the elements of `array` and values computed by `getValue` function.
 * Equivalent to Pythonic `{k: getValue(k) for k in array}` */
function mapArrayToObject(array, getValue) {
    var result = {};
    for (var _i = 0, array_1 = array; _i < array_1.length; _i++) {
        var key = array_1[_i];
        result[key] = getValue(key);
    }
    return result;
}
exports.mapArrayToObject = mapArrayToObject;
function objectForEach(o, f) {
    if (!o)
        return;
    for (var _i = 0, _a = Object.keys(o); _i < _a.length; _i++) {
        var k = _a[_i];
        f(o[k], k);
    }
}
exports.objectForEach = objectForEach;
/** Return an object with keys `keys` and their values same as in `obj` */
function pickObjectKeys(obj, keys) {
    var result = {};
    for (var _i = 0, keys_2 = keys; _i < keys_2.length; _i++) {
        var key = keys_2[_i];
        if (Object.hasOwn(obj, key)) {
            result[key] = obj[key];
        }
    }
    return result;
}
exports.pickObjectKeys = pickObjectKeys;
/** Return an object same as `obj` but without keys `keys` */
function omitObjectKeys(obj, omitKeys) {
    var result = tslib_1.__assign({}, obj);
    for (var _i = 0, omitKeys_1 = omitKeys; _i < omitKeys_1.length; _i++) {
        var key = omitKeys_1[_i];
        delete result[key];
    }
    return result;
}
exports.omitObjectKeys = omitObjectKeys;
/** Create an object from keys and values (first key maps to first value etc.) */
function objectFromKeysAndValues(keys, values) {
    var obj = {};
    for (var i = 0; i < keys.length; i++) {
        obj[keys[i]] = values[i];
    }
    return obj;
}
exports.objectFromKeysAndValues = objectFromKeysAndValues;
/** Decide if `obj` is a good old object (not array or null or other type). */
function isPlainObject(obj) {
    return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
}
exports.isPlainObject = isPlainObject;
/** Like `Promise.all` but with objects instead of arrays */
function promiseAllObj(promisesObj) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var keys, promises, results;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    keys = Object.keys(promisesObj);
                    promises = Object.values(promisesObj);
                    return [4 /*yield*/, Promise.all(promises)];
                case 1:
                    results = _a.sent();
                    return [2 /*return*/, objectFromKeysAndValues(keys, results)];
            }
        });
    });
}
exports.promiseAllObj = promiseAllObj;
