// molstar/lib/extensions/mvs/tree/generic/params-schema.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __spreadArray } from "tslib";
import * as iots from 'io-ts';
import { PathReporter } from 'io-ts/PathReporter';
import { isPlainObject, mapObjectMap } from '../../../../mol-util/object';
import { onelinerJsonString } from '../../../../mol-util/json';
/** Type definition for a string  */
export var str = iots.string;
/** Type definition for an integer  */
export var int = iots.Integer;
/** Type definition for a float or integer number  */
export var float = iots.number;
/** Type definition for a boolean  */
export var bool = iots.boolean;
/** Type definition for a tuple, e.g. `tuple([str, int, int])`  */
export var tuple = iots.tuple;
/** Type definition for a list/array, e.g. `list(str)`  */
export var list = iots.array;
/** Type definition for union types, e.g. `union([str, int])` means string or integer  */
export var union = iots.union;
/** Type definition for nullable types, e.g. `nullable(str)` means string or `null`  */
export function nullable(type) {
    return union([type, iots.null]);
}
/** Type definition for literal types, e.g. `literal('red', 'green', 'blue')` means 'red' or 'green' or 'blue'  */
export function literal() {
    var values = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        values[_i] = arguments[_i];
    }
    if (values.length === 0) {
        throw new Error("literal type must have at least one value");
    }
    var typeName = "(".concat(values.map(function (v) { return onelinerJsonString(v); }).join(' | '), ")");
    return new iots.Type(typeName, (function (value) { return values.includes(value); }), function (value, ctx) { return values.includes(value) ? { _tag: 'Right', right: value } : { _tag: 'Left', left: [{ value: value, context: ctx, message: "\"".concat(value, "\" is not a valid value for literal type ").concat(typeName) }] }; }, function (value) { return value; });
}
export function RequiredField(type, description) {
    return { type: type, required: true, description: description };
}
export function OptionalField(type, description) {
    return { type: type, required: false, description: description };
}
/** Return `undefined` if `value` has correct type for `field`, regardsless of if required or optional.
 * Return description of validation issues, if `value` has wrong type. */
export function fieldValidationIssues(field, value) {
    var validation = field.type.decode(value);
    if (validation._tag === 'Right') {
        return undefined;
    }
    else {
        return PathReporter.report(validation);
    }
}
export function AllRequired(paramsSchema) {
    return mapObjectMap(paramsSchema, function (field) { return RequiredField(field.type, field.description); });
}
/** Return `undefined` if `values` contains correct value types for `schema`,
 * return description of validation issues, if `values` have wrong type.
 * If `options.requireAll`, all parameters (including optional) must have a value provided.
 * If `options.noExtra` is true, presence of any extra parameters is treated as an issue.
 */
export function paramsValidationIssues(schema, values, options) {
    if (options === void 0) { options = {}; }
    if (!isPlainObject(values))
        return ["Parameters must be an object, not ".concat(values)];
    for (var key in schema) {
        var paramDef = schema[key];
        if (Object.hasOwn(values, key)) {
            var value = values[key];
            var issues = fieldValidationIssues(paramDef, value);
            if (issues)
                return __spreadArray(["Invalid type for parameter \"".concat(key, "\":")], issues.map(function (s) { return '  ' + s; }), true);
        }
        else {
            if (paramDef.required)
                return ["Missing required parameter \"".concat(key, "\".")];
            if (options.requireAll)
                return ["Missing optional parameter \"".concat(key, "\".")];
        }
    }
    if (options.noExtra) {
        for (var key in values) {
            if (!Object.hasOwn(schema, key))
                return ["Unknown parameter \"".concat(key, "\".")];
        }
    }
    return undefined;
}
