// molstar/lib/extensions/mvs/tree/generic/tree-utils.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __assign, __spreadArray } from "tslib";
import { canonicalJsonString } from '../../../../mol-util/json';
import { getParams } from './tree-schema';
/** Run DFS (depth-first search) algorithm on a rooted tree.
 * Runs `visit` function when a node is discovered (before visiting any descendants).
 * Runs `postVisit` function when leaving a node (after all descendants have been visited). */
export function dfs(root, visit, postVisit) {
    return _dfs(root, undefined, visit, postVisit);
}
function _dfs(root, parent, visit, postVisit) {
    var _a;
    if (visit)
        visit(root, parent);
    for (var _i = 0, _b = (_a = root.children) !== null && _a !== void 0 ? _a : []; _i < _b.length; _i++) {
        var child = _b[_i];
        _dfs(child, root, visit, postVisit);
    }
    if (postVisit)
        postVisit(root, parent);
}
/** Convert a tree into a pretty-printed string. */
export function treeToString(tree) {
    var level = 0;
    var lines = [];
    dfs(tree, function (node) { var _a; return lines.push('  '.repeat(level++) + "- ".concat(node.kind, " ").concat(formatObject((_a = node.params) !== null && _a !== void 0 ? _a : {}))); }, function (node) { return level--; });
    return lines.join('\n');
}
/** Convert object to a human-friendly string (similar to JSON.stringify but without quoting keys) */
export function formatObject(obj) {
    if (!obj)
        return 'undefined';
    return JSON.stringify(obj).replace(/,("\w+":)/g, ', $1').replace(/"(\w+)":/g, '$1: ');
}
/** Create a copy of a tree node, ignoring children. */
export function copyNodeWithoutChildren(node) {
    return {
        kind: node.kind,
        params: node.params ? __assign({}, node.params) : undefined,
    };
}
/** Create a copy of a tree node, including a shallow copy of children. */
export function copyNode(node) {
    return {
        kind: node.kind,
        params: node.params ? __assign({}, node.params) : undefined,
        children: node.children ? __spreadArray([], node.children, true) : undefined,
    };
}
/** Create a deep copy of a tree. */
export function copyTree(root) {
    return convertTree(root, {});
}
/** Apply a set of conversion rules to a tree to change to a different schema. */
export function convertTree(root, conversions) {
    var mapping = new Map();
    var convertedRoot;
    dfs(root, function (node, parent) {
        var _a, _b;
        var _c;
        var conversion = conversions[node.kind];
        if (conversion) {
            var convertidos = conversion(node, parent);
            if (!parent && convertidos.length === 0)
                throw new Error('Cannot convert root to empty path');
            var convParent = parent ? mapping.get(parent) : undefined;
            for (var _i = 0, convertidos_1 = convertidos; _i < convertidos_1.length; _i++) {
                var conv = convertidos_1[_i];
                if (convParent) {
                    ((_a = convParent.children) !== null && _a !== void 0 ? _a : (convParent.children = [])).push(conv);
                }
                else {
                    convertedRoot = conv;
                }
                convParent = conv;
            }
            mapping.set(node, convParent);
        }
        else {
            var converted = copyNodeWithoutChildren(node);
            if (parent) {
                ((_b = (_c = mapping.get(parent)).children) !== null && _b !== void 0 ? _b : (_c.children = [])).push(converted);
            }
            else {
                convertedRoot = converted;
            }
            mapping.set(node, converted);
        }
    });
    return convertedRoot;
}
/** Create a copy of the tree where twins (siblings of the same kind with the same params) are merged into one node.
 * Applies only to the node kinds listed in `condenseNodes` (or all if undefined) except node kinds in `skipNodes`. */
export function condenseTree(root, condenseNodes, skipNodes) {
    var map = new Map();
    var result = copyTree(root);
    dfs(result, function (node) {
        var _a;
        var _b, _c, _d;
        map.clear();
        var newChildren = [];
        for (var _i = 0, _e = (_b = node.children) !== null && _b !== void 0 ? _b : []; _i < _e.length; _i++) {
            var child = _e[_i];
            var twin = undefined;
            var doApply = (!condenseNodes || condenseNodes.has(child.kind)) && !(skipNodes === null || skipNodes === void 0 ? void 0 : skipNodes.has(child.kind));
            if (doApply) {
                var key = child.kind + canonicalJsonString(getParams(child));
                twin = map.get(key);
                if (!twin)
                    map.set(key, child);
            }
            if (twin) {
                (_a = ((_c = twin.children) !== null && _c !== void 0 ? _c : (twin.children = []))).push.apply(_a, (_d = child.children) !== null && _d !== void 0 ? _d : []);
            }
            else {
                newChildren.push(child);
            }
        }
        node.children = newChildren;
    });
    return result;
}
/** Create a copy of the tree where missing optional params for each node are added based on `defaults`. */
export function addDefaults(tree, defaults) {
    var rules = {};
    var _loop_1 = function (kind) {
        rules[kind] = function (node) { return [{ kind: node.kind, params: __assign(__assign({}, defaults[kind]), node.params) }]; };
    };
    for (var kind in defaults) {
        _loop_1(kind);
    }
    return convertTree(tree, rules);
}
/** Resolve any URI params in a tree, in place. URI params are those listed in `uriParamNames`.
 * Relative URIs are treated as relative to `baseUri`, which can in turn be relative to the window URL (if available). */
export function resolveUris(tree, baseUri, uriParamNames) {
    dfs(tree, function (node) {
        var params = node.params;
        if (!params)
            return;
        for (var name_1 in params) {
            if (uriParamNames.includes(name_1)) {
                var uri = params[name_1];
                if (typeof uri === 'string') {
                    params[name_1] = resolveUri(uri, baseUri, windowUrl());
                }
            }
        }
    });
}
/** Resolve a sequence of URI references (relative URIs), where each reference is either absolute or relative to the next one
 * (i.e. the last one is the base URI). Skip any `undefined`.
 * E.g. `resolveUri('./unexpected.png', '/spanish/inquisition/expectations.html', 'https://example.org/spam/spam/spam')`
 * returns `'https://example.org/spanish/inquisition/unexpected.png'`. */
function resolveUri() {
    var refs = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        refs[_i] = arguments[_i];
    }
    var result = undefined;
    for (var _a = 0, _b = refs.reverse(); _a < _b.length; _a++) {
        var ref = _b[_a];
        if (ref !== undefined) {
            if (result === undefined)
                result = ref;
            else
                result = new URL(ref, result).href;
        }
    }
    return result;
}
/** Return URL of the current page when running in a browser; `undefined` when running in Node. */
function windowUrl() {
    return (typeof window !== 'undefined') ? window.location.href : undefined;
}
