// molstar/lib/commonjs/mol-repr/structure/visual/util/nucleotide.js
"use strict";
/**
 * Copyright (c) 2018-2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasSugarIndices = exports.setSugarIndices = exports.hasPyrimidineIndices = exports.setPyrimidineIndices = exports.hasPurinIndices = exports.setPurinIndices = exports.createNucleicIndices = exports.getNucleotideBaseType = exports.eachNucleotideElement = exports.getNucleotideElementLoci = exports.NucleotideLocationIterator = void 0;
var structure_1 = require("../../../../mol-model/structure");
var loci_1 = require("../../../../mol-model/loci");
var location_iterator_1 = require("../../../../mol-geo/util/location-iterator");
var common_1 = require("./common");
var polymer_1 = require("./polymer");
var types_1 = require("../../../../mol-model/structure/model/types");
var vec3_1 = require("../../../../mol-math/linear-algebra/3d/vec3");
var NucleotideLocationIterator;
(function (NucleotideLocationIterator) {
    function fromGroup(structureGroup) {
        var group = structureGroup.group, structure = structureGroup.structure;
        var u = group.units[0];
        var nucleotideElementIndices = structure_1.Unit.isAtomic(u) ? u.nucleotideElements : [];
        var groupCount = nucleotideElementIndices.length;
        var instanceCount = group.units.length;
        var location = structure_1.StructureElement.Location.create(structure);
        var getLocation = function (groupIndex, instanceIndex) {
            var unit = group.units[instanceIndex];
            location.unit = unit;
            location.element = nucleotideElementIndices[groupIndex];
            return location;
        };
        return (0, location_iterator_1.LocationIterator)(groupCount, instanceCount, 1, getLocation);
    }
    NucleotideLocationIterator.fromGroup = fromGroup;
})(NucleotideLocationIterator || (exports.NucleotideLocationIterator = NucleotideLocationIterator = {}));
function getNucleotideElementLoci(pickingId, structureGroup, id) {
    var objectId = pickingId.objectId, instanceId = pickingId.instanceId, groupId = pickingId.groupId;
    if (id === objectId) {
        var structure = structureGroup.structure, group = structureGroup.group;
        var unit = group.units[instanceId];
        if (structure_1.Unit.isAtomic(unit)) {
            return (0, common_1.getResidueLoci)(structure, unit, unit.nucleotideElements[groupId]);
        }
    }
    return loci_1.EmptyLoci;
}
exports.getNucleotideElementLoci = getNucleotideElementLoci;
function selectNuclotideElements(u) { return u.nucleotideElements; }
/**
 * Mark a nucleotide element (e.g. part of a cartoon block)
 * - mark only when all its residue's elements are in a loci
 */
function eachNucleotideElement(loci, structureGroup, apply) {
    var changed = false;
    if (!structure_1.StructureElement.Loci.is(loci))
        return false;
    var structure = structureGroup.structure, group = structureGroup.group;
    if (!structure_1.Structure.areEquivalent(loci.structure, structure))
        return false;
    var unit = group.units[0];
    if (!structure_1.Unit.isAtomic(unit))
        return false;
    var nucleotideElements = unit.nucleotideElements;
    var groupCount = nucleotideElements.length;
    for (var _i = 0, _a = loci.elements; _i < _a.length; _i++) {
        var e = _a[_i];
        if (!structure_1.Unit.isAtomic(e.unit))
            continue;
        if (!group.unitIndexMap.has(e.unit.id))
            continue;
        var intervalOffset = group.unitIndexMap.get(e.unit.id) * groupCount;
        if (structure_1.Unit.isAtomic(e.unit)) {
            changed = (0, polymer_1.eachAtomicUnitTracedElement)(intervalOffset, groupCount, selectNuclotideElements, apply, e) || changed;
        }
    }
    return changed;
}
exports.eachNucleotideElement = eachNucleotideElement;
//
var pC4 = (0, vec3_1.Vec3)();
var pN9 = (0, vec3_1.Vec3)();
function getNucleotideBaseType(unit, residueIndex) {
    var model = unit.model;
    var _a = model.atomicHierarchy, residueAtomSegments = _a.residueAtomSegments, atoms = _a.atoms, atomicIndex = _a.index;
    var label_comp_id = atoms.label_comp_id;
    var pos = unit.conformation.invariantPosition;
    var compId = label_comp_id.value(residueAtomSegments.offsets[residueIndex]);
    var isPurine = (0, types_1.isPurineBase)(compId);
    var isPyrimidine = (0, types_1.isPyrimidineBase)(compId);
    if (!isPurine && !isPyrimidine) {
        // detect Purine or Pyrimidin based on geometry
        var idxC4 = atomicIndex.findAtomOnResidue(residueIndex, 'C4');
        var idxN9 = atomicIndex.findAtomOnResidue(residueIndex, 'N9');
        if (idxC4 !== -1 && idxN9 !== -1 && vec3_1.Vec3.distance(pos(idxC4, pC4), pos(idxN9, pN9)) < 1.6) {
            isPurine = true;
        }
        else {
            isPyrimidine = true;
        }
    }
    return { isPurine: isPurine, isPyrimidine: isPyrimidine };
}
exports.getNucleotideBaseType = getNucleotideBaseType;
function createNucleicIndices() {
    return {
        trace: -1,
        N1: -1,
        C2: -1,
        N3: -1,
        C4: -1,
        C5: -1,
        C6: -1,
        N7: -1,
        C8: -1,
        N9: -1,
        C1_1: -1,
        C2_1: -1,
        C3_1: -1,
        C4_1: -1,
        O4_1: -1,
    };
}
exports.createNucleicIndices = createNucleicIndices;
function setPurinIndices(idx, unit, residueIndex) {
    var atomicIndex = unit.model.atomicHierarchy.index;
    var traceElementIndex = unit.model.atomicHierarchy.derived.residue.traceElementIndex;
    idx.trace = traceElementIndex[residueIndex];
    idx.N1 = atomicIndex.findAtomOnResidue(residueIndex, 'N1');
    idx.C2 = atomicIndex.findAtomOnResidue(residueIndex, 'C2');
    idx.N3 = atomicIndex.findAtomOnResidue(residueIndex, 'N3');
    idx.C4 = atomicIndex.findAtomOnResidue(residueIndex, 'C4');
    idx.C5 = atomicIndex.findAtomOnResidue(residueIndex, 'C5');
    if (idx.C5 === -1) {
        // modified ring, e.g. DP
        idx.C5 = atomicIndex.findAtomOnResidue(residueIndex, 'N5');
    }
    idx.C6 = atomicIndex.findAtomOnResidue(residueIndex, 'C6');
    idx.N7 = atomicIndex.findAtomOnResidue(residueIndex, 'N7');
    if (idx.N7 === -1) {
        // modified ring, e.g. DP
        idx.N7 = atomicIndex.findAtomOnResidue(residueIndex, 'C7');
    }
    idx.C8 = atomicIndex.findAtomOnResidue(residueIndex, 'C8');
    idx.N9 = atomicIndex.findAtomOnResidue(residueIndex, 'N9');
    return idx;
}
exports.setPurinIndices = setPurinIndices;
function hasPurinIndices(idx) {
    return idx.trace !== -1 && idx.N1 !== -1 && idx.C2 !== -1 && idx.N3 !== -1 && idx.C4 !== -1 && idx.C5 !== -1 && idx.C6 !== -1 && idx.N7 !== -1 && idx.C8 !== -1 && idx.N9 !== -1;
}
exports.hasPurinIndices = hasPurinIndices;
function setPyrimidineIndices(idx, unit, residueIndex) {
    var atomicIndex = unit.model.atomicHierarchy.index;
    var traceElementIndex = unit.model.atomicHierarchy.derived.residue.traceElementIndex;
    idx.trace = traceElementIndex[residueIndex];
    idx.N1 = atomicIndex.findAtomOnResidue(residueIndex, 'N1');
    if (idx.N1 === -1) {
        // modified ring, e.g. DZ
        idx.N1 = atomicIndex.findAtomOnResidue(residueIndex, 'C1');
    }
    idx.C2 = atomicIndex.findAtomOnResidue(residueIndex, 'C2');
    idx.N3 = atomicIndex.findAtomOnResidue(residueIndex, 'N3');
    idx.C4 = atomicIndex.findAtomOnResidue(residueIndex, 'C4');
    idx.C5 = atomicIndex.findAtomOnResidue(residueIndex, 'C5');
    idx.C6 = atomicIndex.findAtomOnResidue(residueIndex, 'C6');
    return idx;
}
exports.setPyrimidineIndices = setPyrimidineIndices;
function hasPyrimidineIndices(idx) {
    return idx.trace !== -1 && idx.N1 !== -1 && idx.C2 !== -1 && idx.N3 !== -1 && idx.C4 !== -1 && idx.C5 !== -1 && idx.C6 !== -1;
}
exports.hasPyrimidineIndices = hasPyrimidineIndices;
function setSugarIndices(idx, unit, residueIndex) {
    var atomicIndex = unit.model.atomicHierarchy.index;
    var traceElementIndex = unit.model.atomicHierarchy.derived.residue.traceElementIndex;
    idx.trace = traceElementIndex[residueIndex];
    idx.C1_1 = atomicIndex.findAtomOnResidue(residueIndex, "C1'");
    idx.C2_1 = atomicIndex.findAtomOnResidue(residueIndex, "C2'");
    idx.C3_1 = atomicIndex.findAtomOnResidue(residueIndex, "C3'");
    idx.C4_1 = atomicIndex.findAtomOnResidue(residueIndex, "C4'");
    idx.O4_1 = atomicIndex.findAtomOnResidue(residueIndex, "O4'");
    return idx;
}
exports.setSugarIndices = setSugarIndices;
function hasSugarIndices(idx) {
    return idx.trace !== -1 && idx.C1_1 !== -1 && idx.C2_1 !== -1 && idx.C3_1 !== -1 && idx.C4_1 !== -1 && idx.O4_1 !== -1;
}
exports.hasSugarIndices = hasSugarIndices;
