// molstar/lib/commonjs/extensions/mvs/tree/mvs/mvs-builder.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.builderDemo = exports.Representation = exports.Component = exports.Structure = exports.Parse = exports.Download = exports.Root = exports.createMVSBuilder = void 0;
var tslib_1 = require("tslib");
var object_1 = require("../../../../mol-util/object");
var mvs_data_1 = require("../../mvs-data");
var mvs_defaults_1 = require("./mvs-defaults");
var mvs_tree_1 = require("./mvs-tree");
/** Create a new MolViewSpec builder containing only a root node. Example of MVS builder usage:
 *
 * ```
 * const builder = createMVSBuilder();
 * builder.canvas({ background_color: 'white' });
 * const struct = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og2_updated.cif' }).parse({ format: 'mmcif' }).modelStructure();
 * struct.component().representation().color({ color: '#3050F8' });
 * console.log(JSON.stringify(builder.getState()));
 * ```
 */
function createMVSBuilder() {
    return new Root();
}
exports.createMVSBuilder = createMVSBuilder;
/** Base class for MVS builder pointing to anything */
var _Base = /** @class */ (function () {
    function _Base(_root, _node) {
        this._root = _root;
        this._node = _node;
    }
    /** Create a new node, append as child to current _node, and return the new node */
    _Base.prototype.addChild = function (kind, params) {
        var _a;
        var _b;
        var allowedParamNames = Object.keys(mvs_tree_1.MVSTreeSchema.nodes[kind].params);
        var node = {
            kind: kind,
            params: (0, object_1.pickObjectKeys)(params, allowedParamNames),
        };
        (_a = (_b = this._node).children) !== null && _a !== void 0 ? _a : (_b.children = []);
        this._node.children.push(node);
        return node;
    };
    return _Base;
}());
/** MVS builder pointing to the 'root' node */
var Root = /** @class */ (function (_super) {
    tslib_1.__extends(Root, _super);
    function Root() {
        var _this = this;
        var node = { kind: 'root' };
        _this = _super.call(this, undefined, node) || this;
        _this._root = _this;
        return _this;
    }
    /** Return the current state of the builder as object in MVS format. */
    Root.prototype.getState = function (metadata) {
        return {
            root: (0, object_1.deepClone)(this._node),
            metadata: tslib_1.__assign(tslib_1.__assign({}, metadata), { version: "".concat(mvs_data_1.MVSData.SupportedVersion), timestamp: utcNowISO() }),
        };
    };
    // omitting `saveState`, filesystem operations are responsibility of the caller code (platform-dependent)
    /** Add a 'camera' node and return builder pointing to the root. 'camera' node instructs to set the camera position and orientation. */
    Root.prototype.camera = function (params) {
        this.addChild('camera', params);
        return this;
    };
    /** Add a 'canvas' node and return builder pointing to the root. 'canvas' node sets canvas properties. */
    Root.prototype.canvas = function (params) {
        this.addChild('canvas', params);
        return this;
    };
    /** Add a 'download' node and return builder pointing to it. 'download' node instructs to retrieve a data resource. */
    Root.prototype.download = function (params) {
        return new Download(this._root, this.addChild('download', params));
    };
    return Root;
}(_Base));
exports.Root = Root;
/** MVS builder pointing to a 'download' node */
var Download = /** @class */ (function (_super) {
    tslib_1.__extends(Download, _super);
    function Download() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /** Add a 'parse' node and return builder pointing to it. 'parse' node instructs to parse a data resource. */
    Download.prototype.parse = function (params) {
        return new Parse(this._root, this.addChild('parse', params));
    };
    return Download;
}(_Base));
exports.Download = Download;
/** Subsets of 'structure' node params which will be passed to individual builder functions. */
var StructureParamsSubsets = {
    model: ['block_header', 'block_index', 'model_index'],
    assembly: ['block_header', 'block_index', 'model_index', 'assembly_id'],
    symmetry: ['block_header', 'block_index', 'model_index', 'ijk_min', 'ijk_max'],
    symmetry_mates: ['block_header', 'block_index', 'model_index', 'radius'],
};
/** MVS builder pointing to a 'parse' node */
var Parse = /** @class */ (function (_super) {
    tslib_1.__extends(Parse, _super);
    function Parse() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /** Add a 'structure' node representing a "model structure", i.e. includes all coordinates from the original model without applying any transformations.
     * Return builder pointing to the new node. */
    Parse.prototype.modelStructure = function (params) {
        if (params === void 0) { params = {}; }
        return new Structure(this._root, this.addChild('structure', tslib_1.__assign({ type: 'model' }, (0, object_1.pickObjectKeys)(params, StructureParamsSubsets.model))));
    };
    /** Add a 'structure' node representing an "assembly structure", i.e. may apply filters and symmetry operators to the original model coordinates.
     * Return builder pointing to the new node. */
    Parse.prototype.assemblyStructure = function (params) {
        if (params === void 0) { params = {}; }
        return new Structure(this._root, this.addChild('structure', tslib_1.__assign({ type: 'assembly' }, (0, object_1.pickObjectKeys)(params, StructureParamsSubsets.assembly))));
    };
    /** Add a 'structure' node representing a "symmetry structure", i.e. applies symmetry operators to build crystal unit cells within given Miller indices.
     * Return builder pointing to the new node. */
    Parse.prototype.symmetryStructure = function (params) {
        if (params === void 0) { params = {}; }
        return new Structure(this._root, this.addChild('structure', tslib_1.__assign({ type: 'symmetry' }, (0, object_1.pickObjectKeys)(params, StructureParamsSubsets.symmetry))));
    };
    /** Add a 'structure' node representing a "symmetry mates structure", i.e. applies symmetry operators to build asymmetric units within a radius from the original model.
     * Return builder pointing to the new node. */
    Parse.prototype.symmetryMatesStructure = function (params) {
        if (params === void 0) { params = {}; }
        return new Structure(this._root, this.addChild('structure', tslib_1.__assign({ type: 'symmetry_mates' }, (0, object_1.pickObjectKeys)(params, StructureParamsSubsets.symmetry_mates))));
    };
    return Parse;
}(_Base));
exports.Parse = Parse;
/** MVS builder pointing to a 'structure' node */
var Structure = /** @class */ (function (_super) {
    tslib_1.__extends(Structure, _super);
    function Structure() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /** Add a 'component' node and return builder pointing to it. 'component' node instructs to create a component (i.e. a subset of the parent structure). */
    Structure.prototype.component = function (params) {
        var _a;
        if (params === void 0) { params = {}; }
        var fullParams = tslib_1.__assign(tslib_1.__assign({}, params), { selector: (_a = params.selector) !== null && _a !== void 0 ? _a : mvs_defaults_1.MVSDefaults.component.selector });
        return new Component(this._root, this.addChild('component', fullParams));
    };
    /** Add a 'component_from_uri' node and return builder pointing to it. 'component_from_uri' node instructs to create a component defined by an external annotation resource. */
    Structure.prototype.componentFromUri = function (params) {
        return new Component(this._root, this.addChild('component_from_uri', params));
    };
    /** Add a 'component_from_source' node and return builder pointing to it. 'component_from_source' node instructs to create a component defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    Structure.prototype.componentFromSource = function (params) {
        return new Component(this._root, this.addChild('component_from_source', params));
    };
    /** Add a 'label_from_uri' node and return builder pointing back to the structure node. 'label_from_uri' node instructs to add labels (textual visual representations) to parts of a structure. The labels are defined by an external annotation resource. */
    Structure.prototype.labelFromUri = function (params) {
        this.addChild('label_from_uri', params);
        return this;
    };
    /** Add a 'label_from_source' node and return builder pointing back to the structure node. 'label_from_source' node instructs to add labels (textual visual representations) to parts of a structure. The labels are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    Structure.prototype.labelFromSource = function (params) {
        this.addChild('label_from_source', params);
        return this;
    };
    /** Add a 'tooltip_from_uri' node and return builder pointing back to the structure node. 'tooltip_from_uri' node instructs to add tooltips to parts of a structure. The tooltips are defined by an external annotation resource. */
    Structure.prototype.tooltipFromUri = function (params) {
        this.addChild('tooltip_from_uri', params);
        return this;
    };
    /** Add a 'tooltip_from_source' node and return builder pointing back to the structure node. 'tooltip_from_source' node instructs to add tooltips to parts of a structure. The tooltips are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    Structure.prototype.tooltipFromSource = function (params) {
        this.addChild('tooltip_from_source', params);
        return this;
    };
    /** Add a 'transform' node and return builder pointing back to the structure node. 'transform' node instructs to rotate and/or translate structure coordinates. */
    Structure.prototype.transform = function (params) {
        if (params === void 0) { params = {}; }
        if (params.rotation && params.rotation.length !== 9) {
            throw new Error('ValueError: `rotation` parameter must be an array of 9 numbers');
        }
        this.addChild('transform', params);
        return this;
    };
    return Structure;
}(_Base));
exports.Structure = Structure;
/** MVS builder pointing to a 'component' or 'component_from_uri' or 'component_from_source' node */
var Component = /** @class */ (function (_super) {
    tslib_1.__extends(Component, _super);
    function Component() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /** Add a 'representation' node and return builder pointing to it. 'representation' node instructs to create a visual representation of a component. */
    Component.prototype.representation = function (params) {
        var _a;
        if (params === void 0) { params = {}; }
        var fullParams = tslib_1.__assign(tslib_1.__assign({}, params), { type: (_a = params.type) !== null && _a !== void 0 ? _a : 'cartoon' });
        return new Representation(this._root, this.addChild('representation', fullParams));
    };
    /** Add a 'label' node and return builder pointing back to the component node. 'label' node instructs to add a label (textual visual representation) to a component. */
    Component.prototype.label = function (params) {
        this.addChild('label', params);
        return this;
    };
    /** Add a 'tooltip' node and return builder pointing back to the component node. 'tooltip' node instructs to add a text which is not a part of the visualization but should be presented to the users when they interact with the component (typically, the tooltip will be shown somewhere on the screen when the user hovers over a visual representation of the component). */
    Component.prototype.tooltip = function (params) {
        this.addChild('tooltip', params);
        return this;
    };
    /** Add a 'focus' node and return builder pointing back to the component node. 'focus' node instructs to set the camera focus to a component (zoom in). */
    Component.prototype.focus = function (params) {
        if (params === void 0) { params = {}; }
        this.addChild('focus', params);
        return this;
    };
    return Component;
}(_Base));
exports.Component = Component;
/** MVS builder pointing to a 'representation' node */
var Representation = /** @class */ (function (_super) {
    tslib_1.__extends(Representation, _super);
    function Representation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /** Add a 'color' node and return builder pointing back to the representation node. 'color' node instructs to apply color to a visual representation. */
    Representation.prototype.color = function (params) {
        this.addChild('color', params);
        return this;
    };
    /** Add a 'color_from_uri' node and return builder pointing back to the representation node. 'color_from_uri' node instructs to apply colors to a visual representation. The colors are defined by an external annotation resource. */
    Representation.prototype.colorFromUri = function (params) {
        this.addChild('color_from_uri', params);
        return this;
    };
    /** Add a 'color_from_source' node and return builder pointing back to the representation node. 'color_from_source' node instructs to apply colors to a visual representation. The colors are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    Representation.prototype.colorFromSource = function (params) {
        this.addChild('color_from_source', params);
        return this;
    };
    return Representation;
}(_Base));
exports.Representation = Representation;
/** Demonstration of usage of MVS builder */
function builderDemo() {
    var builder = createMVSBuilder();
    builder.canvas({ background_color: 'white' });
    var struct = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og2_updated.cif' }).parse({ format: 'mmcif' }).modelStructure();
    struct.component().representation().color({ color: 'white' });
    struct.component({ selector: 'ligand' }).representation({ type: 'ball_and_stick' })
        .color({ color: '#555555' })
        .color({ selector: { type_symbol: 'N' }, color: '#3050F8' })
        .color({ selector: { type_symbol: 'O' }, color: '#FF0D0D' })
        .color({ selector: { type_symbol: 'S' }, color: '#FFFF30' })
        .color({ selector: { type_symbol: 'FE' }, color: '#E06633' });
    builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og5_updated.cif' }).parse({ format: 'mmcif' }).assemblyStructure({ assembly_id: '1' }).component().representation().color({ color: 'cyan' });
    builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og5_updated.cif' }).parse({ format: 'mmcif' }).assemblyStructure({ assembly_id: '2' }).component().representation().color({ color: 'blue' });
    var cif = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1wrf_updated.cif' }).parse({ format: 'mmcif' });
    cif.modelStructure({ model_index: 0 }).component().representation().color({ color: '#CC0000' });
    cif.modelStructure({ model_index: 1 }).component().representation().color({ color: '#EE7700' });
    cif.modelStructure({ model_index: 2 }).component().representation().color({ color: '#FFFF00' });
    cif.modelStructure({ model_index: 0 }).transform({ translation: [30, 0, 0] }).component().representation().color({ color: '#ff88bb' });
    cif.modelStructure({ model_index: 0 }).transform({ translation: [60, 0, 0], rotation: [0, 1, 0, -1, 0, 0, 0, 0, 1] }).component().representation().color({ color: '#aa0077' });
    return builder.getState();
}
exports.builderDemo = builderDemo;
/** Return the current universal time, in ISO format, e.g. '2023-11-24T10:45:49.873Z' */
function utcNowISO() {
    return new Date().toISOString();
}
