// molstar/lib/commonjs/extensions/mvs/components/annotation-structure-component.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateMVSAnnotationStructureComponent = exports.createMVSAnnotationStructureComponent = exports.createMVSAnnotationSubstructure = exports.MVSAnnotationStructureComponent = exports.MVSTransform = exports.MVSAnnotationStructureComponentParams = void 0;
var structure_1 = require("../../../mol-model/structure");
var structure_query_1 = require("../../../mol-plugin-state/helpers/structure-query");
var objects_1 = require("../../../mol-plugin-state/objects");
var mol_state_1 = require("../../../mol-state");
var mol_util_1 = require("../../../mol-util");
var object_1 = require("../../../mol-util/object");
var param_definition_1 = require("../../../mol-util/param-definition");
var selections_1 = require("../helpers/selections");
var annotation_prop_1 = require("./annotation-prop");
/** Parameter definition for `MVSAnnotationStructureComponent` transformer */
exports.MVSAnnotationStructureComponentParams = {
    annotationId: param_definition_1.ParamDefinition.Text('', { description: 'Reference to "Annotation" custom model property' }),
    fieldName: param_definition_1.ParamDefinition.Text('component', { description: 'Annotation field (column) from which to take component identifier' }),
    fieldValues: param_definition_1.ParamDefinition.MappedStatic('all', {
        all: param_definition_1.ParamDefinition.EmptyGroup(),
        selected: param_definition_1.ParamDefinition.ObjectList({
            value: param_definition_1.ParamDefinition.Text(),
        }, function (obj) { return obj.value; }),
    }),
    nullIfEmpty: param_definition_1.ParamDefinition.Optional(param_definition_1.ParamDefinition.Boolean(true, { isHidden: false })),
    label: param_definition_1.ParamDefinition.Text('', { isHidden: false }),
};
/** Transformer builder for MVS extension */
exports.MVSTransform = mol_state_1.StateTransformer.builderFactory('mvs');
exports.MVSAnnotationStructureComponent = (0, exports.MVSTransform)({
    name: 'mvs-structure-component-from-annotation',
    display: { name: 'MVS Annotation Component', description: 'A molecular structure component defined by MVS annotation data.' },
    from: objects_1.PluginStateObject.Molecule.Structure,
    to: objects_1.PluginStateObject.Molecule.Structure,
    params: exports.MVSAnnotationStructureComponentParams,
})({
    apply: function (_a) {
        var a = _a.a, params = _a.params;
        return createMVSAnnotationStructureComponent(a.data, params);
    },
    update: function (_a) {
        var a = _a.a, b = _a.b, oldParams = _a.oldParams, newParams = _a.newParams;
        return updateMVSAnnotationStructureComponent(a.data, b, oldParams, newParams);
    },
    dispose: function (_a) {
        var b = _a.b;
        b === null || b === void 0 ? void 0 : b.data.customPropertyDescriptors.dispose();
    }
});
/** Create a substructure based on `MVSAnnotationStructureComponentProps` */
function createMVSAnnotationSubstructure(structure, params) {
    var annotation = (0, annotation_prop_1.getMVSAnnotationForStructure)(structure, params.annotationId).annotation;
    if (annotation) {
        var rows = annotation.getRows();
        if (params.fieldValues.name === 'selected') {
            var selectedValues_1 = new Set(params.fieldValues.params.map(function (obj) { return obj.value; }));
            rows = rows.filter(function (row, i) { return selectedValues_1.has(annotation.getValueForRow(i, params.fieldName)); });
        }
        var expression = (0, selections_1.rowsToExpression)(rows);
        var selection = structure_query_1.StructureQueryHelper.createAndRun(structure, expression).selection;
        return structure_1.StructureSelection.unionStructure(selection);
    }
    else {
        return structure_1.Structure.Empty;
    }
}
exports.createMVSAnnotationSubstructure = createMVSAnnotationSubstructure;
/** Create a substructure PSO based on `MVSAnnotationStructureComponentProps` */
function createMVSAnnotationStructureComponent(structure, params) {
    var component = createMVSAnnotationSubstructure(structure, params);
    if (params.nullIfEmpty && component.elementCount === 0)
        return mol_state_1.StateObject.Null;
    var label = params.label;
    if (label === undefined || label === '') {
        if (params.fieldValues.name === 'selected' && params.fieldValues.params.length > 0) {
            var values = params.fieldValues.params;
            var valuesStr = "\"".concat(values[0].value, "\"");
            if (values.length === 2) {
                valuesStr += " + \"".concat(values[1].value, "\"");
            }
            else if (values.length > 2) {
                valuesStr += " + ".concat(values.length - 1, " more values");
            }
            label = "MVS Annotation Component (".concat(params.fieldName, ": ").concat(valuesStr, ")");
        }
        else {
            label = 'MVS Annotation Component';
        }
    }
    var props = { label: label, description: structure_1.Structure.elementDescription(component) };
    return new objects_1.PluginStateObject.Molecule.Structure(component, props);
}
exports.createMVSAnnotationStructureComponent = createMVSAnnotationStructureComponent;
/** Update a substructure PSO based on `MVSAnnotationStructureComponentProps` */
function updateMVSAnnotationStructureComponent(a, b, oldParams, newParams) {
    var change = !(0, mol_util_1.deepEqual)(newParams, oldParams);
    var needsRecreate = !(0, mol_util_1.deepEqual)((0, object_1.omitObjectKeys)(newParams, ['label']), (0, object_1.omitObjectKeys)(oldParams, ['label']));
    if (!change) {
        return mol_state_1.StateTransformer.UpdateResult.Unchanged;
    }
    if (!needsRecreate) {
        b.label = newParams.label || b.label;
        return mol_state_1.StateTransformer.UpdateResult.Updated;
    }
    return mol_state_1.StateTransformer.UpdateResult.Recreate;
}
exports.updateMVSAnnotationStructureComponent = updateMVSAnnotationStructureComponent;
