// molstar/lib/commonjs/extensions/mvs/components/annotation-tooltips-prop.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MVSAnnotationTooltipsLabelProvider = exports.MVSAnnotationTooltipsProvider = exports.MVSAnnotationTooltipsParams = void 0;
var tslib_1 = require("tslib");
var custom_structure_property_1 = require("../../../mol-model-props/common/custom-structure-property");
var custom_property_1 = require("../../../mol-model/custom-property");
var structure_1 = require("../../../mol-model/structure");
var param_definition_1 = require("../../../mol-util/param-definition");
var utils_1 = require("../helpers/utils");
var annotation_prop_1 = require("./annotation-prop");
/** Parameter definition for custom structure property "MVSAnnotationTooltips" */
exports.MVSAnnotationTooltipsParams = {
    tooltips: param_definition_1.ParamDefinition.ObjectList({
        annotationId: param_definition_1.ParamDefinition.Text('', { description: 'Reference to "MVS Annotation" custom model property' }),
        fieldName: param_definition_1.ParamDefinition.Text('tooltip', { description: 'Annotation field (column) from which to take color values' }),
    }, function (obj) { return "".concat(obj.annotationId, ":").concat(obj.fieldName); }),
};
/** Provider for custom structure property "MVSAnnotationTooltips" */
exports.MVSAnnotationTooltipsProvider = custom_structure_property_1.CustomStructureProperty.createProvider({
    label: 'MVS Annotation Tooltips',
    descriptor: (0, custom_property_1.CustomPropertyDescriptor)({
        name: 'mvs-annotation-tooltips',
    }),
    type: 'local',
    defaultParams: exports.MVSAnnotationTooltipsParams,
    getParams: function (data) { return exports.MVSAnnotationTooltipsParams; },
    isApplicable: function (data) { return data.root === data; },
    obtain: function (ctx, data, props) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var fullProps;
        return tslib_1.__generator(this, function (_a) {
            fullProps = tslib_1.__assign(tslib_1.__assign({}, param_definition_1.ParamDefinition.getDefaultValues(exports.MVSAnnotationTooltipsParams)), props);
            return [2 /*return*/, { value: fullProps }];
        });
    }); },
});
/** Label provider based on data from "MVS Annotation" custom model property */
exports.MVSAnnotationTooltipsLabelProvider = {
    label: function (loci) {
        switch (loci.kind) {
            case 'element-loci':
                if (!loci.structure.customPropertyDescriptors.hasReference(exports.MVSAnnotationTooltipsProvider.descriptor))
                    return undefined;
                var location_1 = structure_1.StructureElement.Loci.getFirstLocation(loci);
                if (!location_1)
                    return undefined;
                var tooltipProps = exports.MVSAnnotationTooltipsProvider.get(location_1.structure).value;
                if (!tooltipProps || tooltipProps.tooltips.length === 0)
                    return undefined;
                var annotations_1 = annotation_prop_1.MVSAnnotationsProvider.get(location_1.unit.model).value;
                var texts = tooltipProps.tooltips.map(function (p) { var _a; return (_a = annotations_1 === null || annotations_1 === void 0 ? void 0 : annotations_1.getAnnotation(p.annotationId)) === null || _a === void 0 ? void 0 : _a.getValueForLocation(location_1, p.fieldName); });
                return (0, utils_1.filterDefined)(texts).join(' | ');
            default:
                return undefined;
        }
    }
};
