// molstar/lib/commonjs/extensions/mvs/components/annotation-color-theme.js
"use strict";
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MVSAnnotationColorThemeProvider = exports.MVSAnnotationColorTheme = exports.MVSAnnotationColorThemeParams = void 0;
var structure_1 = require("../../../mol-model/structure");
var color_1 = require("../../../mol-theme/color");
var names_1 = require("../../../mol-util/color/names");
var param_definition_1 = require("../../../mol-util/param-definition");
var utils_1 = require("../helpers/utils");
var annotation_prop_1 = require("./annotation-prop");
var is_mvs_model_prop_1 = require("./is-mvs-model-prop");
/** Parameter definition for color theme "MVS Annotation" */
exports.MVSAnnotationColorThemeParams = {
    annotationId: param_definition_1.ParamDefinition.Text('', { description: 'Reference to "Annotation" custom model property' }),
    fieldName: param_definition_1.ParamDefinition.Text('color', { description: 'Annotation field (column) from which to take color values' }),
    background: param_definition_1.ParamDefinition.Color(names_1.ColorNames.gainsboro, { description: 'Color for elements without annotation' }),
};
/** Return color theme that assigns colors based on an annotation file.
 * The annotation file itself is handled by a custom model property (`MVSAnnotationsProvider`),
 * the color theme then just uses this property. */
function MVSAnnotationColorTheme(ctx, props) {
    var color = function () { return props.background; };
    if (ctx.structure && !ctx.structure.isEmpty) {
        var annotation_1 = (0, annotation_prop_1.getMVSAnnotationForStructure)(ctx.structure, props.annotationId).annotation;
        if (annotation_1) {
            var colorForStructureElementLocation_1 = function (location) {
                var _a;
                // if (annot.getAnnotationForLocation(location)?.color !== annot.getAnnotationForLocation_Reference(location)?.color) throw new Error('AssertionError');
                return (_a = (0, utils_1.decodeColor)(annotation_1 === null || annotation_1 === void 0 ? void 0 : annotation_1.getValueForLocation(location, props.fieldName))) !== null && _a !== void 0 ? _a : props.background;
            };
            var auxLocation_1 = structure_1.StructureElement.Location.create(ctx.structure);
            color = function (location) {
                if (structure_1.StructureElement.Location.is(location)) {
                    return colorForStructureElementLocation_1(location);
                }
                else if (structure_1.Bond.isLocation(location)) {
                    // this will be applied for each bond twice, to get color of each half (a* refers to the adjacent atom, b* to the opposite atom)
                    auxLocation_1.unit = location.aUnit;
                    auxLocation_1.element = location.aUnit.elements[location.aIndex];
                    return colorForStructureElementLocation_1(auxLocation_1);
                }
                return props.background;
            };
        }
        else {
            console.error("Annotation source \"".concat(props.annotationId, "\" not present"));
        }
    }
    return {
        factory: MVSAnnotationColorTheme,
        granularity: 'group',
        preferSmoothing: true,
        color: color,
        props: props,
        description: 'Assigns colors based on custom MolViewSpec annotation data.',
    };
}
exports.MVSAnnotationColorTheme = MVSAnnotationColorTheme;
/** A thingy that is needed to register color theme "MVS Annotation" */
exports.MVSAnnotationColorThemeProvider = {
    name: 'mvs-annotation',
    label: 'MVS Annotation',
    category: color_1.ColorTheme.Category.Misc,
    factory: MVSAnnotationColorTheme,
    getParams: function (ctx) { return exports.MVSAnnotationColorThemeParams; },
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.MVSAnnotationColorThemeParams),
    isApplicable: function (ctx) { return !!ctx.structure && (0, is_mvs_model_prop_1.isMVSStructure)(ctx.structure); },
};
