// molstar/lib/commonjs/extensions/mvs/components/custom-tooltips-prop.js
"use strict";
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomTooltipsLabelProvider = exports.CustomTooltipsProvider = exports.CustomTooltipsParams = void 0;
var tslib_1 = require("tslib");
var custom_structure_property_1 = require("../../../mol-model-props/common/custom-structure-property");
var custom_property_1 = require("../../../mol-model/custom-property");
var structure_1 = require("../../../mol-model/structure");
var param_definition_1 = require("../../../mol-util/param-definition");
var utils_1 = require("../helpers/utils");
var selector_1 = require("./selector");
exports.CustomTooltipsParams = {
    tooltips: param_definition_1.ParamDefinition.ObjectList({
        text: param_definition_1.ParamDefinition.Text('', { description: 'Text of the tooltip' }),
        selector: selector_1.SelectorParams,
    }, function (obj) { return obj.text; }),
};
/** Provider for custom structure property "CustomTooltips" */
exports.CustomTooltipsProvider = custom_structure_property_1.CustomStructureProperty.createProvider({
    label: 'MVS Custom Tooltips',
    descriptor: (0, custom_property_1.CustomPropertyDescriptor)({
        name: 'mvs-custom-tooltips',
    }),
    type: 'local',
    defaultParams: exports.CustomTooltipsParams,
    getParams: function (data) { return exports.CustomTooltipsParams; },
    isApplicable: function (data) { return data.root === data; },
    obtain: function (ctx, data, props) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var fullProps, value;
        return tslib_1.__generator(this, function (_a) {
            fullProps = tslib_1.__assign(tslib_1.__assign({}, param_definition_1.ParamDefinition.getDefaultValues(exports.CustomTooltipsParams)), props);
            value = fullProps.tooltips.map(function (t) { return ({
                selector: t.selector,
                text: t.text,
            }); });
            return [2 /*return*/, { value: value }];
        });
    }); },
});
/** Label provider based on custom structure property "CustomTooltips" */
exports.CustomTooltipsLabelProvider = {
    label: function (loci) {
        var _a;
        switch (loci.kind) {
            case 'element-loci':
                if (!loci.structure.customPropertyDescriptors.hasReference(exports.CustomTooltipsProvider.descriptor))
                    return undefined;
                var location_1 = structure_1.StructureElement.Loci.getFirstLocation(loci);
                if (!location_1)
                    return undefined;
                var tooltipData = exports.CustomTooltipsProvider.get(location_1.structure).value;
                if (!tooltipData || tooltipData.length === 0)
                    return undefined;
                var texts = [];
                for (var _i = 0, tooltipData_1 = tooltipData; _i < tooltipData_1.length; _i++) {
                    var tooltip = tooltipData_1[_i];
                    var elements = (_a = tooltip.elementSet) !== null && _a !== void 0 ? _a : (tooltip.elementSet = selector_1.ElementSet.fromSelector(location_1.structure, tooltip.selector));
                    if (selector_1.ElementSet.has(elements, location_1))
                        texts.push(tooltip.text);
                }
                return (0, utils_1.filterDefined)(texts).join(' | ');
            default:
                return undefined;
        }
    }
};
