// molstar/lib/commonjs/extensions/mvs/components/formats.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MVSJFormatProvider = exports.LoadMvsData = exports.ParseMVSJ = exports.Mvs = void 0;
var tslib_1 = require("tslib");
var provider_1 = require("../../../mol-plugin-state/formats/provider");
var objects_1 = require("../../../mol-plugin-state/objects");
var data_1 = require("../../../mol-plugin-state/transforms/data");
var mol_state_1 = require("../../../mol-state");
var mol_task_1 = require("../../../mol-task");
var assets_1 = require("../../../mol-util/assets");
var param_definition_1 = require("../../../mol-util/param-definition");
var load_1 = require("../load");
var mvs_data_1 = require("../mvs-data");
var annotation_structure_component_1 = require("./annotation-structure-component");
/** Plugin state object storing `MVSData` */
var Mvs = /** @class */ (function (_super) {
    tslib_1.__extends(Mvs, _super);
    function Mvs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Mvs;
}(objects_1.PluginStateObject.Create({ name: 'MVS Data', typeClass: 'Data' })));
exports.Mvs = Mvs;
/** Transformer for parsing data in MVSJ format */
exports.ParseMVSJ = (0, annotation_structure_component_1.MVSTransform)({
    name: 'mvs-parse-mvsj',
    display: { name: 'MVS Annotation Component', description: 'A molecular structure component defined by MVS annotation data.' },
    from: objects_1.PluginStateObject.Data.String,
    to: Mvs,
})({
    apply: function (_a, plugin) {
        var a = _a.a;
        var mvsData = mvs_data_1.MVSData.fromMVSJ(a.data);
        var sourceUrl = tryGetDownloadUrl(a, plugin);
        return new Mvs({ mvsData: mvsData, sourceUrl: sourceUrl });
    },
});
/** If the PluginStateObject `pso` comes from a Download transform, try to get its `url` parameter. */
function tryGetDownloadUrl(pso, plugin) {
    var _a;
    var theCell = plugin.state.data.selectQ(function (q) { return q.ofTransformer(data_1.Download); }).find(function (cell) { return cell.obj === pso; });
    var urlParam = (_a = theCell === null || theCell === void 0 ? void 0 : theCell.transform.params) === null || _a === void 0 ? void 0 : _a.url;
    return urlParam ? assets_1.Asset.getUrl(urlParam) : undefined;
}
/** Params for the `LoadMvsData` action */
var LoadMvsDataParams = {
    replaceExisting: param_definition_1.ParamDefinition.Boolean(false, { description: 'If true, the loaded MVS view will replace the current state; if false, the MVS view will be added to the current state.' }),
};
/** State action which loads a MVS view into Mol* */
exports.LoadMvsData = mol_state_1.StateAction.build({
    display: { name: 'Load MVS Data' },
    from: Mvs,
    params: LoadMvsDataParams,
})(function (_a, plugin) {
    var a = _a.a, params = _a.params;
    return mol_task_1.Task.create('Load MVS Data', function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var _a, mvsData, sourceUrl;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = a.data, mvsData = _a.mvsData, sourceUrl = _a.sourceUrl;
                    return [4 /*yield*/, (0, load_1.loadMVS)(plugin, mvsData, { replaceExisting: params.replaceExisting, sourceUrl: sourceUrl })];
                case 1:
                    _b.sent();
                    return [2 /*return*/];
            }
        });
    }); });
});
/** Data format provider for MVSJ format.
 * If Visuals:On, it will load the parsed MVS view;
 * otherwise it will just create a plugin state object with parsed data. */
exports.MVSJFormatProvider = (0, provider_1.DataFormatProvider)({
    label: 'MVSJ',
    description: 'MVSJ',
    category: 'Miscellaneous',
    stringExtensions: ['mvsj'],
    parse: function (plugin, data) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, plugin.state.data.build().to(data).apply(exports.ParseMVSJ).commit()];
        });
    }); },
    visuals: function (plugin, data) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var ref, params;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    ref = mol_state_1.StateObjectRef.resolveRef(data);
                    params = param_definition_1.ParamDefinition.getDefaultValues(LoadMvsDataParams);
                    return [4 /*yield*/, plugin.state.data.applyAction(exports.LoadMvsData, params, ref).run()];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
});
