// molstar/lib/commonjs/extensions/mvs/components/is-mvs-model-prop.js
"use strict";
/**
 * Copyright (c) 2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMVSStructure = exports.isMVSModel = exports.IsMVSModelProvider = exports.IsMVSModelParams = void 0;
var tslib_1 = require("tslib");
var custom_model_property_1 = require("../../../mol-model-props/common/custom-model-property");
var custom_property_1 = require("../../../mol-model/custom-property");
var param_definition_1 = require("../../../mol-util/param-definition");
exports.IsMVSModelParams = {
    isMvs: param_definition_1.ParamDefinition.Boolean(false, { description: 'Flag this model as managed by MolViewSpec and enable MolViewSpec features' }),
};
/** Provider for custom model property "Is MVS" */
exports.IsMVSModelProvider = custom_model_property_1.CustomModelProperty.createProvider({
    label: 'MVS',
    descriptor: (0, custom_property_1.CustomPropertyDescriptor)({
        name: 'mvs-is-mvs-model',
    }),
    type: 'static',
    defaultParams: exports.IsMVSModelParams,
    getParams: function (data) { return exports.IsMVSModelParams; },
    isApplicable: function (data) { return true; },
    obtain: function (ctx, data, props) { return tslib_1.__awaiter(void 0, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
        return [2 /*return*/, ({ value: {} })];
    }); }); },
});
/** Decide if the model is flagged as managed by MolViewSpec */
function isMVSModel(model) {
    var _a;
    return !!((_a = exports.IsMVSModelProvider.props(model)) === null || _a === void 0 ? void 0 : _a.isMvs);
}
exports.isMVSModel = isMVSModel;
/** Decide if the structure is flagged as managed by MolViewSpec */
function isMVSStructure(structure) {
    return structure.models.some(isMVSModel);
}
exports.isMVSStructure = isMVSStructure;
