// molstar/lib/commonjs/extensions/mvs/components/multilayer-color-theme.js
"use strict";
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeMultilayerColorThemeProvider = exports.MultilayerColorThemeName = exports.DefaultMultilayerColorThemeProps = exports.makeMultilayerColorThemeParams = exports.NoColor = void 0;
var structure_1 = require("../../../mol-model/structure");
var color_1 = require("../../../mol-theme/color");
var color_2 = require("../../../mol-util/color");
var names_1 = require("../../../mol-util/color/names");
var param_definition_1 = require("../../../mol-util/param-definition");
var string_1 = require("../../../mol-util/string");
var is_mvs_model_prop_1 = require("./is-mvs-model-prop");
var selector_1 = require("./selector");
/** Special value that can be used as color with null-like semantic (i.e. "no color provided").
 * By some lucky coincidence, Mol* treats -1 as white. */
exports.NoColor = (0, color_2.Color)(-1);
/** Return true if `color` is a real color, false if it is `NoColor`. */
function isValidColor(color) {
    return color >= 0;
}
var DefaultBackgroundColor = names_1.ColorNames.white;
/** Parameter definition for color theme "Multilayer" */
function makeMultilayerColorThemeParams(colorThemeRegistry, ctx) {
    var colorThemeInfo = {
        help: function (value) {
            var name = value.name, params = value.params;
            var p = colorThemeRegistry.get(name);
            var ct = p.factory({}, params);
            return { description: ct.description, legend: ct.legend };
        }
    };
    var nestedThemeTypes = colorThemeRegistry.types.filter(function (_a) {
        var name = _a[0], label = _a[1], category = _a[2];
        return name !== exports.MultilayerColorThemeName && colorThemeRegistry.get(name).isApplicable(ctx);
    }); // Adding 'multilayer' theme itself would cause infinite recursion
    return {
        layers: param_definition_1.ParamDefinition.ObjectList({
            theme: param_definition_1.ParamDefinition.Mapped('uniform', nestedThemeTypes, function (name) { return param_definition_1.ParamDefinition.Group(colorThemeRegistry.get(name).getParams({ structure: structure_1.Structure.Empty })); }, colorThemeInfo),
            selection: selector_1.SelectorParams,
        }, function (obj) { return (0, string_1.stringToWords)(obj.theme.name); }, { description: 'A list of layers, each defining a color theme. The last listed layer is the top layer (applies first). If the top layer does not provide color for a location or its selection does not cover the location, the underneath layers will apply.' }),
        background: param_definition_1.ParamDefinition.Color(DefaultBackgroundColor, { description: 'Color for elements where no layer applies' }),
    };
}
exports.makeMultilayerColorThemeParams = makeMultilayerColorThemeParams;
/** Default values for `MultilayerColorThemeProps` */
exports.DefaultMultilayerColorThemeProps = { layers: [], background: DefaultBackgroundColor };
/** Return color theme that assigns colors based on a list of nested color themes (layers).
 * The last layer in the list whose selection covers the given location
 * and which provides a valid (non-negative) color value will be used.
 * If a nested theme provider has `ensureCustomProperties` methods, these will not be called automatically
 * (the caller must ensure that any required custom properties be attached). */
function makeMultilayerColorTheme(ctx, props, colorThemeRegistry) {
    var _a;
    var colorLayers = []; // undefined elementSet means 'all'
    for (var i = props.layers.length - 1; i >= 0; i--) { // iterate from end to get top layer first, bottom layer last
        var layer = props.layers[i];
        var themeProvider = colorThemeRegistry.get(layer.theme.name);
        if (!themeProvider) {
            console.warn("Skipping color theme '".concat(layer.theme.name, "', cannot find it in registry."));
            continue;
        }
        if ((_a = themeProvider.ensureCustomProperties) === null || _a === void 0 ? void 0 : _a.attach) {
            console.warn("Multilayer color theme: layer \"".concat(themeProvider.name, "\" has ensureCustomProperties.attach method, but Multilayer color theme does not call it. If the layer does not work, make sure you call ensureCustomProperties.attach somewhere."));
        }
        var theme = themeProvider.factory(ctx, layer.theme.params);
        switch (theme.granularity) {
            case 'uniform':
            case 'instance':
            case 'group':
            case 'groupInstance':
            case 'vertex':
            case 'vertexInstance':
                var elementSet = (0, selector_1.isSelectorAll)(layer.selection) ? undefined : selector_1.ElementSet.fromSelector(ctx.structure, layer.selection); // treating 'all' specially for performance reasons (it's expected to be used most often)
                colorLayers.push({ color: theme.color, elementSet: elementSet });
                break;
            default:
                console.warn("Skipping color theme '".concat(layer.theme.name, "', cannot process granularity '").concat(theme.granularity, "'"));
        }
    }
    ;
    function structureElementColor(loc, isSecondary) {
        for (var _i = 0, colorLayers_1 = colorLayers; _i < colorLayers_1.length; _i++) {
            var layer = colorLayers_1[_i];
            var matches = !layer.elementSet || selector_1.ElementSet.has(layer.elementSet, loc);
            if (!matches)
                continue;
            var color_3 = layer.color(loc, isSecondary);
            if (!isValidColor(color_3))
                continue;
            return color_3;
        }
        return props.background;
    }
    var auxLocation = structure_1.StructureElement.Location.create(ctx.structure);
    var color = function (location, isSecondary) {
        if (structure_1.StructureElement.Location.is(location)) {
            return structureElementColor(location, isSecondary);
        }
        else if (structure_1.Bond.isLocation(location)) {
            // this will be applied for each bond twice, to get color of each half (a* refers to the adjacent atom, b* to the opposite atom)
            auxLocation.unit = location.aUnit;
            auxLocation.element = location.aUnit.elements[location.aIndex];
            return structureElementColor(auxLocation, isSecondary);
        }
        return props.background;
    };
    return {
        factory: function (ctx_, props_) { return makeMultilayerColorTheme(ctx_, props_, colorThemeRegistry); },
        granularity: 'group',
        preferSmoothing: true,
        color: color,
        props: props,
        description: 'Combines colors from multiple color themes.',
    };
}
/** Unique name for "Multilayer" color theme */
exports.MultilayerColorThemeName = 'mvs-multilayer';
/** A thingy that is needed to register color theme "Multilayer" */
function makeMultilayerColorThemeProvider(colorThemeRegistry) {
    return {
        name: exports.MultilayerColorThemeName,
        label: 'MVS Multi-layer',
        category: color_1.ColorTheme.Category.Misc,
        factory: function (ctx, props) { return makeMultilayerColorTheme(ctx, props, colorThemeRegistry); },
        getParams: function (ctx) { return makeMultilayerColorThemeParams(colorThemeRegistry, ctx); },
        defaultValues: exports.DefaultMultilayerColorThemeProps,
        isApplicable: function (ctx) { return !!ctx.structure && (0, is_mvs_model_prop_1.isMVSStructure)(ctx.structure); },
    };
}
exports.makeMultilayerColorThemeProvider = makeMultilayerColorThemeProvider;
