// molstar/lib/commonjs/extensions/mvs/components/selector.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.substructureFromSelector = exports.ElementSet = exports.isSelectorAll = exports.SelectorAll = exports.SelectorParams = exports.StaticSelectorChoice = void 0;
var tslib_1 = require("tslib");
var int_1 = require("../../../mol-data/int");
var structure_1 = require("../../../mol-model/structure");
var structure_component_1 = require("../../../mol-plugin-state/helpers/structure-component");
var objects_1 = require("../../../mol-plugin-state/objects");
var builder_1 = require("../../../mol-script/language/builder");
var array_1 = require("../../../mol-util/array");
var object_1 = require("../../../mol-util/object");
var param_choice_1 = require("../../../mol-util/param-choice");
var param_definition_1 = require("../../../mol-util/param-definition");
var string_1 = require("../../../mol-util/string");
var annotation_structure_component_1 = require("./annotation-structure-component");
/** Allowed values for a static selector */
exports.StaticSelectorChoice = new param_choice_1.Choice((0, object_1.mapArrayToObject)(structure_component_1.StaticStructureComponentTypes, function (t) { return (0, string_1.capitalize)(t); }), 'all');
/** Parameter definition for specifying a part of structure (kinda extension of `StructureComponentParams` from mol-plugin-state/helpers/structure-component) */
exports.SelectorParams = param_definition_1.ParamDefinition.MappedStatic('static', {
    static: exports.StaticSelectorChoice.PDSelect(),
    expression: param_definition_1.ParamDefinition.Value(builder_1.MolScriptBuilder.struct.generator.all),
    bundle: param_definition_1.ParamDefinition.Value(structure_1.StructureElement.Bundle.Empty),
    script: param_definition_1.ParamDefinition.Script({ language: 'mol-script', expression: '(sel.atom.all)' }),
    annotation: param_definition_1.ParamDefinition.Group((0, object_1.pickObjectKeys)(annotation_structure_component_1.MVSAnnotationStructureComponentParams, ['annotationId', 'fieldName', 'fieldValues'])),
}, { description: 'Define a part of the structure where this layer applies (use Static:all to apply to the whole structure)' });
/** `Selector` for selecting the whole structure */
exports.SelectorAll = { name: 'static', params: 'all' };
/** Decide whether a selector is `SelectorAll` */
function isSelectorAll(props) {
    return props.name === 'static' && props.params === 'all';
}
exports.isSelectorAll = isSelectorAll;
exports.ElementSet = {
    /** Create an `ElementSet` from the substructure of `structure` defined by `selector` */
    fromSelector: function (structure, selector) {
        var _a;
        var _b;
        if (!structure)
            return {};
        var arrays = {};
        var selection = substructureFromSelector(structure, selector); // using `getAtomRangesForRow` might (might not) be faster here
        for (var _i = 0, _c = selection.units; _i < _c.length; _i++) {
            var unit = _c[_i];
            (0, array_1.arrayExtend)((_a = arrays[_b = unit.model.id]) !== null && _a !== void 0 ? _a : (arrays[_b] = []), unit.elements);
        }
        var result = {};
        for (var modelId in arrays) {
            var array = arrays[modelId];
            (0, array_1.sortIfNeeded)(array, function (a, b) { return a - b; });
            result[modelId] = int_1.SortedArray.ofSortedArray(array);
        }
        return result;
    },
    /** Decide if the element set `set` contains structure element location `location` */
    has: function (set, location) {
        var array = set[location.unit.model.id];
        return array ? int_1.SortedArray.has(array, location.element) : false;
    },
};
/** Return a substructure of `structure` defined by `selector` */
function substructureFromSelector(structure, selector) {
    var pso = (selector.name === 'annotation') ?
        (0, annotation_structure_component_1.createMVSAnnotationStructureComponent)(structure, tslib_1.__assign(tslib_1.__assign({}, selector.params), { label: '', nullIfEmpty: false }))
        : (0, structure_component_1.createStructureComponent)(structure, { type: selector, label: '', nullIfEmpty: false }, { source: structure });
    return objects_1.PluginStateObject.Molecule.Structure.is(pso) ? pso.data : structure_1.Structure.Empty;
}
exports.substructureFromSelector = substructureFromSelector;
