// molstar/lib/extensions/mvs/components/annotation-structure-component.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { Structure, StructureSelection } from '../../../mol-model/structure';
import { StructureQueryHelper } from '../../../mol-plugin-state/helpers/structure-query';
import { PluginStateObject as SO } from '../../../mol-plugin-state/objects';
import { StateObject, StateTransformer } from '../../../mol-state';
import { deepEqual } from '../../../mol-util';
import { omitObjectKeys } from '../../../mol-util/object';
import { ParamDefinition as PD } from '../../../mol-util/param-definition';
import { rowsToExpression } from '../helpers/selections';
import { getMVSAnnotationForStructure } from './annotation-prop';
/** Parameter definition for `MVSAnnotationStructureComponent` transformer */
export var MVSAnnotationStructureComponentParams = {
    annotationId: PD.Text('', { description: 'Reference to "Annotation" custom model property' }),
    fieldName: PD.Text('component', { description: 'Annotation field (column) from which to take component identifier' }),
    fieldValues: PD.MappedStatic('all', {
        all: PD.EmptyGroup(),
        selected: PD.ObjectList({
            value: PD.Text(),
        }, function (obj) { return obj.value; }),
    }),
    nullIfEmpty: PD.Optional(PD.Boolean(true, { isHidden: false })),
    label: PD.Text('', { isHidden: false }),
};
/** Transformer builder for MVS extension */
export var MVSTransform = StateTransformer.builderFactory('mvs');
export var MVSAnnotationStructureComponent = MVSTransform({
    name: 'mvs-structure-component-from-annotation',
    display: { name: 'MVS Annotation Component', description: 'A molecular structure component defined by MVS annotation data.' },
    from: SO.Molecule.Structure,
    to: SO.Molecule.Structure,
    params: MVSAnnotationStructureComponentParams,
})({
    apply: function (_a) {
        var a = _a.a, params = _a.params;
        return createMVSAnnotationStructureComponent(a.data, params);
    },
    update: function (_a) {
        var a = _a.a, b = _a.b, oldParams = _a.oldParams, newParams = _a.newParams;
        return updateMVSAnnotationStructureComponent(a.data, b, oldParams, newParams);
    },
    dispose: function (_a) {
        var b = _a.b;
        b === null || b === void 0 ? void 0 : b.data.customPropertyDescriptors.dispose();
    }
});
/** Create a substructure based on `MVSAnnotationStructureComponentProps` */
export function createMVSAnnotationSubstructure(structure, params) {
    var annotation = getMVSAnnotationForStructure(structure, params.annotationId).annotation;
    if (annotation) {
        var rows = annotation.getRows();
        if (params.fieldValues.name === 'selected') {
            var selectedValues_1 = new Set(params.fieldValues.params.map(function (obj) { return obj.value; }));
            rows = rows.filter(function (row, i) { return selectedValues_1.has(annotation.getValueForRow(i, params.fieldName)); });
        }
        var expression = rowsToExpression(rows);
        var selection = StructureQueryHelper.createAndRun(structure, expression).selection;
        return StructureSelection.unionStructure(selection);
    }
    else {
        return Structure.Empty;
    }
}
/** Create a substructure PSO based on `MVSAnnotationStructureComponentProps` */
export function createMVSAnnotationStructureComponent(structure, params) {
    var component = createMVSAnnotationSubstructure(structure, params);
    if (params.nullIfEmpty && component.elementCount === 0)
        return StateObject.Null;
    var label = params.label;
    if (label === undefined || label === '') {
        if (params.fieldValues.name === 'selected' && params.fieldValues.params.length > 0) {
            var values = params.fieldValues.params;
            var valuesStr = "\"".concat(values[0].value, "\"");
            if (values.length === 2) {
                valuesStr += " + \"".concat(values[1].value, "\"");
            }
            else if (values.length > 2) {
                valuesStr += " + ".concat(values.length - 1, " more values");
            }
            label = "MVS Annotation Component (".concat(params.fieldName, ": ").concat(valuesStr, ")");
        }
        else {
            label = 'MVS Annotation Component';
        }
    }
    var props = { label: label, description: Structure.elementDescription(component) };
    return new SO.Molecule.Structure(component, props);
}
/** Update a substructure PSO based on `MVSAnnotationStructureComponentProps` */
export function updateMVSAnnotationStructureComponent(a, b, oldParams, newParams) {
    var change = !deepEqual(newParams, oldParams);
    var needsRecreate = !deepEqual(omitObjectKeys(newParams, ['label']), omitObjectKeys(oldParams, ['label']));
    if (!change) {
        return StateTransformer.UpdateResult.Unchanged;
    }
    if (!needsRecreate) {
        b.label = newParams.label || b.label;
        return StateTransformer.UpdateResult.Updated;
    }
    return StateTransformer.UpdateResult.Recreate;
}
