// molstar/lib/extensions/mvs/components/annotation-prop.js
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __assign, __awaiter, __generator } from "tslib";
import { Column } from '../../../mol-data/db';
import { CIF, CifBlock, CifFile } from '../../../mol-io/reader/cif';
import { toTable } from '../../../mol-io/reader/cif/schema';
import { MmcifFormat } from '../../../mol-model-formats/structure/mmcif';
import { CustomModelProperty } from '../../../mol-model-props/common/custom-model-property';
import { CustomPropertyDescriptor } from '../../../mol-model/custom-property';
import { arrayExtend } from '../../../mol-util/array';
import { Asset } from '../../../mol-util/assets';
import { canonicalJsonString } from '../../../mol-util/json';
import { pickObjectKeys, promiseAllObj } from '../../../mol-util/object';
import { Choice } from '../../../mol-util/param-choice';
import { ParamDefinition as PD } from '../../../mol-util/param-definition';
import { AtomRanges } from '../helpers/atom-ranges';
import { IndicesAndSortings } from '../helpers/indexing';
import { MaybeStringParamDefinition } from '../helpers/param-definition';
import { MVSAnnotationSchema, getCifAnnotationSchema } from '../helpers/schemas';
import { atomQualifies, getAtomRangesForRow } from '../helpers/selections';
import { safePromise } from '../helpers/utils';
/** Allowed values for the annotation format parameter */
var MVSAnnotationFormat = new Choice({ json: 'json', cif: 'cif', bcif: 'bcif' }, 'json');
var MVSAnnotationFormatTypes = { json: 'string', cif: 'string', bcif: 'binary' };
export var MVSAnnotationsParams = {
    annotations: PD.ObjectList({
        source: PD.MappedStatic('source-cif', {
            'source-cif': PD.EmptyGroup(),
            'url': PD.Group({
                url: PD.Text(''),
                format: MVSAnnotationFormat.PDSelect(),
            }),
        }),
        schema: MVSAnnotationSchema.PDSelect(),
        cifBlock: PD.MappedStatic('index', {
            index: PD.Group({ index: PD.Numeric(0, { min: 0, step: 1 }, { description: '0-based index of the block' }) }),
            header: PD.Group({ header: PD.Text(undefined, { description: 'Block header' }) }),
        }, { description: 'Specify which CIF block contains annotation data (only relevant when format=cif or format=bcif)' }),
        cifCategory: MaybeStringParamDefinition(undefined, { description: 'Specify which CIF category contains annotation data (only relevant when format=cif or format=bcif)' }),
        id: PD.Text('', { description: 'Arbitrary identifier that can be referenced by MVSAnnotationColorTheme' }),
    }, function (obj) { return obj.id; }),
};
/** Provider for custom model property "Annotations" */
export var MVSAnnotationsProvider = CustomModelProperty.createProvider({
    label: 'MVS Annotations',
    descriptor: CustomPropertyDescriptor({
        name: 'mvs-annotations',
    }),
    type: 'static',
    defaultParams: MVSAnnotationsParams,
    getParams: function (data) { return MVSAnnotationsParams; },
    isApplicable: function (data) { return true; },
    obtain: function (ctx, data, props) { return __awaiter(void 0, void 0, void 0, function () {
        var specs, annots;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    props = __assign(__assign({}, PD.getDefaultValues(MVSAnnotationsParams)), props);
                    specs = (_a = props.annotations) !== null && _a !== void 0 ? _a : [];
                    return [4 /*yield*/, MVSAnnotations.fromSpecs(ctx, specs, data)];
                case 1:
                    annots = _b.sent();
                    return [2 /*return*/, { value: annots }];
            }
        });
    }); }
});
/** Represents multiple annotations retrievable by their ID */
var MVSAnnotations = /** @class */ (function () {
    function MVSAnnotations(dict) {
        this.dict = dict;
    }
    MVSAnnotations.fromSpecs = function (ctx, specs, model) {
        return __awaiter(this, void 0, void 0, function () {
            var sources, files, annots, i, spec, file, _a, _b, err_1;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        sources = specs.map(annotationSourceFromSpec);
                        return [4 /*yield*/, getFilesFromSources(ctx, sources, model)];
                    case 1:
                        files = _c.sent();
                        annots = {};
                        i = 0;
                        _c.label = 2;
                    case 2:
                        if (!(i < specs.length)) return [3 /*break*/, 7];
                        spec = specs[i];
                        _c.label = 3;
                    case 3:
                        _c.trys.push([3, 5, , 6]);
                        file = files[i];
                        if (!file.ok)
                            throw file.error;
                        _a = annots;
                        _b = spec.id;
                        return [4 /*yield*/, MVSAnnotation.fromSpec(ctx, spec, file.value)];
                    case 4:
                        _a[_b] = _c.sent();
                        return [3 /*break*/, 6];
                    case 5:
                        err_1 = _c.sent();
                        console.error("Failed to obtain annotation (".concat(err_1, ").\nAnnotation specification:"), spec);
                        annots[spec.id] = MVSAnnotation.createEmpty(spec.schema);
                        return [3 /*break*/, 6];
                    case 6:
                        i++;
                        return [3 /*break*/, 2];
                    case 7: return [2 /*return*/, new MVSAnnotations(annots)];
                }
            });
        });
    };
    MVSAnnotations.prototype.getAnnotation = function (id) {
        return this.dict[id];
    };
    MVSAnnotations.prototype.getAllAnnotations = function () {
        return Object.values(this.dict);
    };
    return MVSAnnotations;
}());
export { MVSAnnotations };
/** Retrieve annotation with given `annotationId` from custom model property "MVS Annotations" and the model from which it comes */
export function getMVSAnnotationForStructure(structure, annotationId) {
    var models = structure.isEmpty ? [] : structure.models;
    for (var _i = 0, models_1 = models; _i < models_1.length; _i++) {
        var model = models_1[_i];
        if (model.customProperties.has(MVSAnnotationsProvider.descriptor)) {
            var annots = MVSAnnotationsProvider.get(model).value;
            var annotation = annots === null || annots === void 0 ? void 0 : annots.getAnnotation(annotationId);
            if (annotation) {
                return { annotation: annotation, model: model };
            }
        }
    }
    return { annotation: undefined, model: undefined };
}
/** Main class for processing MVS annotation */
var MVSAnnotation = /** @class */ (function () {
    function MVSAnnotation(data, schema) {
        this.data = data;
        this.schema = schema;
        /** Store mapping `ElementIndex` -> annotation row index for each `Model`, -1 means no row applies */
        this.indexedModels = new Map();
        this.rows = undefined;
    }
    /** Create a new `MVSAnnotation` based on specification `spec`. Use `file` if provided, otherwise download the file.
     * Throw error if download fails or problem with data. */
    MVSAnnotation.fromSpec = function (ctx, spec, file) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var _b, data, blockSpec_1, block, foundBlock, categoryName, category;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!(file !== null && file !== void 0)) return [3 /*break*/, 1];
                        _b = file;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, getFileFromSource(ctx, annotationSourceFromSpec(spec))];
                    case 2:
                        _b = (file = _c.sent());
                        _c.label = 3;
                    case 3:
                        _b;
                        switch (file.format) {
                            case 'json':
                                data = file;
                                break;
                            case 'cif':
                                if (file.data.blocks.length === 0)
                                    throw new Error('No block in CIF');
                                blockSpec_1 = spec.cifBlock;
                                block = void 0;
                                switch (blockSpec_1.name) {
                                    case 'header':
                                        foundBlock = file.data.blocks.find(function (b) { return b.header === blockSpec_1.params.header; });
                                        if (!foundBlock)
                                            throw new Error("CIF block with header ".concat(blockSpec_1.params.header, " not found"));
                                        block = foundBlock;
                                        break;
                                    case 'index':
                                        block = file.data.blocks[blockSpec_1.params.index];
                                        if (!block)
                                            throw new Error("CIF block with index ".concat(blockSpec_1.params.index, " not found"));
                                        break;
                                }
                                categoryName = (_a = spec.cifCategory) !== null && _a !== void 0 ? _a : Object.keys(block.categories)[0];
                                if (!categoryName)
                                    throw new Error('There are no categories in CIF block');
                                category = block.categories[categoryName];
                                if (!category)
                                    throw new Error("CIF category ".concat(categoryName, " not found"));
                                data = { format: 'cif', data: category };
                                break;
                        }
                        return [2 /*return*/, new MVSAnnotation(data, spec.schema)];
                }
            });
        });
    };
    MVSAnnotation.createEmpty = function (schema) {
        return new MVSAnnotation({ format: 'json', data: [] }, schema);
    };
    /** Reference implementation of `getAnnotationForLocation`, just for checking, DO NOT USE DIRECTLY */
    MVSAnnotation.prototype.getAnnotationForLocation_Reference = function (loc) {
        var model = loc.unit.model;
        var iAtom = loc.element;
        var result = undefined;
        for (var _i = 0, _a = this.getRows(); _i < _a.length; _i++) {
            var row = _a[_i];
            if (atomQualifies(model, iAtom, row))
                result = row;
        }
        return result;
    };
    /** Return value of field `fieldName` assigned to location `loc`, if any */
    MVSAnnotation.prototype.getValueForLocation = function (loc, fieldName) {
        var indexedModel = this.getIndexedModel(loc.unit.model);
        var iRow = indexedModel[loc.element];
        return this.getValueForRow(iRow, fieldName);
    };
    /** Return value of field `fieldName` assigned to `i`-th annotation row, if any */
    MVSAnnotation.prototype.getValueForRow = function (i, fieldName) {
        if (i < 0)
            return undefined;
        switch (this.data.format) {
            case 'json':
                var value = getValueFromJson(i, fieldName, this.data.data);
                if (value === undefined || typeof value === 'string')
                    return value;
                else
                    return "".concat(value);
            case 'cif':
                return getValueFromCif(i, fieldName, this.data.data);
        }
    };
    /** Return cached `ElementIndex` -> `MVSAnnotationRow` mapping for `Model` (or create it if not cached yet) */
    MVSAnnotation.prototype.getIndexedModel = function (model) {
        var key = model.id;
        if (!this.indexedModels.has(key)) {
            var result = this.getRowForEachAtom(model);
            this.indexedModels.set(key, result);
        }
        return this.indexedModels.get(key);
    };
    /** Create `ElementIndex` -> `MVSAnnotationRow` mapping for `Model` */
    MVSAnnotation.prototype.getRowForEachAtom = function (model) {
        var indices = IndicesAndSortings.get(model);
        var nAtoms = model.atomicHierarchy.atoms._rowCount;
        var result = Array(nAtoms).fill(-1);
        var rows = this.getRows();
        var _loop_1 = function (i, nRows) {
            var atomRanges = getAtomRangesForRow(model, rows[i], indices);
            AtomRanges.foreach(atomRanges, function (from, to) { return result.fill(i, from, to); });
        };
        for (var i = 0, nRows = rows.length; i < nRows; i++) {
            _loop_1(i, nRows);
        }
        return result;
    };
    /** Parse and return all annotation rows in this annotation */
    MVSAnnotation.prototype._getRows = function () {
        switch (this.data.format) {
            case 'json':
                return getRowsFromJson(this.data.data, this.schema);
            case 'cif':
                return getRowsFromCif(this.data.data, this.schema);
        }
    };
    /** Parse and return all annotation rows in this annotation, or return cached result if available */
    MVSAnnotation.prototype.getRows = function () {
        var _a;
        return (_a = this.rows) !== null && _a !== void 0 ? _a : (this.rows = this._getRows());
    };
    return MVSAnnotation;
}());
export { MVSAnnotation };
function getValueFromJson(rowIndex, fieldName, data) {
    var _a, _b;
    var js = data;
    if (Array.isArray(js)) {
        var row = (_a = js[rowIndex]) !== null && _a !== void 0 ? _a : {};
        return row[fieldName];
    }
    else {
        var column = (_b = js[fieldName]) !== null && _b !== void 0 ? _b : [];
        return column[rowIndex];
    }
}
function getValueFromCif(rowIndex, fieldName, data) {
    var column = data.getField(fieldName);
    if (!column)
        return undefined;
    if (column.valueKind(rowIndex) !== Column.ValueKind.Present)
        return undefined;
    return column.str(rowIndex);
}
function getRowsFromJson(data, schema) {
    var js = data;
    var cifSchema = getCifAnnotationSchema(schema);
    if (Array.isArray(js)) {
        // array of objects
        return js.map(function (row) { return pickObjectKeys(row, Object.keys(cifSchema)); });
    }
    else {
        // object of arrays
        var rows = [];
        var keys = Object.keys(js).filter(function (key) { return Object.hasOwn(cifSchema, key); });
        if (keys.length > 0) {
            var n_1 = js[keys[0]].length;
            if (keys.some(function (key) { return js[key].length !== n_1; }))
                throw new Error('FormatError: arrays must have the same length.');
            for (var i = 0; i < n_1; i++) {
                var item = {};
                for (var _i = 0, keys_1 = keys; _i < keys_1.length; _i++) {
                    var key = keys_1[_i];
                    item[key] = js[key][i];
                }
                rows.push(item);
            }
        }
        return rows;
    }
}
function getRowsFromCif(data, schema) {
    var rows = [];
    var cifSchema = getCifAnnotationSchema(schema);
    var table = toTable(cifSchema, data);
    arrayExtend(rows, getRowsFromTable(table)); // Avoiding Table.getRows(table) as it replaces . and ? fields by 0 or ''
    return rows;
}
/** Same as `Table.getRows` but omits `.` and `?` fields (instead of using type defaults) */
function getRowsFromTable(table) {
    var rows = [];
    var columns = table._columns;
    var nRows = table._rowCount;
    var Present = Column.ValueKind.Present;
    for (var iRow = 0; iRow < nRows; iRow++) {
        var row = {};
        for (var _i = 0, columns_1 = columns; _i < columns_1.length; _i++) {
            var col = columns_1[_i];
            if (table[col].valueKind(iRow) === Present) {
                row[col] = table[col].value(iRow);
            }
        }
        rows[iRow] = row;
    }
    return rows;
}
function getFileFromSource(ctx, source, model) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, url, dataType, dataWrapper, rawData, _b, json, parsed;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _a = source.kind;
                    switch (_a) {
                        case 'source-cif': return [3 /*break*/, 1];
                        case 'url': return [3 /*break*/, 2];
                    }
                    return [3 /*break*/, 7];
                case 1: return [2 /*return*/, { format: 'cif', data: getSourceFileFromModel(model) }];
                case 2:
                    url = Asset.getUrlAsset(ctx.assetManager, source.url);
                    dataType = MVSAnnotationFormatTypes[source.format];
                    return [4 /*yield*/, ctx.assetManager.resolve(url, dataType).runInContext(ctx.runtime)];
                case 3:
                    dataWrapper = _c.sent();
                    rawData = dataWrapper.data;
                    if (!rawData)
                        throw new Error('Missing data');
                    _b = source.format;
                    switch (_b) {
                        case 'json': return [3 /*break*/, 4];
                        case 'cif': return [3 /*break*/, 5];
                        case 'bcif': return [3 /*break*/, 5];
                    }
                    return [3 /*break*/, 7];
                case 4:
                    json = JSON.parse(rawData);
                    return [2 /*return*/, { format: 'json', data: json }];
                case 5: return [4 /*yield*/, CIF.parse(rawData).run()];
                case 6:
                    parsed = _c.sent();
                    if (parsed.isError)
                        throw new Error("Failed to parse ".concat(source.format));
                    return [2 /*return*/, { format: 'cif', data: parsed.result }];
                case 7: return [2 /*return*/];
            }
        });
    });
}
/** Like `sources.map(s => safePromise(getFileFromSource(ctx, s)))`
 * but downloads a repeating source only once. */
function getFilesFromSources(ctx, sources, model) {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var promises, _i, sources_1, src, key, files;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    promises = {};
                    for (_i = 0, sources_1 = sources; _i < sources_1.length; _i++) {
                        src = sources_1[_i];
                        key = canonicalJsonString(src);
                        (_a = promises[key]) !== null && _a !== void 0 ? _a : (promises[key] = safePromise(getFileFromSource(ctx, src, model)));
                    }
                    return [4 /*yield*/, promiseAllObj(promises)];
                case 1:
                    files = _b.sent();
                    return [2 /*return*/, sources.map(function (src) { return files[canonicalJsonString(src)]; })];
            }
        });
    });
}
function getSourceFileFromModel(model) {
    if (model && MmcifFormat.is(model.sourceData)) {
        if (model.sourceData.data.file) {
            return model.sourceData.data.file;
        }
        else {
            var frame = model.sourceData.data.frame;
            var block = CifBlock(Array.from(frame.categoryNames), frame.categories, frame.header);
            var file = CifFile([block]);
            return file;
        }
    }
    else {
        console.warn('Could not get CifFile from Model, returning empty CifFile');
        return CifFile([]);
    }
}
function annotationSourceFromSpec(s) {
    switch (s.source.name) {
        case 'url':
            return __assign({ kind: 'url' }, s.source.params);
        case 'source-cif':
            return { kind: 'source-cif' };
    }
}
