// molstar/lib/extensions/mvs/components/annotation-tooltips-prop.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __assign, __awaiter, __generator } from "tslib";
import { CustomStructureProperty } from '../../../mol-model-props/common/custom-structure-property';
import { CustomPropertyDescriptor } from '../../../mol-model/custom-property';
import { StructureElement } from '../../../mol-model/structure';
import { ParamDefinition as PD } from '../../../mol-util/param-definition';
import { filterDefined } from '../helpers/utils';
import { MVSAnnotationsProvider } from './annotation-prop';
/** Parameter definition for custom structure property "MVSAnnotationTooltips" */
export var MVSAnnotationTooltipsParams = {
    tooltips: PD.ObjectList({
        annotationId: PD.Text('', { description: 'Reference to "MVS Annotation" custom model property' }),
        fieldName: PD.Text('tooltip', { description: 'Annotation field (column) from which to take color values' }),
    }, function (obj) { return "".concat(obj.annotationId, ":").concat(obj.fieldName); }),
};
/** Provider for custom structure property "MVSAnnotationTooltips" */
export var MVSAnnotationTooltipsProvider = CustomStructureProperty.createProvider({
    label: 'MVS Annotation Tooltips',
    descriptor: CustomPropertyDescriptor({
        name: 'mvs-annotation-tooltips',
    }),
    type: 'local',
    defaultParams: MVSAnnotationTooltipsParams,
    getParams: function (data) { return MVSAnnotationTooltipsParams; },
    isApplicable: function (data) { return data.root === data; },
    obtain: function (ctx, data, props) { return __awaiter(void 0, void 0, void 0, function () {
        var fullProps;
        return __generator(this, function (_a) {
            fullProps = __assign(__assign({}, PD.getDefaultValues(MVSAnnotationTooltipsParams)), props);
            return [2 /*return*/, { value: fullProps }];
        });
    }); },
});
/** Label provider based on data from "MVS Annotation" custom model property */
export var MVSAnnotationTooltipsLabelProvider = {
    label: function (loci) {
        switch (loci.kind) {
            case 'element-loci':
                if (!loci.structure.customPropertyDescriptors.hasReference(MVSAnnotationTooltipsProvider.descriptor))
                    return undefined;
                var location_1 = StructureElement.Loci.getFirstLocation(loci);
                if (!location_1)
                    return undefined;
                var tooltipProps = MVSAnnotationTooltipsProvider.get(location_1.structure).value;
                if (!tooltipProps || tooltipProps.tooltips.length === 0)
                    return undefined;
                var annotations_1 = MVSAnnotationsProvider.get(location_1.unit.model).value;
                var texts = tooltipProps.tooltips.map(function (p) { var _a; return (_a = annotations_1 === null || annotations_1 === void 0 ? void 0 : annotations_1.getAnnotation(p.annotationId)) === null || _a === void 0 ? void 0 : _a.getValueForLocation(location_1, p.fieldName); });
                return filterDefined(texts).join(' | ');
            default:
                return undefined;
        }
    }
};
