// molstar/lib/extensions/mvs/components/annotation-color-theme.js
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { Bond, StructureElement } from '../../../mol-model/structure';
import { ColorTheme } from '../../../mol-theme/color';
import { ColorNames } from '../../../mol-util/color/names';
import { ParamDefinition as PD } from '../../../mol-util/param-definition';
import { decodeColor } from '../helpers/utils';
import { getMVSAnnotationForStructure } from './annotation-prop';
import { isMVSStructure } from './is-mvs-model-prop';
/** Parameter definition for color theme "MVS Annotation" */
export var MVSAnnotationColorThemeParams = {
    annotationId: PD.Text('', { description: 'Reference to "Annotation" custom model property' }),
    fieldName: PD.Text('color', { description: 'Annotation field (column) from which to take color values' }),
    background: PD.Color(ColorNames.gainsboro, { description: 'Color for elements without annotation' }),
};
/** Return color theme that assigns colors based on an annotation file.
 * The annotation file itself is handled by a custom model property (`MVSAnnotationsProvider`),
 * the color theme then just uses this property. */
export function MVSAnnotationColorTheme(ctx, props) {
    var color = function () { return props.background; };
    if (ctx.structure && !ctx.structure.isEmpty) {
        var annotation_1 = getMVSAnnotationForStructure(ctx.structure, props.annotationId).annotation;
        if (annotation_1) {
            var colorForStructureElementLocation_1 = function (location) {
                var _a;
                // if (annot.getAnnotationForLocation(location)?.color !== annot.getAnnotationForLocation_Reference(location)?.color) throw new Error('AssertionError');
                return (_a = decodeColor(annotation_1 === null || annotation_1 === void 0 ? void 0 : annotation_1.getValueForLocation(location, props.fieldName))) !== null && _a !== void 0 ? _a : props.background;
            };
            var auxLocation_1 = StructureElement.Location.create(ctx.structure);
            color = function (location) {
                if (StructureElement.Location.is(location)) {
                    return colorForStructureElementLocation_1(location);
                }
                else if (Bond.isLocation(location)) {
                    // this will be applied for each bond twice, to get color of each half (a* refers to the adjacent atom, b* to the opposite atom)
                    auxLocation_1.unit = location.aUnit;
                    auxLocation_1.element = location.aUnit.elements[location.aIndex];
                    return colorForStructureElementLocation_1(auxLocation_1);
                }
                return props.background;
            };
        }
        else {
            console.error("Annotation source \"".concat(props.annotationId, "\" not present"));
        }
    }
    return {
        factory: MVSAnnotationColorTheme,
        granularity: 'group',
        preferSmoothing: true,
        color: color,
        props: props,
        description: 'Assigns colors based on custom MolViewSpec annotation data.',
    };
}
/** A thingy that is needed to register color theme "MVS Annotation" */
export var MVSAnnotationColorThemeProvider = {
    name: 'mvs-annotation',
    label: 'MVS Annotation',
    category: ColorTheme.Category.Misc,
    factory: MVSAnnotationColorTheme,
    getParams: function (ctx) { return MVSAnnotationColorThemeParams; },
    defaultValues: PD.getDefaultValues(MVSAnnotationColorThemeParams),
    isApplicable: function (ctx) { return !!ctx.structure && isMVSStructure(ctx.structure); },
};
