// molstar/lib/extensions/mvs/components/custom-tooltips-prop.js
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __assign, __awaiter, __generator } from "tslib";
import { CustomStructureProperty } from '../../../mol-model-props/common/custom-structure-property';
import { CustomPropertyDescriptor } from '../../../mol-model/custom-property';
import { StructureElement } from '../../../mol-model/structure';
import { ParamDefinition as PD } from '../../../mol-util/param-definition';
import { filterDefined } from '../helpers/utils';
import { ElementSet, SelectorParams } from './selector';
export var CustomTooltipsParams = {
    tooltips: PD.ObjectList({
        text: PD.Text('', { description: 'Text of the tooltip' }),
        selector: SelectorParams,
    }, function (obj) { return obj.text; }),
};
/** Provider for custom structure property "CustomTooltips" */
export var CustomTooltipsProvider = CustomStructureProperty.createProvider({
    label: 'MVS Custom Tooltips',
    descriptor: CustomPropertyDescriptor({
        name: 'mvs-custom-tooltips',
    }),
    type: 'local',
    defaultParams: CustomTooltipsParams,
    getParams: function (data) { return CustomTooltipsParams; },
    isApplicable: function (data) { return data.root === data; },
    obtain: function (ctx, data, props) { return __awaiter(void 0, void 0, void 0, function () {
        var fullProps, value;
        return __generator(this, function (_a) {
            fullProps = __assign(__assign({}, PD.getDefaultValues(CustomTooltipsParams)), props);
            value = fullProps.tooltips.map(function (t) { return ({
                selector: t.selector,
                text: t.text,
            }); });
            return [2 /*return*/, { value: value }];
        });
    }); },
});
/** Label provider based on custom structure property "CustomTooltips" */
export var CustomTooltipsLabelProvider = {
    label: function (loci) {
        var _a;
        switch (loci.kind) {
            case 'element-loci':
                if (!loci.structure.customPropertyDescriptors.hasReference(CustomTooltipsProvider.descriptor))
                    return undefined;
                var location_1 = StructureElement.Loci.getFirstLocation(loci);
                if (!location_1)
                    return undefined;
                var tooltipData = CustomTooltipsProvider.get(location_1.structure).value;
                if (!tooltipData || tooltipData.length === 0)
                    return undefined;
                var texts = [];
                for (var _i = 0, tooltipData_1 = tooltipData; _i < tooltipData_1.length; _i++) {
                    var tooltip = tooltipData_1[_i];
                    var elements = (_a = tooltip.elementSet) !== null && _a !== void 0 ? _a : (tooltip.elementSet = ElementSet.fromSelector(location_1.structure, tooltip.selector));
                    if (ElementSet.has(elements, location_1))
                        texts.push(tooltip.text);
                }
                return filterDefined(texts).join(' | ');
            default:
                return undefined;
        }
    }
};
