// molstar/lib/extensions/mvs/components/formats.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __awaiter, __extends, __generator } from "tslib";
import { DataFormatProvider } from '../../../mol-plugin-state/formats/provider';
import { PluginStateObject as SO } from '../../../mol-plugin-state/objects';
import { Download } from '../../../mol-plugin-state/transforms/data';
import { StateAction, StateObjectRef } from '../../../mol-state';
import { Task } from '../../../mol-task';
import { Asset } from '../../../mol-util/assets';
import { ParamDefinition as PD } from '../../../mol-util/param-definition';
import { loadMVS } from '../load';
import { MVSData } from '../mvs-data';
import { MVSTransform } from './annotation-structure-component';
/** Plugin state object storing `MVSData` */
var Mvs = /** @class */ (function (_super) {
    __extends(Mvs, _super);
    function Mvs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Mvs;
}(SO.Create({ name: 'MVS Data', typeClass: 'Data' })));
export { Mvs };
/** Transformer for parsing data in MVSJ format */
export var ParseMVSJ = MVSTransform({
    name: 'mvs-parse-mvsj',
    display: { name: 'MVS Annotation Component', description: 'A molecular structure component defined by MVS annotation data.' },
    from: SO.Data.String,
    to: Mvs,
})({
    apply: function (_a, plugin) {
        var a = _a.a;
        var mvsData = MVSData.fromMVSJ(a.data);
        var sourceUrl = tryGetDownloadUrl(a, plugin);
        return new Mvs({ mvsData: mvsData, sourceUrl: sourceUrl });
    },
});
/** If the PluginStateObject `pso` comes from a Download transform, try to get its `url` parameter. */
function tryGetDownloadUrl(pso, plugin) {
    var _a;
    var theCell = plugin.state.data.selectQ(function (q) { return q.ofTransformer(Download); }).find(function (cell) { return cell.obj === pso; });
    var urlParam = (_a = theCell === null || theCell === void 0 ? void 0 : theCell.transform.params) === null || _a === void 0 ? void 0 : _a.url;
    return urlParam ? Asset.getUrl(urlParam) : undefined;
}
/** Params for the `LoadMvsData` action */
var LoadMvsDataParams = {
    replaceExisting: PD.Boolean(false, { description: 'If true, the loaded MVS view will replace the current state; if false, the MVS view will be added to the current state.' }),
};
/** State action which loads a MVS view into Mol* */
export var LoadMvsData = StateAction.build({
    display: { name: 'Load MVS Data' },
    from: Mvs,
    params: LoadMvsDataParams,
})(function (_a, plugin) {
    var a = _a.a, params = _a.params;
    return Task.create('Load MVS Data', function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, mvsData, sourceUrl;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = a.data, mvsData = _a.mvsData, sourceUrl = _a.sourceUrl;
                    return [4 /*yield*/, loadMVS(plugin, mvsData, { replaceExisting: params.replaceExisting, sourceUrl: sourceUrl })];
                case 1:
                    _b.sent();
                    return [2 /*return*/];
            }
        });
    }); });
});
/** Data format provider for MVSJ format.
 * If Visuals:On, it will load the parsed MVS view;
 * otherwise it will just create a plugin state object with parsed data. */
export var MVSJFormatProvider = DataFormatProvider({
    label: 'MVSJ',
    description: 'MVSJ',
    category: 'Miscellaneous',
    stringExtensions: ['mvsj'],
    parse: function (plugin, data) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, plugin.state.data.build().to(data).apply(ParseMVSJ).commit()];
        });
    }); },
    visuals: function (plugin, data) { return __awaiter(void 0, void 0, void 0, function () {
        var ref, params;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    ref = StateObjectRef.resolveRef(data);
                    params = PD.getDefaultValues(LoadMvsDataParams);
                    return [4 /*yield*/, plugin.state.data.applyAction(LoadMvsData, params, ref).run()];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
});
