// molstar/lib/extensions/mvs/helpers/indexing.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { Column } from '../../../mol-data/db';
import { SortedArray } from '../../../mol-data/int';
import { filterInPlace, range, sortIfNeeded } from '../../../mol-util/array';
import { MultiMap, NumberMap } from './utils';
export var IndicesAndSortings = {
    /** Get `IndicesAndSortings` for a model (use a cached value or create if not available yet) */
    get: function (model) {
        var _a;
        var _b;
        return (_a = (_b = model._dynamicPropertyData)['indices-and-sortings']) !== null && _a !== void 0 ? _a : (_b['indices-and-sortings'] = IndicesAndSortings.create(model));
    },
    /** Create `IndicesAndSortings` for a model */
    create: function (model) {
        var h = model.atomicHierarchy;
        var nAtoms = h.atoms._rowCount;
        var nChains = h.chains._rowCount;
        var _a = h.chains, label_entity_id = _a.label_entity_id, label_asym_id = _a.label_asym_id, auth_asym_id = _a.auth_asym_id;
        var _b = h.residues, label_seq_id = _b.label_seq_id, auth_seq_id = _b.auth_seq_id, pdbx_PDB_ins_code = _b.pdbx_PDB_ins_code;
        var Present = Column.ValueKind.Present;
        var chainsByLabelEntityId = new MultiMap();
        var chainsByLabelAsymId = new MultiMap();
        var chainsByAuthAsymId = new MultiMap();
        var residuesSortedByLabelSeqId = new Map();
        var residuesSortedByAuthSeqId = new Map();
        var residuesByInsCode = new Map();
        var atomsById = new NumberMap(nAtoms + 1);
        var atomsByIndex = new NumberMap(nAtoms);
        for (var iChain = 0; iChain < nChains; iChain++) {
            chainsByLabelEntityId.add(label_entity_id.value(iChain), iChain);
            chainsByLabelAsymId.add(label_asym_id.value(iChain), iChain);
            chainsByAuthAsymId.add(auth_asym_id.value(iChain), iChain);
            var iResFrom = h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain]];
            var iResTo = h.residueAtomSegments.index[h.chainAtomSegments.offsets[iChain + 1] - 1] + 1;
            var residuesWithLabelSeqId = filterInPlace(range(iResFrom, iResTo), function (iRes) { return label_seq_id.valueKind(iRes) === Present; });
            residuesSortedByLabelSeqId.set(iChain, Sorting.create(residuesWithLabelSeqId, label_seq_id.value));
            var residuesWithAuthSeqId = filterInPlace(range(iResFrom, iResTo), function (iRes) { return auth_seq_id.valueKind(iRes) === Present; });
            residuesSortedByAuthSeqId.set(iChain, Sorting.create(residuesWithAuthSeqId, auth_seq_id.value));
            var residuesHereByInsCode = new MultiMap();
            for (var iRes = iResFrom; iRes < iResTo; iRes++) {
                if (pdbx_PDB_ins_code.valueKind(iRes) === Present) {
                    residuesHereByInsCode.add(pdbx_PDB_ins_code.value(iRes), iRes);
                }
            }
            residuesByInsCode.set(iChain, residuesHereByInsCode);
        }
        var atomId = model.atomicConformation.atomId.value;
        var atomIndex = h.atomSourceIndex.value;
        for (var iAtom = 0; iAtom < nAtoms; iAtom++) {
            atomsById.set(atomId(iAtom), iAtom);
            atomsByIndex.set(atomIndex(iAtom), iAtom);
        }
        return {
            chainsByLabelEntityId: chainsByLabelEntityId,
            chainsByLabelAsymId: chainsByLabelAsymId,
            chainsByAuthAsymId: chainsByAuthAsymId,
            residuesSortedByLabelSeqId: residuesSortedByLabelSeqId,
            residuesSortedByAuthSeqId: residuesSortedByAuthSeqId,
            residuesByInsCode: residuesByInsCode,
            atomsById: atomsById,
            atomsByIndex: atomsByIndex,
        };
    },
};
export var Sorting = {
    /** Create a `Sorting` from an array of keys and a function returning their corresponding values.
     * If two keys have the same value, the smaller key will come first.
     * This function modifies `keys` - create a copy if you need the original order! */
    create: function (keys, valueFunction) {
        sortIfNeeded(keys, function (a, b) { return valueFunction(a) - valueFunction(b) || a - b; });
        var values = SortedArray.ofSortedArray(keys.map(valueFunction));
        return { keys: keys, values: values };
    },
    /** Return a newly allocated array of keys which have value equal to `target`.
     * The returned keys are sorted by their value. */
    getKeysWithValue: function (sorting, target) {
        return Sorting.getKeysWithValueInRange(sorting, target, target);
    },
    /** Return a newly allocated array of keys which have value within interval `[min, max]` (inclusive).
     * The returned keys are sorted by their value.
     * Undefined `min` is interpreted as negative infitity, undefined `max` is interpreted as positive infinity. */
    getKeysWithValueInRange: function (sorting, min, max) {
        var keys = sorting.keys, values = sorting.values;
        if (!keys)
            return [];
        var n = keys.length;
        var from = (min !== undefined) ? SortedArray.findPredecessorIndex(values, min) : 0;
        var to;
        if (max !== undefined) {
            to = from;
            while (to < n && values[to] <= max)
                to++;
        }
        else {
            to = n;
        }
        return keys.slice(from, to);
    },
};
