// molstar/lib/extensions/mvs/helpers/utils.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __awaiter, __generator } from "tslib";
import { hashString } from '../../../mol-data/util';
import { Color } from '../../../mol-util/color';
import { ColorNames } from '../../../mol-util/color/names';
/** Try to await a promise and return an object with its result (if resolved) or with the error (if rejected) */
export function safePromise(promise) {
    return __awaiter(this, void 0, void 0, function () {
        var value, error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, promise];
                case 1:
                    value = _a.sent();
                    return [2 /*return*/, { ok: true, value: value }];
                case 2:
                    error_1 = _a.sent();
                    return [2 /*return*/, { ok: false, error: error_1 }];
                case 3: return [2 /*return*/];
            }
        });
    });
}
/** A map where values are arrays. Handles missing keys when adding values. */
var MultiMap = /** @class */ (function () {
    function MultiMap() {
        this._map = new Map();
    }
    /** Return the array of values assidned to a key (or `undefined` if no such values) */
    MultiMap.prototype.get = function (key) {
        return this._map.get(key);
    };
    /** Append value to a key (handles missing keys) */
    MultiMap.prototype.add = function (key, value) {
        if (!this._map.has(key)) {
            this._map.set(key, []);
        }
        this._map.get(key).push(value);
    };
    return MultiMap;
}());
export { MultiMap };
/** Implementation of `Map` where keys are integers
 * and most keys are expected to be from interval `[0, limit)`.
 * For the keys within this interval, performance is better than `Map` (implemented by array).
 * For the keys out of this interval, performance is slightly worse than `Map`. */
var NumberMap = /** @class */ (function () {
    function NumberMap(limit) {
        this.limit = limit;
        this.array = new Array(limit);
        this.map = new Map();
    }
    NumberMap.prototype.get = function (key) {
        if (0 <= key && key < this.limit)
            return this.array[key];
        else
            return this.map.get(key);
    };
    NumberMap.prototype.set = function (key, value) {
        if (0 <= key && key < this.limit)
            this.array[key] = value;
        else
            this.map.set(key, value);
    };
    return NumberMap;
}());
export { NumberMap };
/** Return `true` if `value` is not `undefined` or `null`.
 * Prefer this over `value !== undefined`
 * (for maybe if we want to allow `null` in `AnnotationRow` in the future) */
export function isDefined(value) {
    return value !== undefined && value !== null;
}
/** Return `true` if at least one of `values` is not `undefined` or `null`. */
export function isAnyDefined() {
    var values = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        values[_i] = arguments[_i];
    }
    return values.some(function (v) { return isDefined(v); });
}
/** Return filtered array containing all original elements except `undefined` or `null`. */
export function filterDefined(elements) {
    return elements.filter(function (x) { return x !== undefined && x !== null; });
}
/** Create an 8-hex-character hash for a given input string, e.g. 'spanish inquisition' -> '7f9ac4be' */
function stringHash32(input) {
    var uint32hash = hashString(input) >>> 0; // >>>0 converts to uint32, LOL
    return uint32hash.toString(16).padStart(8, '0');
}
/** Create an 16-hex-character hash for a given input string, e.g. 'spanish inquisition' -> '7f9ac4be544330be'*/
export function stringHash(input) {
    var reversed = input.split('').reverse().join('');
    return stringHash32(input) + stringHash32(reversed);
}
/** Convert `colorString` (either X11 color name like 'magenta' or hex code like '#ff00ff') to Color.
 * Return `undefined` if `colorString` cannot be converted. */
export function decodeColor(colorString) {
    if (colorString === undefined)
        return undefined;
    var result;
    if (HexColor.is(colorString)) {
        if (colorString.length === 4) {
            // convert short form to full form (#f0f -> #ff00ff)
            colorString = "#".concat(colorString[1]).concat(colorString[1]).concat(colorString[2]).concat(colorString[2]).concat(colorString[3]).concat(colorString[3]);
        }
        result = Color.fromHexStyle(colorString);
        if (result !== undefined && !isNaN(result))
            return result;
    }
    result = ColorNames[colorString.toLowerCase()];
    if (result !== undefined)
        return result;
    return undefined;
}
/** Regular expression matching a hexadecimal color string, e.g. '#FF1100' or '#f10' */
var hexColorRegex = /^#([0-9A-F]{3}){1,2}$/i;
export var HexColor = {
    /** Decide if a string is a valid hexadecimal color string (6-digit or 3-digit, e.g. '#FF1100' or '#f10') */
    is: function (str) {
        return typeof str === 'string' && hexColorRegex.test(str);
    },
};
