// molstar/lib/commonjs/extensions/mvs/helpers/selections.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.groupRows = exports.rowsToExpression = exports.rowToExpression = exports.atomQualifies = exports.getAtomRangesForRows = exports.getAtomRangesForRow = void 0;
var db_1 = require("../../../mol-data/db");
var builder_1 = require("../../../mol-script/language/builder");
var array_1 = require("../../../mol-util/array");
var atom_ranges_1 = require("./atom-ranges");
var indexing_1 = require("./indexing");
var utils_1 = require("./utils");
var EmptyArray = [];
/** Return atom ranges in `model` which satisfy criteria given by `row` */
function getAtomRangesForRow(model, row, indices) {
    var h = model.atomicHierarchy;
    var nAtoms = h.atoms._rowCount;
    var hasAtomIds = (0, utils_1.isAnyDefined)(row.atom_id, row.atom_index);
    var hasAtomFilter = (0, utils_1.isAnyDefined)(row.label_atom_id, row.auth_atom_id, row.type_symbol);
    var hasResidueFilter = (0, utils_1.isAnyDefined)(row.label_seq_id, row.auth_seq_id, row.pdbx_PDB_ins_code, row.beg_label_seq_id, row.end_label_seq_id, row.beg_auth_seq_id, row.end_auth_seq_id);
    var hasChainFilter = (0, utils_1.isAnyDefined)(row.label_asym_id, row.auth_asym_id, row.label_entity_id);
    if (hasAtomIds) {
        var theAtom = getTheAtomForRow(model, row, indices);
        return theAtom !== undefined ? atom_ranges_1.AtomRanges.single(theAtom, theAtom + 1) : atom_ranges_1.AtomRanges.empty();
    }
    if (!hasChainFilter && !hasResidueFilter && !hasAtomFilter) {
        return atom_ranges_1.AtomRanges.single(0, nAtoms);
    }
    var qualifyingChains = getQualifyingChains(model, row, indices);
    if (!hasResidueFilter && !hasAtomFilter) {
        var chainOffsets = h.chainAtomSegments.offsets;
        var ranges_1 = atom_ranges_1.AtomRanges.empty();
        for (var _i = 0, qualifyingChains_1 = qualifyingChains; _i < qualifyingChains_1.length; _i++) {
            var iChain = qualifyingChains_1[_i];
            atom_ranges_1.AtomRanges.add(ranges_1, chainOffsets[iChain], chainOffsets[iChain + 1]);
        }
        return ranges_1;
    }
    var qualifyingResidues = getQualifyingResidues(model, row, indices, qualifyingChains);
    if (!hasAtomFilter) {
        var residueOffsets = h.residueAtomSegments.offsets;
        var ranges_2 = atom_ranges_1.AtomRanges.empty();
        for (var _a = 0, qualifyingResidues_1 = qualifyingResidues; _a < qualifyingResidues_1.length; _a++) {
            var iRes = qualifyingResidues_1[_a];
            atom_ranges_1.AtomRanges.add(ranges_2, residueOffsets[iRes], residueOffsets[iRes + 1]);
        }
        return ranges_2;
    }
    var qualifyingAtoms = getQualifyingAtoms(model, row, indices, qualifyingResidues);
    var ranges = atom_ranges_1.AtomRanges.empty();
    for (var _b = 0, qualifyingAtoms_1 = qualifyingAtoms; _b < qualifyingAtoms_1.length; _b++) {
        var iAtom = qualifyingAtoms_1[_b];
        atom_ranges_1.AtomRanges.add(ranges, iAtom, iAtom + 1);
    }
    return ranges;
}
exports.getAtomRangesForRow = getAtomRangesForRow;
/** Return atom ranges in `model` which satisfy criteria given by any of `rows` (atoms that satisfy more rows are still included only once) */
function getAtomRangesForRows(model, rows, indices) {
    if (Array.isArray(rows)) {
        return atom_ranges_1.AtomRanges.union(rows.map(function (row) { return getAtomRangesForRow(model, row, indices); }));
    }
    else {
        return getAtomRangesForRow(model, rows, indices);
    }
}
exports.getAtomRangesForRows = getAtomRangesForRows;
/** Return an array of chain indexes which satisfy criteria given by `row` */
function getQualifyingChains(model, row, indices) {
    var _a, _b, _c;
    var _d = model.atomicHierarchy.chains, auth_asym_id = _d.auth_asym_id, label_entity_id = _d.label_entity_id, nChains = _d._rowCount;
    var result = undefined;
    if ((0, utils_1.isDefined)(row.label_asym_id)) {
        result = (_a = indices.chainsByLabelAsymId.get(row.label_asym_id)) !== null && _a !== void 0 ? _a : EmptyArray;
    }
    if ((0, utils_1.isDefined)(row.auth_asym_id)) {
        if (result) {
            result = result.filter(function (i) { return auth_asym_id.value(i) === row.auth_asym_id; });
        }
        else {
            result = (_b = indices.chainsByAuthAsymId.get(row.auth_asym_id)) !== null && _b !== void 0 ? _b : EmptyArray;
        }
    }
    if ((0, utils_1.isDefined)(row.label_entity_id)) {
        if (result) {
            result = result.filter(function (i) { return label_entity_id.value(i) === row.label_entity_id; });
        }
        else {
            result = (_c = indices.chainsByLabelEntityId.get(row.label_entity_id)) !== null && _c !== void 0 ? _c : EmptyArray;
        }
    }
    result !== null && result !== void 0 ? result : (result = (0, array_1.range)(nChains));
    return result;
}
/** Return an array of residue indexes which satisfy criteria given by `row` */
function getQualifyingResidues(model, row, indices, fromChains) {
    var _a;
    var _b = model.atomicHierarchy.residues, label_seq_id = _b.label_seq_id, auth_seq_id = _b.auth_seq_id, pdbx_PDB_ins_code = _b.pdbx_PDB_ins_code;
    var Present = db_1.Column.ValueKind.Present;
    var result = [];
    for (var _i = 0, fromChains_1 = fromChains; _i < fromChains_1.length; _i++) {
        var iChain = fromChains_1[_i];
        var residuesHere = undefined;
        if ((0, utils_1.isDefined)(row.label_seq_id)) {
            var sorting = indices.residuesSortedByLabelSeqId.get(iChain);
            residuesHere = indexing_1.Sorting.getKeysWithValue(sorting, row.label_seq_id);
        }
        if ((0, utils_1.isDefined)(row.auth_seq_id)) {
            if (residuesHere) {
                residuesHere = residuesHere.filter(function (i) { return auth_seq_id.valueKind(i) === Present && auth_seq_id.value(i) === row.auth_seq_id; });
            }
            else {
                var sorting = indices.residuesSortedByAuthSeqId.get(iChain);
                residuesHere = indexing_1.Sorting.getKeysWithValue(sorting, row.auth_seq_id);
            }
        }
        if ((0, utils_1.isDefined)(row.pdbx_PDB_ins_code)) {
            if (residuesHere) {
                residuesHere = residuesHere.filter(function (i) { return pdbx_PDB_ins_code.value(i) === row.pdbx_PDB_ins_code; });
            }
            else {
                residuesHere = (_a = indices.residuesByInsCode.get(iChain).get(row.pdbx_PDB_ins_code)) !== null && _a !== void 0 ? _a : EmptyArray;
            }
        }
        if ((0, utils_1.isDefined)(row.beg_label_seq_id) || (0, utils_1.isDefined)(row.end_label_seq_id)) {
            if (residuesHere) {
                if ((0, utils_1.isDefined)(row.beg_label_seq_id)) {
                    residuesHere = residuesHere.filter(function (i) { return label_seq_id.valueKind(i) === Present && label_seq_id.value(i) >= row.beg_label_seq_id; });
                }
                if ((0, utils_1.isDefined)(row.end_label_seq_id)) {
                    residuesHere = residuesHere.filter(function (i) { return label_seq_id.valueKind(i) === Present && label_seq_id.value(i) <= row.end_label_seq_id; });
                }
            }
            else {
                var sorting = indices.residuesSortedByLabelSeqId.get(iChain);
                residuesHere = indexing_1.Sorting.getKeysWithValueInRange(sorting, row.beg_label_seq_id, row.end_label_seq_id);
            }
        }
        if ((0, utils_1.isDefined)(row.beg_auth_seq_id) || (0, utils_1.isDefined)(row.end_auth_seq_id)) {
            if (residuesHere) {
                if ((0, utils_1.isDefined)(row.beg_auth_seq_id)) {
                    residuesHere = residuesHere.filter(function (i) { return auth_seq_id.valueKind(i) === Present && auth_seq_id.value(i) >= row.beg_auth_seq_id; });
                }
                if ((0, utils_1.isDefined)(row.end_auth_seq_id)) {
                    residuesHere = residuesHere.filter(function (i) { return auth_seq_id.valueKind(i) === Present && auth_seq_id.value(i) <= row.end_auth_seq_id; });
                }
            }
            else {
                var sorting = indices.residuesSortedByAuthSeqId.get(iChain);
                residuesHere = indexing_1.Sorting.getKeysWithValueInRange(sorting, row.beg_auth_seq_id, row.end_auth_seq_id);
            }
        }
        if (!residuesHere) {
            var _c = model.atomicHierarchy, residueAtomSegments = _c.residueAtomSegments, chainAtomSegments = _c.chainAtomSegments;
            var firstResidueForChain = residueAtomSegments.index[chainAtomSegments.offsets[iChain]];
            var firstResidueAfterChain = residueAtomSegments.index[chainAtomSegments.offsets[iChain + 1] - 1] + 1;
            residuesHere = (0, array_1.range)(firstResidueForChain, firstResidueAfterChain);
        }
        (0, array_1.arrayExtend)(result, residuesHere);
    }
    return result;
}
/** Return an array of atom indexes which satisfy criteria given by `row` */
function getQualifyingAtoms(model, row, indices, fromResidues) {
    var _a = model.atomicHierarchy.atoms, label_atom_id = _a.label_atom_id, auth_atom_id = _a.auth_atom_id, type_symbol = _a.type_symbol;
    var residueAtomSegments_offsets = model.atomicHierarchy.residueAtomSegments.offsets;
    var result = [];
    for (var _i = 0, fromResidues_1 = fromResidues; _i < fromResidues_1.length; _i++) {
        var iRes = fromResidues_1[_i];
        var atomIdcs = (0, array_1.range)(residueAtomSegments_offsets[iRes], residueAtomSegments_offsets[iRes + 1]);
        if ((0, utils_1.isDefined)(row.label_atom_id)) {
            (0, array_1.filterInPlace)(atomIdcs, function (iAtom) { return label_atom_id.value(iAtom) === row.label_atom_id; });
        }
        if ((0, utils_1.isDefined)(row.auth_atom_id)) {
            (0, array_1.filterInPlace)(atomIdcs, function (iAtom) { return auth_atom_id.value(iAtom) === row.auth_atom_id; });
        }
        if ((0, utils_1.isDefined)(row.type_symbol)) {
            (0, array_1.filterInPlace)(atomIdcs, function (iAtom) { var _a; return type_symbol.value(iAtom) === ((_a = row.type_symbol) === null || _a === void 0 ? void 0 : _a.toUpperCase()); });
        }
        (0, array_1.arrayExtend)(result, atomIdcs);
    }
    return result;
}
/** Return index of atom in `model` which satistfies criteria given by `row`, if any.
 * Only works when `row.atom_id` and/or `row.atom_index` is defined (otherwise use `getAtomRangesForRow`). */
function getTheAtomForRow(model, row, indices) {
    var iAtom = undefined;
    if (!(0, utils_1.isDefined)(row.atom_id) && !(0, utils_1.isDefined)(row.atom_index))
        throw new Error('ArgumentError: at least one of row.atom_id, row.atom_index must be defined.');
    if ((0, utils_1.isDefined)(row.atom_id) && (0, utils_1.isDefined)(row.atom_index)) {
        var a1 = indices.atomsById.get(row.atom_id);
        var a2 = indices.atomsByIndex.get(row.atom_index);
        if (a1 !== a2)
            return undefined;
        iAtom = a1;
    }
    if ((0, utils_1.isDefined)(row.atom_id)) {
        iAtom = indices.atomsById.get(row.atom_id);
    }
    if ((0, utils_1.isDefined)(row.atom_index)) {
        iAtom = indices.atomsByIndex.get(row.atom_index);
    }
    if (iAtom === undefined)
        return undefined;
    if (!atomQualifies(model, iAtom, row))
        return undefined;
    return iAtom;
}
/** Return true if `iAtom`-th atom in `model` satisfies all selection criteria given by `row`. */
function atomQualifies(model, iAtom, row) {
    var _a;
    var h = model.atomicHierarchy;
    var iChain = h.chainAtomSegments.index[iAtom];
    var label_asym_id = h.chains.label_asym_id.value(iChain);
    var auth_asym_id = h.chains.auth_asym_id.value(iChain);
    var label_entity_id = h.chains.label_entity_id.value(iChain);
    if (!matches(row.label_asym_id, label_asym_id))
        return false;
    if (!matches(row.auth_asym_id, auth_asym_id))
        return false;
    if (!matches(row.label_entity_id, label_entity_id))
        return false;
    var iRes = h.residueAtomSegments.index[iAtom];
    var label_seq_id = (h.residues.label_seq_id.valueKind(iRes) === db_1.Column.ValueKind.Present) ? h.residues.label_seq_id.value(iRes) : undefined;
    var auth_seq_id = (h.residues.auth_seq_id.valueKind(iRes) === db_1.Column.ValueKind.Present) ? h.residues.auth_seq_id.value(iRes) : undefined;
    var pdbx_PDB_ins_code = h.residues.pdbx_PDB_ins_code.value(iRes);
    if (!matches(row.label_seq_id, label_seq_id))
        return false;
    if (!matches(row.auth_seq_id, auth_seq_id))
        return false;
    if (!matches(row.pdbx_PDB_ins_code, pdbx_PDB_ins_code))
        return false;
    if (!matchesRange(row.beg_label_seq_id, row.end_label_seq_id, label_seq_id))
        return false;
    if (!matchesRange(row.beg_auth_seq_id, row.end_auth_seq_id, auth_seq_id))
        return false;
    var label_atom_id = h.atoms.label_atom_id.value(iAtom);
    var auth_atom_id = h.atoms.auth_atom_id.value(iAtom);
    var type_symbol = h.atoms.type_symbol.value(iAtom);
    var atom_id = model.atomicConformation.atomId.value(iAtom);
    var atom_index = h.atomSourceIndex.value(iAtom);
    if (!matches(row.label_atom_id, label_atom_id))
        return false;
    if (!matches(row.auth_atom_id, auth_atom_id))
        return false;
    if (!matches((_a = row.type_symbol) === null || _a === void 0 ? void 0 : _a.toUpperCase(), type_symbol))
        return false;
    if (!matches(row.atom_id, atom_id))
        return false;
    if (!matches(row.atom_index, atom_index))
        return false;
    return true;
}
exports.atomQualifies = atomQualifies;
/** Return true if `value` equals `requiredValue` or if `requiredValue` if not defined.  */
function matches(requiredValue, value) {
    return !(0, utils_1.isDefined)(requiredValue) || value === requiredValue;
}
/** Return true if `requiredMin <= value <= requiredMax`.
 * Undefined `requiredMin` behaves like negative infinity.
 * Undefined `requiredMax` behaves like positive infinity. */
function matchesRange(requiredMin, requiredMax, value) {
    if ((0, utils_1.isDefined)(requiredMin) && (!(0, utils_1.isDefined)(value) || value < requiredMin))
        return false;
    if ((0, utils_1.isDefined)(requiredMax) && (!(0, utils_1.isDefined)(value) || value > requiredMax))
        return false;
    return true;
}
/** Convert an annotation row into a MolScript expression */
function rowToExpression(row) {
    var and = builder_1.MolScriptBuilder.core.logic.and;
    var _a = builder_1.MolScriptBuilder.core.rel, eq = _a.eq, gte = _a.gre, lte = _a.lte;
    var macromolecular = builder_1.MolScriptBuilder.struct.atomProperty.macromolecular;
    var propTests = {};
    if ((0, utils_1.isDefined)(row.label_entity_id)) {
        propTests['entity-test'] = eq([macromolecular.label_entity_id(), row.label_entity_id]);
    }
    var chainTests = [];
    if ((0, utils_1.isDefined)(row.label_asym_id))
        chainTests.push(eq([macromolecular.label_asym_id(), row.label_asym_id]));
    if ((0, utils_1.isDefined)(row.auth_asym_id))
        chainTests.push(eq([macromolecular.auth_asym_id(), row.auth_asym_id]));
    if (chainTests.length === 1) {
        propTests['chain-test'] = chainTests[0];
    }
    else if (chainTests.length > 1) {
        propTests['chain-test'] = and(chainTests);
    }
    var residueTests = [];
    if ((0, utils_1.isDefined)(row.label_seq_id))
        residueTests.push(eq([macromolecular.label_seq_id(), row.label_seq_id]));
    if ((0, utils_1.isDefined)(row.auth_seq_id))
        residueTests.push(eq([macromolecular.auth_seq_id(), row.auth_seq_id]));
    if ((0, utils_1.isDefined)(row.pdbx_PDB_ins_code))
        residueTests.push(eq([macromolecular.pdbx_PDB_ins_code(), row.pdbx_PDB_ins_code]));
    if ((0, utils_1.isDefined)(row.beg_label_seq_id))
        residueTests.push(gte([macromolecular.label_seq_id(), row.beg_label_seq_id]));
    if ((0, utils_1.isDefined)(row.end_label_seq_id))
        residueTests.push(lte([macromolecular.label_seq_id(), row.end_label_seq_id]));
    if ((0, utils_1.isDefined)(row.beg_auth_seq_id))
        residueTests.push(gte([macromolecular.auth_seq_id(), row.beg_auth_seq_id]));
    if ((0, utils_1.isDefined)(row.end_auth_seq_id))
        residueTests.push(lte([macromolecular.auth_seq_id(), row.end_auth_seq_id]));
    if (residueTests.length === 1) {
        propTests['residue-test'] = residueTests[0];
    }
    else if (residueTests.length > 1) {
        propTests['residue-test'] = and(residueTests);
    }
    var atomTests = [];
    if ((0, utils_1.isDefined)(row.atom_id))
        atomTests.push(eq([macromolecular.id(), row.atom_id]));
    if ((0, utils_1.isDefined)(row.atom_index))
        atomTests.push(eq([builder_1.MolScriptBuilder.struct.atomProperty.core.sourceIndex(), row.atom_index]));
    if ((0, utils_1.isDefined)(row.label_atom_id))
        atomTests.push(eq([macromolecular.label_atom_id(), row.label_atom_id]));
    if ((0, utils_1.isDefined)(row.auth_atom_id))
        atomTests.push(eq([macromolecular.auth_atom_id(), row.auth_atom_id]));
    if ((0, utils_1.isDefined)(row.type_symbol))
        atomTests.push(eq([builder_1.MolScriptBuilder.struct.atomProperty.core.elementSymbol(), row.type_symbol.toUpperCase()]));
    if (atomTests.length === 1) {
        propTests['atom-test'] = atomTests[0];
    }
    else if (atomTests.length > 1) {
        propTests['atom-test'] = and(atomTests);
    }
    return builder_1.MolScriptBuilder.struct.generator.atomGroups(propTests);
}
exports.rowToExpression = rowToExpression;
/** Convert multiple annotation rows into a MolScript expression.
 * (with union semantics, i.e. an atom qualifies if it qualifies for at least one of the rows) */
function rowsToExpression(rows) {
    return unionExpression(rows.map(rowToExpression));
}
exports.rowsToExpression = rowsToExpression;
/** Create MolScript expression covering the set union of the given expressions */
function unionExpression(expressions) {
    return builder_1.MolScriptBuilder.struct.combinator.merge(expressions.map(function (e) { return builder_1.MolScriptBuilder.struct.modifier.union([e]); }));
}
/** Return row indices grouped by `row.group_id`. Rows with `row.group_id===undefined` are treated as separate groups. */
function groupRows(rows) {
    var counter = 0;
    var groupMap = new Map();
    var groups = [];
    for (var i = 0; i < rows.length; i++) {
        var group_id = rows[i].group_id;
        if (group_id === undefined) {
            groups.push(counter++);
        }
        else {
            var groupIndex = groupMap.get(group_id);
            if (groupIndex === undefined) {
                groupMap.set(group_id, counter);
                groups.push(counter);
                counter++;
            }
            else {
                groups.push(groupIndex);
            }
        }
    }
    var rowIndices = (0, array_1.range)(rows.length).sort(function (i, j) { return groups[i] - groups[j]; });
    var offsets = [];
    for (var i = 0; i < rows.length; i++) {
        if (i === 0 || groups[rowIndices[i]] !== groups[rowIndices[i - 1]])
            offsets.push(i);
    }
    offsets.push(rowIndices.length);
    return { count: offsets.length - 1, offsets: offsets, grouped: rowIndices };
}
exports.groupRows = groupRows;
