// molstar/lib/commonjs/extensions/mvs/camera.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setCanvas = exports.setFocus = exports.setCamera = void 0;
var tslib_1 = require("tslib");
var boundary_helper_1 = require("../../mol-math/geometry/boundary-helper");
var linear_algebra_1 = require("../../mol-math/linear-algebra");
var loci_1 = require("../../mol-model/loci");
var structure_1 = require("../../mol-model/structure");
var objects_1 = require("../../mol-plugin-state/objects");
var commands_1 = require("../../mol-plugin/commands");
var names_1 = require("../../mol-util/color/names");
var utils_1 = require("./helpers/utils");
var mvs_defaults_1 = require("./tree/mvs/mvs-defaults");
var DefaultFocusOptions = {
    minRadius: 5,
    extraRadiusForFocus: 0,
    extraRadiusForZoomAll: 0,
};
var DefaultCanvasBackgroundColor = names_1.ColorNames.white;
var _tmpVec = (0, linear_algebra_1.Vec3)();
/** Set the camera based on a camera node params. */
function setCamera(plugin, params) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var target, position, up, snapshot;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    target = linear_algebra_1.Vec3.create.apply(linear_algebra_1.Vec3, params.target);
                    position = linear_algebra_1.Vec3.create.apply(linear_algebra_1.Vec3, params.position);
                    if (plugin.canvas3d)
                        position = fovAdjustedPosition(target, position, plugin.canvas3d.camera.state.mode, plugin.canvas3d.camera.state.fov);
                    up = linear_algebra_1.Vec3.create.apply(linear_algebra_1.Vec3, params.up);
                    linear_algebra_1.Vec3.orthogonalize(up, linear_algebra_1.Vec3.sub(_tmpVec, target, position), up);
                    snapshot = { target: target, position: position, up: up, radius: 10000, 'radiusMax': 10000 };
                    return [4 /*yield*/, commands_1.PluginCommands.Camera.SetSnapshot(plugin, { snapshot: snapshot })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.setCamera = setCamera;
/** Focus the camera on the bounding sphere of a (sub)structure (or on the whole scene if `structureNodeSelector` is null).
 * Orient the camera based on a focus node params. */
function setFocus(plugin, structureNodeSelector, params) {
    var _a;
    if (params === void 0) { params = mvs_defaults_1.MVSDefaults.focus; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var structure, cell, boundingSphere, extraRadius, direction, up, snapshot;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    structure = undefined;
                    if (structureNodeSelector) {
                        cell = plugin.state.data.cells.get(structureNodeSelector.ref);
                        structure = (_a = cell === null || cell === void 0 ? void 0 : cell.obj) === null || _a === void 0 ? void 0 : _a.data;
                        if (!structure)
                            console.warn('Focus: no structure');
                        if (!(structure instanceof structure_1.Structure)) {
                            console.warn('Focus: cannot apply to a non-structure node');
                            structure = undefined;
                        }
                    }
                    boundingSphere = structure ? loci_1.Loci.getBoundingSphere(structure_1.Structure.Loci(structure)) : getPluginBoundingSphere(plugin);
                    if (!(boundingSphere && plugin.canvas3d)) return [3 /*break*/, 2];
                    extraRadius = structure ? DefaultFocusOptions.extraRadiusForFocus : DefaultFocusOptions.extraRadiusForZoomAll;
                    direction = linear_algebra_1.Vec3.create.apply(linear_algebra_1.Vec3, params.direction);
                    up = linear_algebra_1.Vec3.create.apply(linear_algebra_1.Vec3, params.up);
                    linear_algebra_1.Vec3.orthogonalize(up, direction, up);
                    snapshot = snapshotFromSphereAndDirections(plugin.canvas3d.camera, {
                        center: boundingSphere.center,
                        radius: boundingSphere.radius + extraRadius,
                        up: up,
                        direction: direction,
                    });
                    return [4 /*yield*/, commands_1.PluginCommands.Camera.SetSnapshot(plugin, { snapshot: snapshot })];
                case 1:
                    _b.sent();
                    _b.label = 2;
                case 2: return [2 /*return*/];
            }
        });
    });
}
exports.setFocus = setFocus;
/** Return camera snapshot for focusing a sphere with given `center` and `radius`,
 * while ensuring given view `direction` (aligns with vector position->target)
 * and `up` (aligns with screen Y axis). */
function snapshotFromSphereAndDirections(camera, options) {
    // This might seem to repeat `plugin.canvas3d.camera.getFocus` but avoid flipping
    var center = options.center, direction = options.direction, up = options.up;
    var radius = Math.max(options.radius, DefaultFocusOptions.minRadius);
    var distance = camera.getTargetDistance(radius);
    var deltaDirection = linear_algebra_1.Vec3.setMagnitude(_tmpVec, direction, distance);
    var position = linear_algebra_1.Vec3.sub((0, linear_algebra_1.Vec3)(), center, deltaDirection);
    return { target: center, position: position, up: up, radius: radius };
}
/** Return the distance adjustment ratio for conversion from the "reference camera"
 * to a camera with an arbitrary field of view `fov`. */
function distanceAdjustment(mode, fov) {
    if (mode === 'orthographic')
        return 1 / (2 * Math.tan(fov / 2));
    else
        return 1 / (2 * Math.sin(fov / 2));
}
/** Return the position for a camera with an arbitrary field of view `fov`
 * necessary to just fit into view the same sphere (with center at `target`)
 * as the "reference camera" placed at `refPosition` would fit, while keeping the camera orientation.
 * The "reference camera" is a camera which can just fit into view a sphere of radius R with center at distance 2R
 * (this corresponds to FOV = 2 * asin(1/2) in perspective mode or FOV = 2 * atan(1/2) in orthographic mode). */
function fovAdjustedPosition(target, refPosition, mode, fov) {
    var delta = linear_algebra_1.Vec3.sub((0, linear_algebra_1.Vec3)(), refPosition, target);
    var adjustment = distanceAdjustment(mode, fov);
    return linear_algebra_1.Vec3.scaleAndAdd(delta, target, delta, adjustment); // return target + delta * adjustment
}
/** Compute the bounding sphere of the whole scene. */
function getPluginBoundingSphere(plugin) {
    var renderObjects = getRenderObjects(plugin, false);
    var spheres = renderObjects.map(function (r) { return r.values.boundingSphere.ref.value; }).filter(function (sphere) { return sphere.radius > 0; });
    return boundingSphereOfSpheres(spheres);
}
function getRenderObjects(plugin, includeHidden) {
    var reprCells = Array.from(plugin.state.data.cells.values()).filter(function (cell) { return cell.obj && objects_1.PluginStateObject.isRepresentation3D(cell.obj); });
    if (!includeHidden)
        reprCells = reprCells.filter(function (cell) { return !cell.state.isHidden; });
    var renderables = reprCells.flatMap(function (cell) { return cell.obj.data.repr.renderObjects; });
    return renderables;
}
var boundaryHelper = undefined;
function boundingSphereOfSpheres(spheres) {
    boundaryHelper !== null && boundaryHelper !== void 0 ? boundaryHelper : (boundaryHelper = new boundary_helper_1.BoundaryHelper('98'));
    boundaryHelper.reset();
    for (var _i = 0, spheres_1 = spheres; _i < spheres_1.length; _i++) {
        var s = spheres_1[_i];
        boundaryHelper.includeSphere(s);
    }
    boundaryHelper.finishedIncludeStep();
    for (var _a = 0, spheres_2 = spheres; _a < spheres_2.length; _a++) {
        var s = spheres_2[_a];
        boundaryHelper.radiusSphere(s);
    }
    return boundaryHelper.getSphere();
}
/** Set canvas properties based on a canvas node params. */
function setCanvas(plugin, params) {
    var _a, _b, _c;
    var backgroundColor = (_a = (0, utils_1.decodeColor)(params === null || params === void 0 ? void 0 : params.background_color)) !== null && _a !== void 0 ? _a : DefaultCanvasBackgroundColor;
    if (backgroundColor !== ((_b = plugin.canvas3d) === null || _b === void 0 ? void 0 : _b.props.renderer.backgroundColor)) {
        (_c = plugin.canvas3d) === null || _c === void 0 ? void 0 : _c.setProps(function (old) { return (tslib_1.__assign(tslib_1.__assign({}, old), { renderer: tslib_1.__assign(tslib_1.__assign({}, old.renderer), { backgroundColor: backgroundColor }) })); });
    }
}
exports.setCanvas = setCanvas;
