// molstar/lib/commonjs/extensions/mvs/load.js
"use strict";
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadMVS = void 0;
var tslib_1 = require("tslib");
var data_1 = require("../../mol-plugin-state/transforms/data");
var model_1 = require("../../mol-plugin-state/transforms/model");
var representation_1 = require("../../mol-plugin-state/transforms/representation");
var behavior_1 = require("./behavior");
var camera_1 = require("./camera");
var annotation_prop_1 = require("./components/annotation-prop");
var annotation_structure_component_1 = require("./components/annotation-structure-component");
var annotation_tooltips_prop_1 = require("./components/annotation-tooltips-prop");
var representation_2 = require("./components/custom-label/representation");
var custom_tooltips_prop_1 = require("./components/custom-tooltips-prop");
var is_mvs_model_prop_1 = require("./components/is-mvs-model-prop");
var load_helpers_1 = require("./load-helpers");
var tree_schema_1 = require("./tree/generic/tree-schema");
var conversion_1 = require("./tree/molstar/conversion");
var molstar_tree_1 = require("./tree/molstar/molstar-tree");
var mvs_tree_1 = require("./tree/mvs/mvs-tree");
/** Load a MolViewSpec (MVS) tree into the Mol* plugin.
 * If `options.replaceExisting`, remove all objects in the current Mol* state; otherwise add to the current state.
 * If `options.sanityChecks`, run some sanity checks and print potential issues to the console.
 * `options.sourceUrl` serves as the base for resolving relative URLs/URIs and may itself be relative to the window URL. */
function loadMVS(plugin, data, options) {
    if (options === void 0) { options = {}; }
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var molstarTree, err_1;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    // console.log(`MVS tree:\n${MVSData.toPrettyString(data)}`)
                    (0, tree_schema_1.validateTree)(mvs_tree_1.MVSTreeSchema, data.root, 'MVS');
                    if (options.sanityChecks)
                        (0, conversion_1.mvsSanityCheck)(data.root);
                    molstarTree = (0, conversion_1.convertMvsToMolstar)(data.root, options.sourceUrl);
                    // console.log(`Converted MolStar tree:\n${MVSData.toPrettyString({ root: molstarTree, metadata: { version: 'x', timestamp: 'x' } })}`)
                    (0, tree_schema_1.validateTree)(molstar_tree_1.MolstarTreeSchema, molstarTree, 'Converted Molstar');
                    return [4 /*yield*/, loadMolstarTree(plugin, molstarTree, options)];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    err_1 = _a.sent();
                    plugin.log.error("".concat(err_1));
                    throw err_1;
                case 3: return [2 /*return*/];
            }
        });
    });
}
exports.loadMVS = loadMVS;
/** Load a `MolstarTree` into the Mol* plugin.
 * If `replaceExisting`, remove all objects in the current Mol* state; otherwise add to the current state. */
function loadMolstarTree(plugin, tree, options) {
    var _a, _b;
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var mvsExtensionLoaded, context;
        return tslib_1.__generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    mvsExtensionLoaded = plugin.state.hasBehavior(behavior_1.MolViewSpec);
                    if (!mvsExtensionLoaded)
                        throw new Error('MolViewSpec extension is not loaded.');
                    context = {};
                    return [4 /*yield*/, (0, load_helpers_1.loadTree)(plugin, tree, MolstarLoadingActions, context, options)];
                case 1:
                    _c.sent();
                    (0, camera_1.setCanvas)(plugin, context.canvas);
                    if (!(((_a = context.focus) === null || _a === void 0 ? void 0 : _a.kind) === 'camera')) return [3 /*break*/, 3];
                    return [4 /*yield*/, (0, camera_1.setCamera)(plugin, context.focus.params)];
                case 2:
                    _c.sent();
                    return [3 /*break*/, 7];
                case 3:
                    if (!(((_b = context.focus) === null || _b === void 0 ? void 0 : _b.kind) === 'focus')) return [3 /*break*/, 5];
                    return [4 /*yield*/, (0, camera_1.setFocus)(plugin, context.focus.focusTarget, context.focus.params)];
                case 4:
                    _c.sent();
                    return [3 /*break*/, 7];
                case 5: return [4 /*yield*/, (0, camera_1.setFocus)(plugin, undefined, undefined)];
                case 6:
                    _c.sent();
                    _c.label = 7;
                case 7: return [2 /*return*/];
            }
        });
    });
}
/** Loading actions for loading a `MolstarTree`, per node kind. */
var MolstarLoadingActions = {
    root: function (updateParent, node, context) {
        context.nearestReprMap = (0, load_helpers_1.makeNearestReprMap)(node);
        return updateParent;
    },
    download: function (updateParent, node) {
        return load_helpers_1.UpdateTarget.apply(updateParent, data_1.Download, {
            url: node.params.url,
            isBinary: node.params.is_binary,
        });
    },
    parse: function (updateParent, node) {
        var format = node.params.format;
        if (format === 'cif') {
            return load_helpers_1.UpdateTarget.apply(updateParent, data_1.ParseCif, {});
        }
        else if (format === 'pdb') {
            return updateParent;
        }
        else {
            console.error("Unknown format in \"parse\" node: \"".concat(format, "\""));
            return undefined;
        }
    },
    trajectory: function (updateParent, node) {
        var _a, _b;
        var format = node.params.format;
        if (format === 'cif') {
            return load_helpers_1.UpdateTarget.apply(updateParent, model_1.TrajectoryFromMmCif, {
                blockHeader: (_a = node.params.block_header) !== null && _a !== void 0 ? _a : '', // Must set to '' because just undefined would get overwritten by createDefaults
                blockIndex: (_b = node.params.block_index) !== null && _b !== void 0 ? _b : undefined,
            });
        }
        else if (format === 'pdb') {
            return load_helpers_1.UpdateTarget.apply(updateParent, model_1.TrajectoryFromPDB, {});
        }
        else {
            console.error("Unknown format in \"trajectory\" node: \"".concat(format, "\""));
            return undefined;
        }
    },
    model: function (updateParent, node, context) {
        var _a;
        var annotations = (0, load_helpers_1.collectAnnotationReferences)(node, context);
        var model = load_helpers_1.UpdateTarget.apply(updateParent, model_1.ModelFromTrajectory, {
            modelIndex: node.params.model_index,
        });
        load_helpers_1.UpdateTarget.apply(model, model_1.CustomModelProperties, {
            properties: (_a = {},
                _a[is_mvs_model_prop_1.IsMVSModelProvider.descriptor.name] = { isMvs: true },
                _a[annotation_prop_1.MVSAnnotationsProvider.descriptor.name] = { annotations: annotations },
                _a),
            autoAttach: [
                is_mvs_model_prop_1.IsMVSModelProvider.descriptor.name,
                annotation_prop_1.MVSAnnotationsProvider.descriptor.name,
            ],
        });
        return model;
    },
    structure: function (updateParent, node, context) {
        var _a;
        var _b;
        var props = (0, load_helpers_1.structureProps)(node);
        var struct = load_helpers_1.UpdateTarget.apply(updateParent, model_1.StructureFromModel, props);
        var transformed = struct;
        for (var _i = 0, _c = (0, load_helpers_1.transformProps)(node); _i < _c.length; _i++) {
            var t = _c[_i];
            transformed = load_helpers_1.UpdateTarget.apply(transformed, model_1.TransformStructureConformation, t); // applying to the result of previous transform, to get the correct transform order
        }
        var annotationTooltips = (0, load_helpers_1.collectAnnotationTooltips)(node, context);
        var inlineTooltips = (0, load_helpers_1.collectInlineTooltips)(node, context);
        if (annotationTooltips.length + inlineTooltips.length > 0) {
            load_helpers_1.UpdateTarget.apply(struct, model_1.CustomStructureProperties, {
                properties: (_a = {},
                    _a[annotation_tooltips_prop_1.MVSAnnotationTooltipsProvider.descriptor.name] = { tooltips: annotationTooltips },
                    _a[custom_tooltips_prop_1.CustomTooltipsProvider.descriptor.name] = { tooltips: inlineTooltips },
                    _a),
                autoAttach: [
                    annotation_tooltips_prop_1.MVSAnnotationTooltipsProvider.descriptor.name,
                    custom_tooltips_prop_1.CustomTooltipsProvider.descriptor.name,
                ],
            });
        }
        var inlineLabels = (0, load_helpers_1.collectInlineLabels)(node, context);
        if (inlineLabels.length > 0) {
            var nearestReprNode = (_b = context.nearestReprMap) === null || _b === void 0 ? void 0 : _b.get(node);
            load_helpers_1.UpdateTarget.apply(struct, representation_1.StructureRepresentation3D, {
                type: {
                    name: representation_2.CustomLabelRepresentationProvider.name,
                    params: { items: inlineLabels },
                },
                colorTheme: (0, load_helpers_1.colorThemeForNode)(nearestReprNode, context),
            });
        }
        return struct;
    },
    tooltip: undefined, // No action needed, already loaded in `structure`
    tooltip_from_uri: undefined, // No action needed, already loaded in `structure`
    tooltip_from_source: undefined, // No action needed, already loaded in `structure`
    component: function (updateParent, node) {
        if ((0, load_helpers_1.isPhantomComponent)(node)) {
            return updateParent;
        }
        var selector = node.params.selector;
        return load_helpers_1.UpdateTarget.apply(updateParent, model_1.StructureComponent, {
            type: (0, load_helpers_1.componentPropsFromSelector)(selector),
            label: (0, load_helpers_1.prettyNameFromSelector)(selector),
            nullIfEmpty: false,
        });
    },
    component_from_uri: function (updateParent, node, context) {
        if ((0, load_helpers_1.isPhantomComponent)(node))
            return undefined;
        var props = (0, load_helpers_1.componentFromXProps)(node, context);
        return load_helpers_1.UpdateTarget.apply(updateParent, annotation_structure_component_1.MVSAnnotationStructureComponent, props);
    },
    component_from_source: function (updateParent, node, context) {
        if ((0, load_helpers_1.isPhantomComponent)(node))
            return undefined;
        var props = (0, load_helpers_1.componentFromXProps)(node, context);
        return load_helpers_1.UpdateTarget.apply(updateParent, annotation_structure_component_1.MVSAnnotationStructureComponent, props);
    },
    representation: function (updateParent, node, context) {
        return load_helpers_1.UpdateTarget.apply(updateParent, representation_1.StructureRepresentation3D, tslib_1.__assign(tslib_1.__assign({}, (0, load_helpers_1.representationProps)(node.params)), { colorTheme: (0, load_helpers_1.colorThemeForNode)(node, context) }));
    },
    color: undefined, // No action needed, already loaded in `representation`
    color_from_uri: undefined, // No action needed, already loaded in `representation`
    color_from_source: undefined, // No action needed, already loaded in `representation`
    label: undefined, // No action needed, already loaded in `structure`
    label_from_uri: function (updateParent, node, context) {
        var props = (0, load_helpers_1.labelFromXProps)(node, context);
        return load_helpers_1.UpdateTarget.apply(updateParent, representation_1.StructureRepresentation3D, props);
    },
    label_from_source: function (updateParent, node, context) {
        var props = (0, load_helpers_1.labelFromXProps)(node, context);
        return load_helpers_1.UpdateTarget.apply(updateParent, representation_1.StructureRepresentation3D, props);
    },
    focus: function (updateParent, node, context) {
        context.focus = { kind: 'focus', focusTarget: updateParent.selector, params: node.params };
        return updateParent;
    },
    camera: function (updateParent, node, context) {
        context.focus = { kind: 'camera', params: node.params };
        return updateParent;
    },
    canvas: function (updateParent, node, context) {
        context.canvas = node.params;
        return updateParent;
    },
};
