// molstar/lib/commonjs/mol-plugin-ui/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Log = exports.DefaultViewport = exports.ControlsWrapper = exports.PluginContextContainer = exports.Plugin = void 0;
var tslib_1 = require("tslib");
var jsx_runtime_1 = require("react/jsx-runtime");
var React = tslib_1.__importStar(require("react"));
var mol_util_1 = require("../mol-util");
var base_1 = require("./base");
var controls_1 = require("./controls");
var left_panel_1 = require("./left-panel");
var sequence_1 = require("./sequence");
var task_1 = require("./task");
var toast_1 = require("./toast");
var viewport_1 = require("./viewport");
var commands_1 = require("../mol-plugin/commands");
var file_1 = require("../mol-plugin-state/actions/file");
var assets_1 = require("../mol-util/assets");
var rxjs_1 = require("rxjs");
var use_behavior_1 = require("./hooks/use-behavior");
function Plugin(_a) {
    var plugin = _a.plugin;
    if (plugin.isInitialized) {
        return (0, jsx_runtime_1.jsx)(base_1.PluginReactContext.Provider, { value: plugin, children: (0, jsx_runtime_1.jsx)(Layout, {}) });
    }
    return (0, jsx_runtime_1.jsx)(PluginInitWrapper, { plugin: plugin });
}
exports.Plugin = Plugin;
function PluginInitWrapper(_a) {
    var plugin = _a.plugin;
    var _b = React.useState({ kind: 'pending' }), state = _b[0], setState = _b[1];
    React.useEffect(function () {
        setState({ kind: 'pending' });
        var mounted = true;
        plugin.initialized.then(function () {
            if (mounted)
                setState({ kind: 'initialized' });
        }).catch(function (err) {
            if (mounted)
                setState({ kind: 'error', message: "".concat(err) });
        });
        return function () { mounted = false; };
    }, [plugin]);
    if (state.kind === 'pending')
        return null;
    if (state.kind === 'error') {
        return (0, jsx_runtime_1.jsx)("div", { className: 'msp-plugin', children: (0, jsx_runtime_1.jsxs)("div", { className: 'msp-plugin-init-error', children: ["Initialization error: ", state.message] }) });
    }
    return (0, jsx_runtime_1.jsx)(base_1.PluginReactContext.Provider, { value: plugin, children: (0, jsx_runtime_1.jsx)(Layout, {}) });
}
var PluginContextContainer = /** @class */ (function (_super) {
    tslib_1.__extends(PluginContextContainer, _super);
    function PluginContextContainer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PluginContextContainer.prototype.render = function () {
        return (0, jsx_runtime_1.jsx)(base_1.PluginReactContext.Provider, { value: this.props.plugin, children: (0, jsx_runtime_1.jsx)("div", { className: 'msp-plugin', children: this.props.children }) });
    };
    return PluginContextContainer;
}(React.Component));
exports.PluginContextContainer = PluginContextContainer;
var Layout = /** @class */ (function (_super) {
    tslib_1.__extends(Layout, _super);
    function Layout() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onDrop = function (ev) {
            ev.preventDefault();
            var files = [];
            if (ev.dataTransfer.items) {
                // Use DataTransferItemList interface to access the file(s)
                for (var i = 0; i < ev.dataTransfer.items.length; i++) {
                    if (ev.dataTransfer.items[i].kind !== 'file')
                        continue;
                    var file = ev.dataTransfer.items[i].getAsFile();
                    if (file)
                        files.push(file);
                }
            }
            else {
                for (var i = 0; i < ev.dataTransfer.files.length; i++) {
                    var file = ev.dataTransfer.files[i];
                    if (file)
                        files.push(file);
                }
            }
            var sessions = files.filter(function (f) {
                var fn = f.name.toLowerCase();
                return fn.endsWith('.molx') || fn.endsWith('.molj');
            });
            if (sessions.length > 0) {
                commands_1.PluginCommands.State.Snapshots.OpenFile(_this.plugin, { file: sessions[0] });
            }
            else {
                _this.plugin.runTask(_this.plugin.state.data.applyAction(file_1.OpenFiles, {
                    files: files.map(function (f) { return assets_1.Asset.File(f); }),
                    format: { name: 'auto', params: {} },
                    visuals: true
                }));
            }
        };
        _this.onDragOver = function (ev) {
            ev.preventDefault();
        };
        _this.showDragOverlay = new rxjs_1.BehaviorSubject(false);
        _this.onDragEnter = function (ev) {
            var hasFile = false;
            if (ev.dataTransfer.items && ev.dataTransfer.items.length > 0) {
                for (var i = 0; i < ev.dataTransfer.items.length; i++) {
                    if (ev.dataTransfer.items[i].kind !== 'file')
                        continue;
                    hasFile = true;
                    break;
                }
            }
            else {
                for (var i = 0; i < ev.dataTransfer.types.length; i++) {
                    if (ev.dataTransfer.types[i] !== 'Files')
                        continue;
                    hasFile = true;
                    break;
                }
            }
            if (hasFile) {
                _this.showDragOverlay.next(true);
            }
        };
        return _this;
    }
    Layout.prototype.componentDidMount = function () {
        var _this = this;
        this.subscribe(this.plugin.layout.events.updated, function () { return _this.forceUpdate(); });
    };
    Layout.prototype.region = function (kind, Element) {
        return (0, jsx_runtime_1.jsx)("div", { className: "msp-layout-region msp-layout-".concat(kind), children: (0, jsx_runtime_1.jsx)("div", { className: 'msp-layout-static', children: Element ? (0, jsx_runtime_1.jsx)(Element, {}) : null }) });
    };
    Object.defineProperty(Layout.prototype, "layoutVisibilityClassName", {
        get: function () {
            var _a, _b;
            var layout = this.plugin.layout.state;
            var controls = (_b = (_a = this.plugin.spec.components) === null || _a === void 0 ? void 0 : _a.controls) !== null && _b !== void 0 ? _b : {};
            var classList = [];
            if (controls.top === 'none' || !layout.showControls || layout.regionState.top === 'hidden') {
                classList.push('msp-layout-hide-top');
            }
            if (controls.left === 'none' || !layout.showControls || layout.regionState.left === 'hidden') {
                classList.push('msp-layout-hide-left');
            }
            else if (layout.regionState.left === 'collapsed') {
                classList.push('msp-layout-collapse-left');
            }
            if (controls.right === 'none' || !layout.showControls || layout.regionState.right === 'hidden') {
                classList.push('msp-layout-hide-right');
            }
            if (controls.bottom === 'none' || !layout.showControls || layout.regionState.bottom === 'hidden') {
                classList.push('msp-layout-hide-bottom');
            }
            return classList.join(' ');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Layout.prototype, "layoutClassName", {
        get: function () {
            var layout = this.plugin.layout.state;
            var classList = ['msp-plugin-content'];
            if (layout.isExpanded) {
                classList.push('msp-layout-expanded');
            }
            else {
                classList.push('msp-layout-standard', "msp-layout-standard-".concat(layout.controlsDisplay));
            }
            return classList.join(' ');
        },
        enumerable: false,
        configurable: true
    });
    Layout.prototype.render = function () {
        var _a, _b, _c, _d, _e, _f, _g;
        var layout = this.plugin.layout.state;
        var controls = ((_a = this.plugin.spec.components) === null || _a === void 0 ? void 0 : _a.controls) || {};
        var viewport = ((_c = (_b = this.plugin.spec.components) === null || _b === void 0 ? void 0 : _b.viewport) === null || _c === void 0 ? void 0 : _c.view) || DefaultViewport;
        var sequenceView = ((_e = (_d = this.plugin.spec.components) === null || _d === void 0 ? void 0 : _d.sequenceViewer) === null || _e === void 0 ? void 0 : _e.view) || sequence_1.SequenceView;
        return (0, jsx_runtime_1.jsx)("div", { className: 'msp-plugin', children: (0, jsx_runtime_1.jsxs)("div", { className: this.layoutClassName, onDragEnter: this.onDragEnter, children: [(0, jsx_runtime_1.jsxs)("div", { className: this.layoutVisibilityClassName, children: [this.region('main', viewport), layout.showControls && controls.top !== 'none' && this.region('top', controls.top || sequenceView), layout.showControls && controls.left !== 'none' && this.region('left', controls.left || left_panel_1.LeftPanelControls), layout.showControls && controls.right !== 'none' && this.region('right', controls.right || ControlsWrapper), layout.showControls && controls.bottom !== 'none' && this.region('bottom', controls.bottom || Log)] }), !((_f = this.plugin.spec.components) === null || _f === void 0 ? void 0 : _f.hideTaskOverlay) && (0, jsx_runtime_1.jsx)(task_1.OverlayTaskProgress, {}), !((_g = this.plugin.spec.components) === null || _g === void 0 ? void 0 : _g.disableDragOverlay) && (0, jsx_runtime_1.jsx)(DragOverlay, { plugin: this.plugin, showDragOverlay: this.showDragOverlay })] }) });
    };
    return Layout;
}(base_1.PluginUIComponent));
function dropFiles(ev, plugin, showDragOverlay) {
    ev.preventDefault();
    ev.stopPropagation();
    showDragOverlay.next(false);
    var files = [];
    if (ev.dataTransfer.items) {
        // Use DataTransferItemList interface to access the file(s)
        for (var i = 0; i < ev.dataTransfer.items.length; i++) {
            if (ev.dataTransfer.items[i].kind !== 'file')
                continue;
            var file = ev.dataTransfer.items[i].getAsFile();
            if (file)
                files.push(file);
        }
    }
    else {
        for (var i = 0; i < ev.dataTransfer.files.length; i++) {
            var file = ev.dataTransfer.files[i];
            if (file)
                files.push(file);
        }
    }
    plugin.managers.dragAndDrop.handle(files);
}
function DragOverlay(_a) {
    var plugin = _a.plugin, showDragOverlay = _a.showDragOverlay;
    var show = (0, use_behavior_1.useBehavior)(showDragOverlay);
    var preventDrag = function (e) {
        e.dataTransfer.dropEffect = 'copy';
        e.preventDefault();
        e.stopPropagation();
    };
    return (0, jsx_runtime_1.jsx)("div", { className: 'msp-drag-drop-overlay', style: { display: show ? 'flex' : 'none' }, onDragEnter: preventDrag, onDragOver: preventDrag, onDragLeave: function () { return showDragOverlay.next(false); }, onDrop: function (e) { return dropFiles(e, plugin, showDragOverlay); }, children: "Load File(s)" });
}
var ControlsWrapper = /** @class */ (function (_super) {
    tslib_1.__extends(ControlsWrapper, _super);
    function ControlsWrapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ControlsWrapper.prototype.render = function () {
        var _a;
        var StructureTools = ((_a = this.plugin.spec.components) === null || _a === void 0 ? void 0 : _a.structureTools) || controls_1.DefaultStructureTools;
        return (0, jsx_runtime_1.jsx)("div", { className: 'msp-scrollable-container', children: (0, jsx_runtime_1.jsx)(StructureTools, {}) });
    };
    return ControlsWrapper;
}(base_1.PluginUIComponent));
exports.ControlsWrapper = ControlsWrapper;
var DefaultViewport = /** @class */ (function (_super) {
    tslib_1.__extends(DefaultViewport, _super);
    function DefaultViewport() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DefaultViewport.prototype.render = function () {
        var _a, _b;
        var VPControls = ((_b = (_a = this.plugin.spec.components) === null || _a === void 0 ? void 0 : _a.viewport) === null || _b === void 0 ? void 0 : _b.controls) || viewport_1.ViewportControls;
        return (0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(viewport_1.Viewport, {}), (0, jsx_runtime_1.jsxs)("div", { className: 'msp-viewport-top-left-controls', children: [(0, jsx_runtime_1.jsx)(controls_1.AnimationViewportControls, {}), (0, jsx_runtime_1.jsx)(controls_1.TrajectoryViewportControls, {}), (0, jsx_runtime_1.jsx)(controls_1.StateSnapshotViewportControls, {}), (0, jsx_runtime_1.jsx)(controls_1.ViewportSnapshotDescription, {})] }), (0, jsx_runtime_1.jsx)(controls_1.SelectionViewportControls, {}), (0, jsx_runtime_1.jsx)(VPControls, {}), (0, jsx_runtime_1.jsx)(task_1.BackgroundTaskProgress, {}), (0, jsx_runtime_1.jsxs)("div", { className: 'msp-highlight-toast-wrapper', children: [(0, jsx_runtime_1.jsx)(controls_1.LociLabels, {}), (0, jsx_runtime_1.jsx)(toast_1.Toasts, {})] })] });
    };
    return DefaultViewport;
}(base_1.PluginUIComponent));
exports.DefaultViewport = DefaultViewport;
var Log = /** @class */ (function (_super) {
    tslib_1.__extends(Log, _super);
    function Log() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.wrapper = React.createRef();
        _this.state = { entries: _this.plugin.log.entries };
        return _this;
    }
    Log.prototype.componentDidMount = function () {
        var _this = this;
        this.subscribe(this.plugin.events.log, function () { return _this.setState({ entries: _this.plugin.log.entries }); });
    };
    Log.prototype.componentDidUpdate = function () {
        this.scrollToBottom();
    };
    Log.prototype.scrollToBottom = function () {
        var log = this.wrapper.current;
        if (log)
            log.scrollTop = log.scrollHeight - log.clientHeight - 1;
    };
    Log.prototype.render = function () {
        // TODO: ability to show full log
        // showing more entries dramatically slows animations.
        var maxEntries = 10;
        var xs = this.state.entries, l = xs.size;
        var entries = [];
        for (var i = Math.max(0, l - maxEntries), o = 0; i < l; i++) {
            var e = xs.get(i);
            entries.push((0, jsx_runtime_1.jsxs)("li", { children: [(0, jsx_runtime_1.jsx)("div", { className: 'msp-log-entry-badge msp-log-entry-' + e.type }), (0, jsx_runtime_1.jsx)("div", { className: 'msp-log-timestamp', children: (0, mol_util_1.formatTime)(e.timestamp) }), (0, jsx_runtime_1.jsx)("div", { className: 'msp-log-entry', children: e.message })] }, o++));
        }
        return (0, jsx_runtime_1.jsx)("div", { ref: this.wrapper, className: 'msp-log', style: { position: 'absolute', top: '0', right: '0', bottom: '0', left: '0', overflowY: 'auto' }, children: (0, jsx_runtime_1.jsx)("ul", { className: 'msp-list-unstyled', children: entries }) });
    };
    return Log;
}(base_1.PluginUIComponent));
exports.Log = Log;
