// molstar/lib/commonjs/extensions/mvs/components/custom-label/visual.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomLabelTextVisual = exports.CustomLabelTextParams = void 0;
var tslib_1 = require("tslib");
var int_1 = require("../../../../mol-data/int");
var text_builder_1 = require("../../../../mol-geo/geometry/text/text-builder");
var complex_visual_1 = require("../../../../mol-repr/structure/complex-visual");
var Original = tslib_1.__importStar(require("../../../../mol-repr/structure/visual/label-text"));
var element_1 = require("../../../../mol-repr/structure/visual/util/element");
var mol_util_1 = require("../../../../mol-util");
var names_1 = require("../../../../mol-util/color/names");
var object_1 = require("../../../../mol-util/object");
var param_definition_1 = require("../../../../mol-util/param-definition");
var label_text_1 = require("../../helpers/label-text");
var selector_1 = require("../selector");
exports.CustomLabelTextParams = tslib_1.__assign(tslib_1.__assign({ items: param_definition_1.ParamDefinition.ObjectList({
        text: param_definition_1.ParamDefinition.Text('¯\\_(ツ)_/¯'),
        position: param_definition_1.ParamDefinition.MappedStatic('selection', {
            x_y_z: param_definition_1.ParamDefinition.Group({
                x: param_definition_1.ParamDefinition.Numeric(0),
                y: param_definition_1.ParamDefinition.Numeric(0),
                z: param_definition_1.ParamDefinition.Numeric(0),
                scale: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 20, step: 0.1 })
            }),
            selection: param_definition_1.ParamDefinition.Group({
                selector: selector_1.SelectorParams,
            }),
        }),
    }, function (obj) { return obj.text; }, { isEssential: true }) }, (0, object_1.omitObjectKeys)(Original.LabelTextParams, ['level', 'chainScale', 'residueScale', 'elementScale'])), { borderColor: tslib_1.__assign(tslib_1.__assign({}, Original.LabelTextParams.borderColor), { defaultValue: names_1.ColorNames.black }) });
/** Create "label-text" visual for "Custom Label" representation */
function CustomLabelTextVisual(materialId) {
    return (0, complex_visual_1.ComplexTextVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.CustomLabelTextParams),
        createGeometry: createLabelText,
        createLocationIterator: element_1.ElementIterator.fromStructure,
        getLoci: element_1.getSerialElementLoci,
        eachLocation: element_1.eachSerialElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = !(0, mol_util_1.deepEqual)(newProps.items, currentProps.items);
        }
    }, materialId);
}
exports.CustomLabelTextVisual = CustomLabelTextVisual;
function createLabelText(ctx, structure, theme, props, text) {
    var _a;
    var count = props.items.length;
    var builder = text_builder_1.TextBuilder.create(props, count, count / 2, text);
    for (var _i = 0, _b = props.items; _i < _b.length; _i++) {
        var item = _b[_i];
        switch (item.position.name) {
            case 'x_y_z':
                var scale = item.position.params.scale;
                builder.add(item.text, item.position.params.x, item.position.params.y, item.position.params.z, scale, scale, 0);
                break;
            case 'selection':
                var substructure = (0, selector_1.substructureFromSelector)(structure, item.position.params.selector);
                var p = (0, label_text_1.textPropsForSelection)(substructure, theme.size.size, {});
                var group = (_a = serialIndexOfSubstructure(structure, substructure)) !== null && _a !== void 0 ? _a : 0;
                if (p)
                    builder.add(item.text, p.center[0], p.center[1], p.center[2], p.depth, p.scale, group);
                break;
        }
    }
    return builder.getText();
}
/** Return the serial index within `structure` of the first element of `substructure` (or `undefined` in that element is not in `structure`)  */
function serialIndexOfSubstructure(structure, substructure) {
    if (substructure.isEmpty)
        return undefined;
    var theUnit = substructure.units[0];
    var theElement = theUnit.elements[0];
    for (var _i = 0, _a = structure.units; _i < _a.length; _i++) {
        var unit = _a[_i];
        if (unit.model.id === theUnit.model.id && int_1.SortedArray.has(unit.elements, theElement)) {
            return structure.serialMapping.getSerialIndex(unit, theElement);
        }
    }
    return undefined;
}
