// molstar/lib/extensions/mvs/behavior.js
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __awaiter, __extends, __generator } from "tslib";
import { PluginBehavior } from '../../mol-plugin/behavior/behavior';
import { ParamDefinition as PD } from '../../mol-util/param-definition';
import { MVSAnnotationColorThemeProvider } from './components/annotation-color-theme';
import { MVSAnnotationLabelRepresentationProvider } from './components/annotation-label/representation';
import { MVSAnnotationsProvider } from './components/annotation-prop';
import { MVSAnnotationTooltipsLabelProvider, MVSAnnotationTooltipsProvider } from './components/annotation-tooltips-prop';
import { CustomLabelRepresentationProvider } from './components/custom-label/representation';
import { CustomTooltipsLabelProvider, CustomTooltipsProvider } from './components/custom-tooltips-prop';
import { LoadMvsData, MVSJFormatProvider } from './components/formats';
import { IsMVSModelProvider } from './components/is-mvs-model-prop';
import { makeMultilayerColorThemeProvider } from './components/multilayer-color-theme';
import { loadMVS } from './load';
import { MVSData } from './mvs-data';
/** Registers everything needed for loading MolViewSpec files */
export var MolViewSpec = PluginBehavior.create({
    name: 'molviewspec',
    category: 'misc',
    display: {
        name: 'MolViewSpec',
        description: 'MolViewSpec extension',
    },
    ctor: /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.registrables = {
                customModelProperties: [
                    IsMVSModelProvider,
                    MVSAnnotationsProvider,
                ],
                customStructureProperties: [
                    CustomTooltipsProvider,
                    MVSAnnotationTooltipsProvider,
                ],
                representations: [
                    CustomLabelRepresentationProvider,
                    MVSAnnotationLabelRepresentationProvider,
                ],
                colorThemes: [
                    MVSAnnotationColorThemeProvider,
                    makeMultilayerColorThemeProvider(_this.ctx.representation.structure.themes.colorThemeRegistry),
                ],
                lociLabels: [
                    CustomTooltipsLabelProvider,
                    MVSAnnotationTooltipsLabelProvider,
                ],
                dragAndDropHandlers: [
                    MVSDragAndDropHandler,
                ],
                dataFormats: [
                    { name: 'MVSJ', provider: MVSJFormatProvider },
                ],
                actions: [
                    LoadMvsData,
                ]
            };
            return _this;
        }
        class_1.prototype.register = function () {
            var _a, _b, _c, _d, _e, _f, _g, _h;
            for (var _i = 0, _j = (_a = this.registrables.customModelProperties) !== null && _a !== void 0 ? _a : []; _i < _j.length; _i++) {
                var prop = _j[_i];
                this.ctx.customModelProperties.register(prop, this.params.autoAttach);
            }
            for (var _k = 0, _l = (_b = this.registrables.customStructureProperties) !== null && _b !== void 0 ? _b : []; _k < _l.length; _k++) {
                var prop = _l[_k];
                this.ctx.customStructureProperties.register(prop, this.params.autoAttach);
            }
            for (var _m = 0, _o = (_c = this.registrables.representations) !== null && _c !== void 0 ? _c : []; _m < _o.length; _m++) {
                var repr = _o[_m];
                this.ctx.representation.structure.registry.add(repr);
            }
            for (var _p = 0, _q = (_d = this.registrables.colorThemes) !== null && _d !== void 0 ? _d : []; _p < _q.length; _p++) {
                var theme = _q[_p];
                this.ctx.representation.structure.themes.colorThemeRegistry.add(theme);
            }
            for (var _r = 0, _s = (_e = this.registrables.lociLabels) !== null && _e !== void 0 ? _e : []; _r < _s.length; _r++) {
                var provider = _s[_r];
                this.ctx.managers.lociLabels.addProvider(provider);
            }
            for (var _t = 0, _u = (_f = this.registrables.dragAndDropHandlers) !== null && _f !== void 0 ? _f : []; _t < _u.length; _t++) {
                var handler = _u[_t];
                this.ctx.managers.dragAndDrop.addHandler(handler.name, handler.handle);
            }
            for (var _v = 0, _w = (_g = this.registrables.dataFormats) !== null && _g !== void 0 ? _g : []; _v < _w.length; _v++) {
                var format = _w[_v];
                this.ctx.dataFormats.add(format.name, format.provider);
            }
            for (var _x = 0, _y = (_h = this.registrables.actions) !== null && _h !== void 0 ? _h : []; _x < _y.length; _x++) {
                var action = _y[_x];
                this.ctx.state.data.actions.add(action);
            }
        };
        class_1.prototype.update = function (p) {
            var _a, _b;
            var updated = this.params.autoAttach !== p.autoAttach;
            this.params.autoAttach = p.autoAttach;
            for (var _i = 0, _c = (_a = this.registrables.customModelProperties) !== null && _a !== void 0 ? _a : []; _i < _c.length; _i++) {
                var prop = _c[_i];
                this.ctx.customModelProperties.setDefaultAutoAttach(prop.descriptor.name, this.params.autoAttach);
            }
            for (var _d = 0, _e = (_b = this.registrables.customStructureProperties) !== null && _b !== void 0 ? _b : []; _d < _e.length; _d++) {
                var prop = _e[_d];
                this.ctx.customStructureProperties.setDefaultAutoAttach(prop.descriptor.name, this.params.autoAttach);
            }
            return updated;
        };
        class_1.prototype.unregister = function () {
            var _a, _b, _c, _d, _e, _f, _g, _h;
            for (var _i = 0, _j = (_a = this.registrables.customModelProperties) !== null && _a !== void 0 ? _a : []; _i < _j.length; _i++) {
                var prop = _j[_i];
                this.ctx.customModelProperties.unregister(prop.descriptor.name);
            }
            for (var _k = 0, _l = (_b = this.registrables.customStructureProperties) !== null && _b !== void 0 ? _b : []; _k < _l.length; _k++) {
                var prop = _l[_k];
                this.ctx.customStructureProperties.unregister(prop.descriptor.name);
            }
            for (var _m = 0, _o = (_c = this.registrables.representations) !== null && _c !== void 0 ? _c : []; _m < _o.length; _m++) {
                var repr = _o[_m];
                this.ctx.representation.structure.registry.remove(repr);
            }
            for (var _p = 0, _q = (_d = this.registrables.colorThemes) !== null && _d !== void 0 ? _d : []; _p < _q.length; _p++) {
                var theme = _q[_p];
                this.ctx.representation.structure.themes.colorThemeRegistry.remove(theme);
            }
            for (var _r = 0, _s = (_e = this.registrables.lociLabels) !== null && _e !== void 0 ? _e : []; _r < _s.length; _r++) {
                var labelProvider = _s[_r];
                this.ctx.managers.lociLabels.removeProvider(labelProvider);
            }
            for (var _t = 0, _u = (_f = this.registrables.dragAndDropHandlers) !== null && _f !== void 0 ? _f : []; _t < _u.length; _t++) {
                var handler = _u[_t];
                this.ctx.managers.dragAndDrop.removeHandler(handler.name);
            }
            for (var _v = 0, _w = (_g = this.registrables.dataFormats) !== null && _g !== void 0 ? _g : []; _v < _w.length; _v++) {
                var format = _w[_v];
                this.ctx.dataFormats.remove(format.name);
            }
            for (var _x = 0, _y = (_h = this.registrables.actions) !== null && _h !== void 0 ? _h : []; _x < _y.length; _x++) {
                var action = _y[_x];
                this.ctx.state.data.actions.remove(action);
            }
        };
        return class_1;
    }(PluginBehavior.Handler)),
    params: function () { return ({
        autoAttach: PD.Boolean(false),
    }); }
});
/** DragAndDropHandler handler for `.mvsj` files */
var MVSDragAndDropHandler = {
    name: 'mvs-mvsj',
    /** Load .mvsj files. Delete previous plugin state before loading.
     * If multiple files are provided, merge their MVS data into one state. */
    handle: function (files, plugin) {
        return __awaiter(this, void 0, void 0, function () {
            var applied, _i, files_1, file, data, mvsData;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        applied = false;
                        _i = 0, files_1 = files;
                        _a.label = 1;
                    case 1:
                        if (!(_i < files_1.length)) return [3 /*break*/, 5];
                        file = files_1[_i];
                        if (!file.name.toLowerCase().endsWith('.mvsj')) return [3 /*break*/, 4];
                        return [4 /*yield*/, file.text()];
                    case 2:
                        data = _a.sent();
                        mvsData = MVSData.fromMVSJ(data);
                        return [4 /*yield*/, loadMVS(plugin, mvsData, { sanityChecks: true, replaceExisting: !applied, sourceUrl: undefined })];
                    case 3:
                        _a.sent();
                        applied = true;
                        _a.label = 4;
                    case 4:
                        _i++;
                        return [3 /*break*/, 1];
                    case 5: return [2 /*return*/, applied];
                }
            });
        });
    },
};
