// molstar/lib/extensions/mvs/load-helpers.js
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __assign, __awaiter, __generator } from "tslib";
import { Mat3, Mat4, Vec3 } from '../../mol-math/linear-algebra';
import { StructureRepresentation3D } from '../../mol-plugin-state/transforms/representation';
import { arrayDistinct } from '../../mol-util/array';
import { canonicalJsonString } from '../../mol-util/json';
import { stringToWords } from '../../mol-util/string';
import { MVSAnnotationColorThemeProvider } from './components/annotation-color-theme';
import { MVSAnnotationLabelRepresentationProvider } from './components/annotation-label/representation';
import { MultilayerColorThemeName, NoColor } from './components/multilayer-color-theme';
import { SelectorAll } from './components/selector';
import { rowToExpression, rowsToExpression } from './helpers/selections';
import { decodeColor, isDefined, stringHash } from './helpers/utils';
import { getChildren } from './tree/generic/tree-schema';
import { dfs, formatObject } from './tree/generic/tree-utils';
import { DefaultColor } from './tree/mvs/mvs-defaults';
/** Load a tree into Mol*, by applying loading actions in DFS order and then commiting at once.
 * If `options.replaceExisting`, remove all objects in the current Mol* state; otherwise add to the current state. */
export function loadTree(plugin, tree, loadingActions, context, options) {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var mapping, updateRoot;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    mapping = new Map();
                    updateRoot = UpdateTarget.create(plugin, (_a = options === null || options === void 0 ? void 0 : options.replaceExisting) !== null && _a !== void 0 ? _a : false);
                    if (options === null || options === void 0 ? void 0 : options.replaceExisting) {
                        UpdateTarget.deleteChildren(updateRoot);
                    }
                    dfs(tree, function (node, parent) {
                        var kind = node.kind;
                        var action = loadingActions[kind];
                        if (action) {
                            var updateParent = parent ? mapping.get(parent) : updateRoot;
                            if (updateParent) {
                                var msNode = action(updateParent, node, context);
                                mapping.set(node, msNode);
                            }
                            else {
                                console.warn("No target found for this \"".concat(node.kind, "\" node"));
                                return;
                            }
                        }
                    });
                    return [4 /*yield*/, UpdateTarget.commit(updateRoot)];
                case 1:
                    _b.sent();
                    return [2 /*return*/];
            }
        });
    });
}
export var UpdateTarget = {
    /** Create a new update, with `selector` pointing to the root. */
    create: function (plugin, replaceExisting) {
        var update = plugin.build();
        var msTarget = update.toRoot().selector;
        var refManager = new RefManager(plugin, replaceExisting);
        return { update: update, selector: msTarget, refManager: refManager };
    },
    /** Add a child node to `target.selector`, return a new `UpdateTarget` pointing to the new child. */
    apply: function (target, transformer, params, options) {
        var _a, _b;
        var refSuffix = transformer.id;
        if (transformer.id === StructureRepresentation3D.id) {
            var reprType = (_b = (_a = params === null || params === void 0 ? void 0 : params.type) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : '';
            refSuffix += ":".concat(reprType);
        }
        var ref = target.refManager.getChildRef(target.selector, refSuffix);
        var msResult = target.update.to(target.selector).apply(transformer, params, __assign(__assign({}, options), { ref: ref })).selector;
        return __assign(__assign({}, target), { selector: msResult });
    },
    /** Delete all children of `target.selector`. */
    deleteChildren: function (target) {
        var children = target.update.currentTree.children.get(target.selector.ref);
        children.forEach(function (child) { return target.update.delete(child); });
        return target;
    },
    /** Commit all changes done in the current update. */
    commit: function (target) {
        return target.update.commit();
    },
};
/** Manages transform refs in a deterministic way. Uses refs like !mvs:3ce3664304d32c5d:0 */
var RefManager = /** @class */ (function () {
    function RefManager(plugin, replaceExisting) {
        var _this = this;
        /** For each hash (e.g. 3ce3664304d32c5d), store the number of already used refs with that hash. */
        this._counter = {};
        if (!replaceExisting) {
            plugin.state.data.cells.forEach(function (cell) {
                var _a;
                var ref = cell.transform.ref;
                if (ref.startsWith('!mvs:')) {
                    var _b = ref.split(':'), _ = _b[0], hash = _b[1], idNumber = _b[2];
                    var nextIdNumber = parseInt(idNumber) + 1;
                    if (nextIdNumber > ((_a = _this._counter[hash]) !== null && _a !== void 0 ? _a : 0)) {
                        _this._counter[hash] = nextIdNumber;
                    }
                }
            });
        }
    }
    /** Return ref for a new node with given `hash`; update the counter accordingly. */
    RefManager.prototype.nextRef = function (hash) {
        var _a;
        var _b;
        (_a = (_b = this._counter)[hash]) !== null && _a !== void 0 ? _a : (_b[hash] = 0);
        var idNumber = this._counter[hash]++;
        return "!mvs:".concat(hash, ":").concat(idNumber);
    };
    /** Return ref for a new node based on parent and desired suffix. */
    RefManager.prototype.getChildRef = function (parent, suffix) {
        var hashBase = parent.ref.replace(/^!mvs:/, '') + ':' + suffix;
        var hash = stringHash(hashBase);
        var result = this.nextRef(hash);
        return result;
    };
    return RefManager;
}());
export var AnnotationFromUriKinds = new Set(['color_from_uri', 'component_from_uri', 'label_from_uri', 'tooltip_from_uri']);
export var AnnotationFromSourceKinds = new Set(['color_from_source', 'component_from_source', 'label_from_source', 'tooltip_from_source']);
/** Return a 4x4 matrix representing a rotation followed by a translation */
export function transformFromRotationTranslation(rotation, translation) {
    if (rotation && rotation.length !== 9)
        throw new Error("'rotation' param for 'transform' node must be array of 9 elements, found ".concat(rotation));
    if (translation && translation.length !== 3)
        throw new Error("'translation' param for 'transform' node must be array of 3 elements, found ".concat(translation));
    var T = Mat4.identity();
    if (rotation) {
        var rotMatrix = Mat3.fromArray(Mat3(), rotation, 0);
        ensureRotationMatrix(rotMatrix, rotMatrix);
        Mat4.fromMat3(T, rotMatrix);
    }
    if (translation) {
        Mat4.setTranslation(T, Vec3.fromArray(Vec3(), translation, 0));
    }
    if (!Mat4.isRotationAndTranslation(T))
        throw new Error("'rotation' param for 'transform' is not a valid rotation matrix: ".concat(rotation));
    return T;
}
/** Adjust values in a close-to-rotation matrix `a` to ensure it is a proper rotation matrix
 * (i.e. its columns and rows are orthonormal and determinant equal to 1, within available precission). */
function ensureRotationMatrix(out, a) {
    var x = Vec3.fromArray(_tmpVecX, a, 0);
    var y = Vec3.fromArray(_tmpVecY, a, 3);
    var z = Vec3.fromArray(_tmpVecZ, a, 6);
    Vec3.normalize(x, x);
    Vec3.orthogonalize(y, x, y);
    Vec3.normalize(z, Vec3.cross(z, x, y));
    Mat3.fromColumns(out, x, y, z);
    return out;
}
var _tmpVecX = Vec3();
var _tmpVecY = Vec3();
var _tmpVecZ = Vec3();
/** Create an array of props for `TransformStructureConformation` transformers from all 'transform' nodes applied to a 'structure' node. */
export function transformProps(node) {
    var result = [];
    var transforms = getChildren(node).filter(function (c) { return c.kind === 'transform'; });
    for (var _i = 0, transforms_1 = transforms; _i < transforms_1.length; _i++) {
        var transform = transforms_1[_i];
        var _a = transform.params, rotation = _a.rotation, translation = _a.translation;
        var matrix = transformFromRotationTranslation(rotation, translation);
        result.push({ transform: { name: 'matrix', params: { data: matrix, transpose: false } } });
    }
    return result;
}
/** Collect distinct annotation specs from all nodes in `tree` and set `context.annotationMap[node]` to respective annotationIds */
export function collectAnnotationReferences(tree, context) {
    var distinctSpecs = {};
    dfs(tree, function (node) {
        var _a, _b, _c, _d;
        var spec = undefined;
        if (AnnotationFromUriKinds.has(node.kind)) {
            var p = node.params;
            spec = { source: { name: 'url', params: { url: p.uri, format: p.format } }, schema: p.schema, cifBlock: blockSpec(p.block_header, p.block_index), cifCategory: (_a = p.category_name) !== null && _a !== void 0 ? _a : undefined };
        }
        else if (AnnotationFromSourceKinds.has(node.kind)) {
            var p = node.params;
            spec = { source: { name: 'source-cif', params: {} }, schema: p.schema, cifBlock: blockSpec(p.block_header, p.block_index), cifCategory: (_b = p.category_name) !== null && _b !== void 0 ? _b : undefined };
        }
        if (spec) {
            var key = canonicalJsonString(spec);
            (_c = distinctSpecs[key]) !== null && _c !== void 0 ? _c : (distinctSpecs[key] = __assign(__assign({}, spec), { id: stringHash(key) }));
            ((_d = context.annotationMap) !== null && _d !== void 0 ? _d : (context.annotationMap = new Map())).set(node, distinctSpecs[key].id);
        }
    });
    return Object.values(distinctSpecs);
}
function blockSpec(header, index) {
    if (isDefined(header)) {
        return { name: 'header', params: { header: header } };
    }
    else {
        return { name: 'index', params: { index: index !== null && index !== void 0 ? index : 0 } };
    }
}
/** Collect annotation tooltips from all nodes in `tree` and map them to annotationIds. */
export function collectAnnotationTooltips(tree, context) {
    var annotationTooltips = [];
    dfs(tree, function (node) {
        var _a;
        if (node.kind === 'tooltip_from_uri' || node.kind === 'tooltip_from_source') {
            var annotationId = (_a = context.annotationMap) === null || _a === void 0 ? void 0 : _a.get(node);
            if (annotationId) {
                annotationTooltips.push({ annotationId: annotationId, fieldName: node.params.field_name });
            }
            ;
        }
    });
    return arrayDistinct(annotationTooltips);
}
/** Collect inline tooltips from all nodes in `tree`. */
export function collectInlineTooltips(tree, context) {
    var inlineTooltips = [];
    dfs(tree, function (node, parent) {
        if (node.kind === 'tooltip') {
            if ((parent === null || parent === void 0 ? void 0 : parent.kind) === 'component') {
                inlineTooltips.push({
                    text: node.params.text,
                    selector: componentPropsFromSelector(parent.params.selector),
                });
            }
            else if ((parent === null || parent === void 0 ? void 0 : parent.kind) === 'component_from_uri' || (parent === null || parent === void 0 ? void 0 : parent.kind) === 'component_from_source') {
                var p = componentFromXProps(parent, context);
                if (isDefined(p.annotationId) && isDefined(p.fieldName) && isDefined(p.fieldValues)) {
                    inlineTooltips.push({
                        text: node.params.text,
                        selector: {
                            name: 'annotation',
                            params: { annotationId: p.annotationId, fieldName: p.fieldName, fieldValues: p.fieldValues },
                        },
                    });
                }
            }
        }
    });
    return inlineTooltips;
}
/** Collect inline labels from all nodes in `tree`. */
export function collectInlineLabels(tree, context) {
    var inlineLabels = [];
    dfs(tree, function (node, parent) {
        if (node.kind === 'label') {
            if ((parent === null || parent === void 0 ? void 0 : parent.kind) === 'component') {
                inlineLabels.push({
                    text: node.params.text,
                    position: {
                        name: 'selection',
                        params: {
                            selector: componentPropsFromSelector(parent.params.selector),
                        },
                    },
                });
            }
            else if ((parent === null || parent === void 0 ? void 0 : parent.kind) === 'component_from_uri' || (parent === null || parent === void 0 ? void 0 : parent.kind) === 'component_from_source') {
                var p = componentFromXProps(parent, context);
                if (isDefined(p.annotationId) && isDefined(p.fieldName) && isDefined(p.fieldValues)) {
                    inlineLabels.push({
                        text: node.params.text,
                        position: {
                            name: 'selection',
                            params: {
                                selector: {
                                    name: 'annotation',
                                    params: { annotationId: p.annotationId, fieldName: p.fieldName, fieldValues: p.fieldValues },
                                },
                            },
                        },
                    });
                }
            }
        }
    });
    return inlineLabels;
}
/** Return `true` for components nodes which only serve for tooltip placement (not to be created in the MolStar object hierarchy) */
export function isPhantomComponent(node) {
    return node.children && node.children.every(function (child) { return child.kind === 'tooltip' || child.kind === 'label'; });
    // These nodes could theoretically be removed when converting MVS to Molstar tree, but would get very tricky if we allow nested components
}
/** Create props for `StructureFromModel` transformer from a structure node. */
export function structureProps(node) {
    var _a;
    var params = node.params;
    switch (params.type) {
        case 'model':
            return {
                type: {
                    name: 'model',
                    params: {}
                },
            };
        case 'assembly':
            return {
                type: {
                    name: 'assembly',
                    params: { id: (_a = params.assembly_id) !== null && _a !== void 0 ? _a : undefined }
                },
            };
        case 'symmetry':
            return {
                type: {
                    name: 'symmetry',
                    params: { ijkMin: Vec3.ofArray(params.ijk_min), ijkMax: Vec3.ofArray(params.ijk_max) }
                },
            };
        case 'symmetry_mates':
            return {
                type: {
                    name: 'symmetry-mates',
                    params: { radius: params.radius }
                }
            };
        default:
            throw new Error("NotImplementedError: Loading action for \"structure\" node, type \"".concat(params.type, "\""));
    }
}
/** Create value for `type` prop for `StructureComponent` transformer based on a MVS selector. */
export function componentPropsFromSelector(selector) {
    if (selector === undefined) {
        return SelectorAll;
    }
    else if (typeof selector === 'string') {
        return { name: 'static', params: selector };
    }
    else if (Array.isArray(selector)) {
        return { name: 'expression', params: rowsToExpression(selector) };
    }
    else {
        return { name: 'expression', params: rowToExpression(selector) };
    }
}
/** Return a pretty name for a value of selector param, e.g.  "protein" -> 'Protein', {label_asym_id: "A"} -> 'Custom Selection: {label_asym_id: "A"}' */
export function prettyNameFromSelector(selector) {
    if (selector === undefined) {
        return 'All';
    }
    else if (typeof selector === 'string') {
        return stringToWords(selector);
    }
    else if (Array.isArray(selector)) {
        return "Custom Selection: [".concat(selector.map(formatObject).join(', '), "]");
    }
    else {
        return "Custom Selection: ".concat(formatObject(selector));
    }
}
/** Create props for `StructureRepresentation3D` transformer from a label_from_* node. */
export function labelFromXProps(node, context) {
    var _a, _b;
    var annotationId = (_a = context.annotationMap) === null || _a === void 0 ? void 0 : _a.get(node);
    var fieldName = node.params.field_name;
    var nearestReprNode = (_b = context.nearestReprMap) === null || _b === void 0 ? void 0 : _b.get(node);
    return {
        type: { name: MVSAnnotationLabelRepresentationProvider.name, params: { annotationId: annotationId, fieldName: fieldName } },
        colorTheme: colorThemeForNode(nearestReprNode, context),
    };
}
/** Create props for `AnnotationStructureComponent` transformer from a component_from_* node. */
export function componentFromXProps(node, context) {
    var _a;
    var annotationId = (_a = context.annotationMap) === null || _a === void 0 ? void 0 : _a.get(node);
    var _b = node.params, field_name = _b.field_name, field_values = _b.field_values;
    return {
        annotationId: annotationId,
        fieldName: field_name,
        fieldValues: field_values ? { name: 'selected', params: field_values.map(function (v) { return ({ value: v }); }) } : { name: 'all', params: {} },
        nullIfEmpty: false,
    };
}
/** Create props for `StructureRepresentation3D` transformer from a representation node. */
export function representationProps(params) {
    switch (params.type) {
        case 'cartoon':
            return {
                type: { name: 'cartoon', params: {} },
            };
        case 'ball_and_stick':
            return {
                type: { name: 'ball-and-stick', params: { sizeFactor: 0.5, sizeAspectRatio: 0.5 } },
            };
        case 'surface':
            return {
                type: { name: 'molecular-surface', params: {} },
                sizeTheme: { name: 'physical', params: { scale: 1 } },
            };
        default:
            throw new Error('NotImplementedError');
    }
}
/** Create value for `colorTheme` prop for `StructureRepresentation3D` transformer from a representation node based on color* nodes in its subtree. */
export function colorThemeForNode(node, context) {
    var _a;
    if ((node === null || node === void 0 ? void 0 : node.kind) === 'representation') {
        var children = getChildren(node).filter(function (c) { return c.kind === 'color' || c.kind === 'color_from_uri' || c.kind === 'color_from_source'; });
        if (children.length === 0) {
            return {
                name: 'uniform',
                params: { value: decodeColor(DefaultColor) },
            };
        }
        else if (children.length === 1 && appliesColorToWholeRepr(children[0])) {
            return colorThemeForNode(children[0], context);
        }
        else {
            var layers = children.map(function (c) { return ({ theme: colorThemeForNode(c, context), selection: componentPropsFromSelector(c.kind === 'color' ? c.params.selector : undefined) }); });
            return {
                name: MultilayerColorThemeName,
                params: { layers: layers },
            };
        }
    }
    var annotationId = undefined;
    var fieldName = undefined;
    var color = undefined;
    switch (node === null || node === void 0 ? void 0 : node.kind) {
        case 'color_from_uri':
        case 'color_from_source':
            annotationId = (_a = context.annotationMap) === null || _a === void 0 ? void 0 : _a.get(node);
            fieldName = node.params.field_name;
            break;
        case 'color':
            color = node.params.color;
            break;
    }
    if (annotationId) {
        return {
            name: MVSAnnotationColorThemeProvider.name,
            params: { annotationId: annotationId, fieldName: fieldName, background: NoColor },
        };
    }
    else {
        return {
            name: 'uniform',
            params: { value: decodeColor(color) },
        };
    }
}
function appliesColorToWholeRepr(node) {
    if (node.kind === 'color') {
        return !isDefined(node.params.selector) || node.params.selector === 'all';
    }
    else {
        return true;
    }
}
/** Create a mapping of nearest representation nodes for each node in the tree
 * (to transfer coloring to label nodes smartly).
 * Only considers nodes within the same 'structure' subtree. */
export function makeNearestReprMap(root) {
    var map = new Map();
    // Propagate up:
    dfs(root, undefined, function (node, parent) {
        if (node.kind === 'representation') {
            map.set(node, node);
        }
        if (node.kind !== 'structure' && map.has(node) && parent && !map.has(parent)) { // do not propagate above the lowest structure node
            map.set(parent, map.get(node));
        }
    });
    // Propagate down:
    dfs(root, function (node, parent) {
        if (!map.has(node) && parent && map.has(parent)) {
            map.set(node, map.get(parent));
        }
    });
    return map;
}
