// molstar/lib/extensions/mvs/load.js
/**
 * Copyright (c) 2023-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
import { __assign, __awaiter, __generator } from "tslib";
import { Download, ParseCif } from '../../mol-plugin-state/transforms/data';
import { CustomModelProperties, CustomStructureProperties, ModelFromTrajectory, StructureComponent, StructureFromModel, TrajectoryFromMmCif, TrajectoryFromPDB, TransformStructureConformation } from '../../mol-plugin-state/transforms/model';
import { StructureRepresentation3D } from '../../mol-plugin-state/transforms/representation';
import { MolViewSpec } from './behavior';
import { setCamera, setCanvas, setFocus } from './camera';
import { MVSAnnotationsProvider } from './components/annotation-prop';
import { MVSAnnotationStructureComponent } from './components/annotation-structure-component';
import { MVSAnnotationTooltipsProvider } from './components/annotation-tooltips-prop';
import { CustomLabelRepresentationProvider } from './components/custom-label/representation';
import { CustomTooltipsProvider } from './components/custom-tooltips-prop';
import { IsMVSModelProvider } from './components/is-mvs-model-prop';
import { UpdateTarget, collectAnnotationReferences, collectAnnotationTooltips, collectInlineLabels, collectInlineTooltips, colorThemeForNode, componentFromXProps, componentPropsFromSelector, isPhantomComponent, labelFromXProps, loadTree, makeNearestReprMap, prettyNameFromSelector, representationProps, structureProps, transformProps } from './load-helpers';
import { validateTree } from './tree/generic/tree-schema';
import { convertMvsToMolstar, mvsSanityCheck } from './tree/molstar/conversion';
import { MolstarTreeSchema } from './tree/molstar/molstar-tree';
import { MVSTreeSchema } from './tree/mvs/mvs-tree';
/** Load a MolViewSpec (MVS) tree into the Mol* plugin.
 * If `options.replaceExisting`, remove all objects in the current Mol* state; otherwise add to the current state.
 * If `options.sanityChecks`, run some sanity checks and print potential issues to the console.
 * `options.sourceUrl` serves as the base for resolving relative URLs/URIs and may itself be relative to the window URL. */
export function loadMVS(plugin, data, options) {
    if (options === void 0) { options = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var molstarTree, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    // console.log(`MVS tree:\n${MVSData.toPrettyString(data)}`)
                    validateTree(MVSTreeSchema, data.root, 'MVS');
                    if (options.sanityChecks)
                        mvsSanityCheck(data.root);
                    molstarTree = convertMvsToMolstar(data.root, options.sourceUrl);
                    // console.log(`Converted MolStar tree:\n${MVSData.toPrettyString({ root: molstarTree, metadata: { version: 'x', timestamp: 'x' } })}`)
                    validateTree(MolstarTreeSchema, molstarTree, 'Converted Molstar');
                    return [4 /*yield*/, loadMolstarTree(plugin, molstarTree, options)];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    err_1 = _a.sent();
                    plugin.log.error("".concat(err_1));
                    throw err_1;
                case 3: return [2 /*return*/];
            }
        });
    });
}
/** Load a `MolstarTree` into the Mol* plugin.
 * If `replaceExisting`, remove all objects in the current Mol* state; otherwise add to the current state. */
function loadMolstarTree(plugin, tree, options) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function () {
        var mvsExtensionLoaded, context;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    mvsExtensionLoaded = plugin.state.hasBehavior(MolViewSpec);
                    if (!mvsExtensionLoaded)
                        throw new Error('MolViewSpec extension is not loaded.');
                    context = {};
                    return [4 /*yield*/, loadTree(plugin, tree, MolstarLoadingActions, context, options)];
                case 1:
                    _c.sent();
                    setCanvas(plugin, context.canvas);
                    if (!(((_a = context.focus) === null || _a === void 0 ? void 0 : _a.kind) === 'camera')) return [3 /*break*/, 3];
                    return [4 /*yield*/, setCamera(plugin, context.focus.params)];
                case 2:
                    _c.sent();
                    return [3 /*break*/, 7];
                case 3:
                    if (!(((_b = context.focus) === null || _b === void 0 ? void 0 : _b.kind) === 'focus')) return [3 /*break*/, 5];
                    return [4 /*yield*/, setFocus(plugin, context.focus.focusTarget, context.focus.params)];
                case 4:
                    _c.sent();
                    return [3 /*break*/, 7];
                case 5: return [4 /*yield*/, setFocus(plugin, undefined, undefined)];
                case 6:
                    _c.sent();
                    _c.label = 7;
                case 7: return [2 /*return*/];
            }
        });
    });
}
/** Loading actions for loading a `MolstarTree`, per node kind. */
var MolstarLoadingActions = {
    root: function (updateParent, node, context) {
        context.nearestReprMap = makeNearestReprMap(node);
        return updateParent;
    },
    download: function (updateParent, node) {
        return UpdateTarget.apply(updateParent, Download, {
            url: node.params.url,
            isBinary: node.params.is_binary,
        });
    },
    parse: function (updateParent, node) {
        var format = node.params.format;
        if (format === 'cif') {
            return UpdateTarget.apply(updateParent, ParseCif, {});
        }
        else if (format === 'pdb') {
            return updateParent;
        }
        else {
            console.error("Unknown format in \"parse\" node: \"".concat(format, "\""));
            return undefined;
        }
    },
    trajectory: function (updateParent, node) {
        var _a, _b;
        var format = node.params.format;
        if (format === 'cif') {
            return UpdateTarget.apply(updateParent, TrajectoryFromMmCif, {
                blockHeader: (_a = node.params.block_header) !== null && _a !== void 0 ? _a : '', // Must set to '' because just undefined would get overwritten by createDefaults
                blockIndex: (_b = node.params.block_index) !== null && _b !== void 0 ? _b : undefined,
            });
        }
        else if (format === 'pdb') {
            return UpdateTarget.apply(updateParent, TrajectoryFromPDB, {});
        }
        else {
            console.error("Unknown format in \"trajectory\" node: \"".concat(format, "\""));
            return undefined;
        }
    },
    model: function (updateParent, node, context) {
        var _a;
        var annotations = collectAnnotationReferences(node, context);
        var model = UpdateTarget.apply(updateParent, ModelFromTrajectory, {
            modelIndex: node.params.model_index,
        });
        UpdateTarget.apply(model, CustomModelProperties, {
            properties: (_a = {},
                _a[IsMVSModelProvider.descriptor.name] = { isMvs: true },
                _a[MVSAnnotationsProvider.descriptor.name] = { annotations: annotations },
                _a),
            autoAttach: [
                IsMVSModelProvider.descriptor.name,
                MVSAnnotationsProvider.descriptor.name,
            ],
        });
        return model;
    },
    structure: function (updateParent, node, context) {
        var _a;
        var _b;
        var props = structureProps(node);
        var struct = UpdateTarget.apply(updateParent, StructureFromModel, props);
        var transformed = struct;
        for (var _i = 0, _c = transformProps(node); _i < _c.length; _i++) {
            var t = _c[_i];
            transformed = UpdateTarget.apply(transformed, TransformStructureConformation, t); // applying to the result of previous transform, to get the correct transform order
        }
        var annotationTooltips = collectAnnotationTooltips(node, context);
        var inlineTooltips = collectInlineTooltips(node, context);
        if (annotationTooltips.length + inlineTooltips.length > 0) {
            UpdateTarget.apply(struct, CustomStructureProperties, {
                properties: (_a = {},
                    _a[MVSAnnotationTooltipsProvider.descriptor.name] = { tooltips: annotationTooltips },
                    _a[CustomTooltipsProvider.descriptor.name] = { tooltips: inlineTooltips },
                    _a),
                autoAttach: [
                    MVSAnnotationTooltipsProvider.descriptor.name,
                    CustomTooltipsProvider.descriptor.name,
                ],
            });
        }
        var inlineLabels = collectInlineLabels(node, context);
        if (inlineLabels.length > 0) {
            var nearestReprNode = (_b = context.nearestReprMap) === null || _b === void 0 ? void 0 : _b.get(node);
            UpdateTarget.apply(struct, StructureRepresentation3D, {
                type: {
                    name: CustomLabelRepresentationProvider.name,
                    params: { items: inlineLabels },
                },
                colorTheme: colorThemeForNode(nearestReprNode, context),
            });
        }
        return struct;
    },
    tooltip: undefined, // No action needed, already loaded in `structure`
    tooltip_from_uri: undefined, // No action needed, already loaded in `structure`
    tooltip_from_source: undefined, // No action needed, already loaded in `structure`
    component: function (updateParent, node) {
        if (isPhantomComponent(node)) {
            return updateParent;
        }
        var selector = node.params.selector;
        return UpdateTarget.apply(updateParent, StructureComponent, {
            type: componentPropsFromSelector(selector),
            label: prettyNameFromSelector(selector),
            nullIfEmpty: false,
        });
    },
    component_from_uri: function (updateParent, node, context) {
        if (isPhantomComponent(node))
            return undefined;
        var props = componentFromXProps(node, context);
        return UpdateTarget.apply(updateParent, MVSAnnotationStructureComponent, props);
    },
    component_from_source: function (updateParent, node, context) {
        if (isPhantomComponent(node))
            return undefined;
        var props = componentFromXProps(node, context);
        return UpdateTarget.apply(updateParent, MVSAnnotationStructureComponent, props);
    },
    representation: function (updateParent, node, context) {
        return UpdateTarget.apply(updateParent, StructureRepresentation3D, __assign(__assign({}, representationProps(node.params)), { colorTheme: colorThemeForNode(node, context) }));
    },
    color: undefined, // No action needed, already loaded in `representation`
    color_from_uri: undefined, // No action needed, already loaded in `representation`
    color_from_source: undefined, // No action needed, already loaded in `representation`
    label: undefined, // No action needed, already loaded in `structure`
    label_from_uri: function (updateParent, node, context) {
        var props = labelFromXProps(node, context);
        return UpdateTarget.apply(updateParent, StructureRepresentation3D, props);
    },
    label_from_source: function (updateParent, node, context) {
        var props = labelFromXProps(node, context);
        return UpdateTarget.apply(updateParent, StructureRepresentation3D, props);
    },
    focus: function (updateParent, node, context) {
        context.focus = { kind: 'focus', focusTarget: updateParent.selector, params: node.params };
        return updateParent;
    },
    camera: function (updateParent, node, context) {
        context.focus = { kind: 'camera', params: node.params };
        return updateParent;
    },
    canvas: function (updateParent, node, context) {
        context.canvas = node.params;
        return updateParent;
    },
};
